<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Effectively using the DelayedArray framework as a user to support the analysis of large datasets • DelayedArrayWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Effectively using the DelayedArray framework as a user to support the analysis of large datasets">
<meta property="og:description" content="DelayedArrayWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DelayedArrayWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Effectively_using_the_DelayedArray_framework_for_users.html">Effectively using the DelayedArray framework as a user to support the analysis of large datasets</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PeteHaitch/BioC2020_DelayedArray_workshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Effectively_using_the_DelayedArray_framework_for_users_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="Effectively_using_the_DelayedArray_framework_for_users_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="Effectively_using_the_DelayedArray_framework_for_users_files/jquery-1.12.4/jquery.min.js"></script><link href="Effectively_using_the_DelayedArray_framework_for_users_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="Effectively_using_the_DelayedArray_framework_for_users_files/datatables-binding-0.14/datatables.js"></script><link href="Effectively_using_the_DelayedArray_framework_for_users_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="Effectively_using_the_DelayedArray_framework_for_users_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="Effectively_using_the_DelayedArray_framework_for_users_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="Effectively_using_the_DelayedArray_framework_for_users_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="Effectively_using_the_DelayedArray_framework_for_users_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Effectively using the DelayedArray framework as a user to support the analysis of large datasets</h1>
            <h3 class="subtitle">Presented at BioC 2020 (27-31 July)</h3>
                        <h4 class="author">Peter Hickey</h4>
            
            <h4 class="date">Last modified: July 28, 2020; Compiled: July 28, 2020</h4>
      
      
      <div class="hidden name"><code>Effectively_using_the_DelayedArray_framework_for_users.Rmd</code></div>

    </div>

    
    
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop description</h2>
<p>This workshop gives an introductory overview of the DelayedArray framework, which can be used by R / Bioconductor packages to support the analysis of large array-like datasets. A <em>DelayedArray</em> is like an ordinary array in R, but allows for the data to be in-memory, on-disk in a file, or even hosted on a remote server.</p>
<p>Workshop participants will learn where they might encounter a <em>DelayedArray</em> in the wild while using Bioconductor and help them understand the fundamental concepts underlying the framework. This workshop will feature <a href="https://docs.google.com/presentation/d/1_v4IuKLFs781pp7x0BwUxv7mwCfPLBTWu320g3lQmME/edit?usp=sharing">introductory material</a>, ‘live’ coding, and Q&amp;A.</p>
<div id="instructor" class="section level3">
<h3 class="hasAnchor">
<a href="#instructor" class="anchor"></a>Instructor</h3>
<ul>
<li>
<a href="https://peterhickey.org/">Peter Hickey</a> (<a href="mailto:hickey@wehi.edu.au" class="email">hickey@wehi.edu.au</a>)</li>
</ul>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li>Basic knowledge of R syntax.</li>
<li>Familiarity with common operations on matrices in R, such as <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code> and <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMeans()</a></code>.</li>
<li>Some familiarity with S4 objects may be helpful but is not required.</li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h3>
<p>Students will be able to run code examples from the workshop material. There will be a Q&amp;A session in the second half of the workshop.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<p>These packages are the focus of this workshop:</p>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em></li>
</ul>
<p>Please see the workshop <code>DESCRIPTION</code> for a full list of dependencies.</p>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introductory material</td>
<td>5 min</td>
</tr>
<tr class="even">
<td>First contact</td>
<td>20 min</td>
</tr>
<tr class="odd">
<td>Workflow tips for DelayedArray-backed analyses</td>
<td>15 min</td>
</tr>
<tr class="even">
<td>Q&amp;A</td>
<td>15 min</td>
</tr>
</tbody>
</table>
</div>
<div id="workshop-goals-and-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h3>
<div id="learning-goals" class="section level4">
<h4 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h4>
<ul>
<li>Learn of existing packages and functions that <em>use</em> the DelayedArray framework.</li>
<li>Develop a high-level understanding of classes and packages that <em>implement</em> the DelayedArray framework.</li>
<li>Become familiar with the fundamental concepts of delayed operations, block processing, and realization.</li>
<li>Reason about potential bottlenecks, and how to avoid or reduce these, in algorithms operating on <em>DelayedArray</em> objects.</li>
</ul>
</div>
<div id="learning-objectives" class="section level4">
<h4 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h4>
<ul>
<li>Identify when an object is a <em>DelayedArray</em> or one of its derivatives.</li>
<li>Be able to recognise when it is useful to use a <em>DelayedArray</em> instead of an ordinary array or other array-like data structure.</li>
<li>Learn how to load and save a DelayedArray-backed object.</li>
<li>Learn how the ‘block size’ and ‘chunking’ of the dataset affect performance when operating on <em>DelayedArray</em> objects.</li>
<li>Take away some miscellaneous tips and tricks I’ve learnt over the years when working with DelayedArray-backed objects.</li>
</ul>
</div>
</div>
</div>
<div id="introductory-material" class="section level2">
<h2 class="hasAnchor">
<a href="#introductory-material" class="anchor"></a>Introductory material</h2>
<p>Data from a high-throughput biological assay, such as single-cell RNA-sequencing (scRNA-seq), will often be summarised as a matrix of counts, where rows correspond to features and columns to samples<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. Within <strong>Bioconductor</strong>, the <em>SummarizedExperiment</em> class is the recommended container for such data, offering a rich interface that tightly links assay measurements to data on the features and the samples.</p>
<p>The <em>SummarizedExperiment</em> class is used to store rectangular arrays of experimental results (<em>assays</em>). Here, each <em>assay</em> is drawn as a matrix but higher-dimensional arrays are also supported.</p>
<p><img src="images/SummarizedExperiment.svg"><!-- --></p>
<p>Traditionally, the assay data are stored in-memory as an ordinary <em>array</em> object<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. Storing the data in-memory becomes a real pain with the ever-growing size of ’omics datasets. It is now not uncommon to collect <span class="math inline">\(10,000-100,000,000\)</span> measurements on <span class="math inline">\(100 - 1,000,000\)</span> samples, which would occupy <span class="math inline">\(10-1,000\)</span> gigabytes (Gb) if stored in-memory as ordinary R arrays.</p>
<p>The DelayedArray framework offers a solution to this problem. Wrapping an array-like object (typically an on-disk object) in a <em>DelayedArray</em> object allows one to perform common array operations on it without loading the object in memory. In order to reduce memory usage and optimize performance, operations on the object are either delayed or executed using a block processing mechanism.</p>
<div id="projects-enabled-by-delayedarray" class="section level3">
<h3 class="hasAnchor">
<a href="#projects-enabled-by-delayedarray" class="anchor"></a>Projects enabled by DelayedArray</h3>
<p>The DelayedArray framework enables the analysis of datasets that are too large to be stored or processed in-memory. This has become particularly relevant with the advent of large single-cell RNA-sequencing (scRNA-seq) studies containing tens of thousands to millions of cells.</p>
<p>In my own research I have made extensive use of the DelayedArray framework when analysing whole genome bisulfite sequencing (WGBS) datasets. To give a recent example, we profiled 45 human brain samples using WGBS to measure DNA methylation <span class="citation">(Rizzardi et al. 2019)</span>. For most tissues, we focus on so-called so-called CpG methylation, which requires analysing matrices with roughly 20 million rows (CpG loci) and 45 columns (samples). This sized data is challenging, but well within the realms of high performance computing available at a modern research institute. However, the brain also has extensive non-CpG methylation and there are an order of magnitude more non-CpG loci. This necessitated extensive re-factoring of our software tools and we successfully adopted the DelayedArray framework to enable this research.</p>
</div>
<div id="package-ecosystem" class="section level3">
<h3 class="hasAnchor">
<a href="#package-ecosystem" class="anchor"></a>Package ecosystem</h3>
<p>The DelayedArray framework is, unsurprisingly, implemented in the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package. However, there are several other key packages that are an important part of the broader ‘ecosystem’. More importantly, as a user of Bioconductor software, it is increasingly likely that you will encounter <em>DelayedArray</em> objects during a data analysis, especially if you are analysing single-cell data<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. The following table lists packages that depend upon the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="no">dep_tbl</span> <span class="kw">&lt;-</span> <span class="kw pkg">BiocPkgTools</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocPkgTools/man/buildPkgDependencyDataFrame.html">buildPkgDependencyDataFrame</a></span>()
<span class="no">da_dep_tbl</span> <span class="kw">&lt;-</span> <span class="no">dep_tbl</span>[<span class="no">dep_tbl</span>$<span class="no">dependency</span> <span class="kw">==</span> <span class="st">"DelayedArray"</span>,
                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Package"</span>, <span class="st">"edgetype"</span>)]
<span class="no">da_dep_tbl</span> <span class="kw">&lt;-</span> <span class="no">da_dep_tbl</span>[<span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span>(<span class="no">da_dep_tbl</span>, <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="no">edgetype</span>, <span class="no">Package</span>)), ]
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">da_dep_tbl</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Package"</span>, <span class="st">"Dependency Type"</span>)
<span class="kw pkg">DT</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span>(<span class="no">da_dep_tbl</span>)</pre></body></html></div>
<div id="htmlwidget-f6b798897adcf402f11d" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-f6b798897adcf402f11d">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52"],["DelayedDataFrame","DelayedMatrixStats","GDSArray","HDF5Array","restfulSE","rhdf5client","singleCellTK","SummarizedExperiment","VCFArray","batchelor","beachmat","bigPint","BiocSingular","bsseq","CAGEr","celaref","celda","clusterExperiment","DEScan2","DSS","ELMER","FRASER","GenoGAM","GenomicScores","glmGamPoi","gramm4R","hipathia","LoomExperiment","mbkmeans","methrix","methylSig","minfi","netSmooth","PCAtools","RTCGAToolbox","scater","scDblFinder","scMerge","scmeth","scran","scuttle","signatureSearch","SingleR","VariantExperiment","weitrix","BiocGenerics","ChIPpeakAnno","gwascat","iSEE","MAST","S4Vectors","SQLDataFrame"],["Depends","Depends","Depends","Depends","Depends","Depends","Depends","Depends","Depends","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Imports","Suggests","Suggests","Suggests","Suggests","Suggests","Suggests","Suggests"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Package<\/th>\n      <th>Dependency Type<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script><p>We will briefly highlight some of the key packages in this table, broadly categorising these as ‘user-focused’/‘user-facing’ or ‘developer-focused’ packages and those that span the spectrum.</p>
<div id="packages-that-both-users-and-developers-should-probably-know-about" class="section level4">
<h4 class="hasAnchor">
<a href="#packages-that-both-users-and-developers-should-probably-know-about" class="anchor"></a>Packages that both users and developers should probably know about</h4>
<div id="delayedarray" class="section level5">
<h5 class="hasAnchor">
<a href="#delayedarray" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em>
</h5>
<p>Well, duh. Implements the <em>DelayedArray</em> and <em>RleArray</em> classes, along with all the fundamentals the enable the delayed operations, block processing, and realization that underpin the DelayedArray framework.</p>
</div>
<div id="hdf5array" class="section level5">
<h5 class="hasAnchor">
<a href="#hdf5array" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em>
</h5>
<p>Implements the <em>HDF5Array</em> and <em>TENxMatrix</em> classes, two convenient and memory-efficient array-like containers for on-disk representation of HDF5 datasets. <em>HDF5Array</em> is for datasets that use the conventional (i.e. dense) HDF5 representation. <em>TENxMatrix</em> is for datasets that use the HDF5-based sparse matrix representation from 10x Genomics.</p>
</div>
<div id="tiledbarray" class="section level5">
<h5 class="hasAnchor">
<a href="#tiledbarray" class="anchor"></a><em><a href="https://github.com/LTLA/TileDBArray">TileDBArray</a></em>
</h5>
<p>Implements a DelayedArray backend for <a href="https://tiledb.com/">TileDB</a> to read, write and store dense and sparse arrays. The resulting <em>TileDBArray</em> objects are directly compatible with any Bioconductor package that accepts <em>DelayedArray</em> objects, serving as a swap-in replacement for the predominant <em>HDF5Array</em> that is currently used throughout the Bioconductor ecosystem for representing large datasets.</p>
<p><strong>NB</strong>: <em><a href="https://github.com/LTLA/TileDBArray">TileDBArray</a></em> is not yet available from Bioconductor.</p>
</div>
<div id="delayedmatrixstats" class="section level5">
<h5 class="hasAnchor">
<a href="#delayedmatrixstats" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em>
</h5>
<p>A port of the <em><a href="https://CRAN.R-project.org/package=matrixStats">matrixStats</a></em> API for use with <em>DelayedMatrix</em> objects. High-performing functions operating on rows and columns of <em>DelayedMatrix</em> objects, e.g. <code>col</code> / <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colMedians.html">rowMedians()</a></code>, <code>col</code> / <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colRanks.html">rowRanks()</a></code>, and <code>col</code> / <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colMads.html">rowSds()</a></code>. Functions optimized per data type and for subsetted calculations such that both memory usage and processing time is minimized.</p>
<p>Disclaimer: I wrote this one.</p>
</div>
<div id="biocsingular" class="section level5">
<h5 class="hasAnchor">
<a href="#biocsingular" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/BiocSingular">BiocSingular</a></em>
</h5>
<p>Implements exact and approximate methods for singular value decomposition and principal components analysis using a framework that allows them to be easily switched within Bioconductor packages or workflows. These methods work on <em>DelayedMatrix</em> objects as well as ordinary <em>matrix</em> objects and some sparse matrix objects from the <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> package.</p>
<p><em><a href="https://bioconductor.org/packages/3.12/BiocSingular">BiocSingular</a></em> defines a few interesting specialized <em>DelayedMatrix</em> subclasses that aim to preserve sparsity of the original matrix:</p>
<ul>
<li>
<em>DeferredMatrix</em>: Supports deferred centering and scaling of the columns of a matrix prior to principal components analysis.</li>
<li>
<em>LowRankMatrix</em>: Provides a memory-efficient representation of a low-rank reconstruction, e.g., after a principal components analysis.</li>
<li>
<em>ResidualMatrix</em>: Supports delayed calculation of the residuals from a linear model fit, usually prior to principal components analysis.</li>
</ul>
</div>
<div id="vcfarray-and-gdsarray" class="section level5">
<h5 class="hasAnchor">
<a href="#vcfarray-and-gdsarray" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/VCFArray">VCFArray</a></em> and <em><a href="https://bioconductor.org/packages/3.12/GDSArray">GDSArray</a></em>
</h5>
<p>Implements the <em>VCFArray</em> and <em>GDSArray</em> classes, types of <em>DelayedArray</em>, to represent VCF files and GDS-files in an array-like representation. VCF and GDS files are widely used to represent genotyping or sequence data.</p>
</div>
<div id="rhdf5client-and-restfulse" class="section level5">
<h5 class="hasAnchor">
<a href="#rhdf5client-and-restfulse" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/rhdf5client">rhdf5client</a></em> and <em><a href="https://bioconductor.org/packages/3.12/restfulSE">restfulSE</a></em>
</h5>
<p>Provide functions and classes to interface with remote data stores by operating on <em>SummarizedExperiment</em>-like objects. These data are HDF5 files living on a remote server running <code>h5serv</code>, a REST-based service for HDF5 data.</p>
</div>
</div>
<div id="user-focuseduser-facing-packages" class="section level4">
<h4 class="hasAnchor">
<a href="#user-focuseduser-facing-packages" class="anchor"></a>User-focused/user-facing packages</h4>
<p>These are the packages that as a user you might directly load/attach with <code>library()</code> as part of a data analysis. Alternatively, these may be loaded/attached as a dependency<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a> of another package you load/attach as part of an analysis.</p>
<div id="dropletutils" class="section level5">
<h5 class="hasAnchor">
<a href="#dropletutils" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/DropletUtils">DropletUtils</a></em>
</h5>
<p>Provides a number of utility functions for handling single-cell (RNA-seq) data from droplet technologies such as 10X Genomics. This includes <code>read10xCounts()</code> for data loading from the count matrices produced by 10x Genomics’ <strong>CellRanger</strong> software, which may be stored in an HDF5 file. To do this, it makes use of the <em>TENxMatrix</em> class.</p>
</div>
<div id="loomexperiment" class="section level5">
<h5 class="hasAnchor">
<a href="#loomexperiment" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/LoomExperiment">LoomExperiment</a></em>
</h5>
<p>Provides a means to convert from ‘loom’ files to standard Bioconductor classes and back again. The <a href="http://linnarssonlab.org/loompy/index.html">Loom file format</a> uses HDF5 to store experimental data and is used by some tools and labs producing data using single-cell assays. This includes the <code>import()</code> function for data loading from loom files into an <em>HDF5Matrix</em>.</p>
</div>
<div id="scuttle" class="section level5">
<h5 class="hasAnchor">
<a href="#scuttle" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/scuttle">scuttle</a></em>
</h5>
<p>Provides basic utility functions for performing single-cell analyses, focusing on simple normalization, quality control and data transformations. Also provides some helper functions to assist development of other packages. These methods work on <em>DelayedMatrix</em> objects as well as ordinary <em>matrix</em> objects and some sparse matrix objects from the <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> package.</p>
</div>
<div id="batchelor" class="section level5">
<h5 class="hasAnchor">
<a href="#batchelor" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/batchelor">batchelor</a></em>
</h5>
<p>Implements a variety of methods for batch correction of single-cell (RNA sequencing) data, such as <code>multiBatchPCA()</code> and <code>fastMNN()</code>. These methods work on <em>DelayedMatrix</em> objects as well as ordinary <em>matrix</em> objects and some sparse matrix objects from the <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> package.</p>
</div>
<div id="bsseq" class="section level5">
<h5 class="hasAnchor">
<a href="#bsseq" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/bsseq">bsseq</a></em>
</h5>
<p>A collection of tools for analyzing and visualizing bisulfite sequencing data. This was one of the first packages to make use of the DelayedArray framework and it supports these throughout the package. This was needed in order to store and analyse large non-CpG methylation datasets (&gt; 300 million loci, hundreds of samples) using HDF5 files.</p>
<p>Disclaimer: I did this re-write of <em><a href="https://bioconductor.org/packages/3.12/bsseq">bsseq</a></em> and learnt a lot along the way.</p>
</div>
<div id="minfi" class="section level5">
<h5 class="hasAnchor">
<a href="#minfi" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/minfi">minfi</a></em>
</h5>
<p>Tools to analyze &amp; visualize Illumina Infinium methylation arrays. This doesn’t have the same level of support for <em>DelayedMatrix</em> objects as <em><a href="https://bioconductor.org/packages/3.12/bsseq">bsseq</a></em>, but perhaps one day. This is needed in order to store and analyse large methylation datasets (&gt; 850,000 loci, tens of thousands of) using HDF5 files.</p>
<p>Disclaimer: This was the second package, after <em><a href="https://bioconductor.org/packages/3.12/bsseq">bsseq</a></em>, I started to re-write to support the DelayedArray framework. Here, it is rather more difficult because it is a ‘widely’ used package and has code from lots of different authors with different styles.</p>
</div>
</div>
<div id="developer-focused-packages" class="section level4">
<h4 class="hasAnchor">
<a href="#developer-focused-packages" class="anchor"></a>Developer-focused packages</h4>
<div id="beachmat" class="section level5">
<h5 class="hasAnchor">
<a href="#beachmat" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/beachmat">beachmat</a></em>
</h5>
<p>Provides a consistent C++ class interface for reading from and writing data to a variety of commonly used matrix types. Ordinary matrices and several sparse/dense <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> classes are directly supported, third-party S4 classes may be supported by external linkage (such as the <em>HDF5Matrix</em> class), while all other matrices are handled by DelayedArray block processing.</p>
</div>
</div>
</div>
</div>
<div id="first-contact" class="section level2">
<h2 class="hasAnchor">
<a href="#first-contact" class="anchor"></a>First contact</h2>
<div id="motivation" class="section level3">
<h3 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h3>
<p>The heart of the DelayedArray framework is implemented in the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package, which we now load and attach.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DelayedArray</span>)
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; Loading required package: Matrix</span>
<span class="co">#&gt; Loading required package: matrixStats</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class="co">#&gt;     union, unique, unsplit, which.max, which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following object is masked from 'package:Matrix':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'DelayedArray'</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     aperm, apply, rowsum</span></pre></body></html></div>
<p>We’ll also load and attach the <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> package, which extends the DelayedArray framework to support on-disk HDF5 files.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">HDF5Array</span>)
<span class="co">#&gt; Loading required package: rhdf5</span></pre></body></html></div>
<p>We will begin with an example using some scRNA-seq data on 1.3 million brain cells from embryonic mice, generated by 10X Genomics. This dataset is available from <em><a href="https://bioconductor.org/packages/3.12/ExperimentHub">ExperimentHub</a></em><a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ExperimentHub</span>)
<span class="co">#&gt; Loading required package: AnnotationHub</span>
<span class="co">#&gt; Loading required package: BiocFileCache</span>
<span class="co">#&gt; Loading required package: dbplyr</span>
<span class="no">hub</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html">ExperimentHub</a></span>()
<span class="co">#&gt; using temporary cache /tmp/RtmpQlTtWh/BiocFileCache</span>
<span class="co">#&gt; snapshotDate(): 2020-07-10</span>

<span class="co"># Query ExperimentHub to find the relevant resource.</span>
<span class="co"># This dataset is available in two formats: a 'dense matrix' format and a</span>
<span class="co"># 'HDF5-based 10X Genomics' format. We'll use the 'dense matrix' version for </span>
<span class="co"># this workshop.</span>
<span class="fu">query</span>(<span class="no">hub</span>, <span class="st">"TENxBrainData"</span>)
<span class="co">#&gt; ExperimentHub with 8 records</span>
<span class="co">#&gt; # snapshotDate(): 2020-07-10</span>
<span class="co">#&gt; # $dataprovider: 10X Genomics</span>
<span class="co">#&gt; # $species: Mus musculus</span>
<span class="co">#&gt; # $rdataclass: character</span>
<span class="co">#&gt; # additional mcols(): taxonomyid, genome, description,</span>
<span class="co">#&gt; #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span>
<span class="co">#&gt; #   rdatapath, sourceurl, sourcetype </span>
<span class="co">#&gt; # retrieve records with, e.g., 'object[["EH1039"]]' </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;            title                                                            </span>
<span class="co">#&gt;   EH1039 | Brain scRNA-seq data, 'HDF5-based 10X Genomics' format           </span>
<span class="co">#&gt;   EH1040 | Brain scRNA-seq data, 'dense matrix' format                      </span>
<span class="co">#&gt;   EH1041 | Brain scRNA-seq data, sample (column) annotation                 </span>
<span class="co">#&gt;   EH1042 | Brain scRNA-seq data, gene (row) annotation                      </span>
<span class="co">#&gt;   EH1689 | Brain scRNA-seq data 20k subset, 'HDF5-based 10x Genomics' format</span>
<span class="co">#&gt;   EH1690 | Brain scRNA-seq data 20k subset, 'dense matrix' format           </span>
<span class="co">#&gt;   EH1691 | Brain scRNA-seq data 20k subset, sample (column) annotation      </span>
<span class="co">#&gt;   EH1692 | Brain scRNA-seq data 20k subset, gene (row) annotation</span>
<span class="co"># Load the relevant resource.</span>
<span class="co"># This will download the data and may take a little while on the first run. </span>
<span class="co"># The result will be cached, however, so subsequent runs avoid re-downloading </span>
<span class="co"># the data.</span>
<span class="no">fname</span> <span class="kw">&lt;-</span> <span class="no">hub</span><span class="kw">[[</span><span class="st">"EH1040"</span>]]
<span class="co">#&gt; Bioconductor version 3.12 (BiocManager 1.30.10), R 4.0.0 (2020-04-24)</span>
<span class="co">#&gt; Installing package(s) 'TENxBrainData'</span>
<span class="co">#&gt; Old packages: 'boot', 'class', 'foreign', 'KernSmooth', 'MASS', 'nlme', 'nnet',</span>
<span class="co">#&gt;   'spatial'</span>
<span class="co">#&gt; see ?TENxBrainData and browseVignettes('TENxBrainData') for documentation</span>
<span class="co">#&gt; downloading 1 resources</span>
<span class="co">#&gt; retrieving 1 resource</span>
<span class="co">#&gt; loading from cache</span>

<span class="co"># The structure of this HDF5 file can be seen using the h5ls() command</span>
<span class="co"># from the rhdf5 package:</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rhdf5</span>)
<span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5ls.html">h5ls</a></span>(<span class="no">fname</span>)
<span class="co">#&gt;   group   name       otype  dclass             dim</span>
<span class="co">#&gt; 0     / counts H5I_DATASET INTEGER 27998 x 1306127</span>

<span class="co"># The 1.3 Million Brain Cell Dataset is represented by the "counts" group. </span>
<span class="co"># We point the HDF5Array() constructor to this group to create a HDF5Matrix </span>
<span class="co"># object representing the dataset:</span>
<span class="no">tenx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/HDF5Array-class.html">HDF5Array</a></span>(<span class="kw">filepath</span> <span class="kw">=</span> <span class="no">fname</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">"counts"</span>)

<span class="co"># The data contain counts on nearly 28,000 gene for more than </span>
<span class="co"># 1.3 million cells.</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">tenx</span>)
<span class="co">#&gt; [1]   27998 1306127</span></pre></body></html></div>
<p>With data from 1.3 million cells, this is roughly 100,000-times more samples than a typical bulk RNA-seq dataset and would require over 140 GB of RAM to hold as a matrix and around 30 GB as a sparse matrix.</p>
<p>With so much data, we might expect that it would feel sluggish to interact with this object, but this is not the case. For example, let’s do something that would ordinarily be a terrible idea, and something that’s frustrated me way too many times: let’s ‘accidentally’ print out the entire counts matrix.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">tenx</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span></pre></body></html></div>
<p>Hallelujah! Unlike what you may have experienced when printing out a large matrix, this didn’t overwhelm the screen with thousands of lines of output nor did it cause the R session to hang indefinitely. In fact, this gives us a rather pretty printing of the counts matrix<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a>. No need for panicked mashing of <code>Ctrl-c</code> or <code>Esc</code>.</p>
<div id="a-peak-behind-the-curtain" class="section level4">
<h4 class="hasAnchor">
<a href="#a-peak-behind-the-curtain" class="anchor"></a>A peak behind the curtain</h4>
<p>We might now recognise that <code>tenx</code> is no ordinary <em>matrix</em>. In fact, it is an <em>HDF5Matrix</em>, which is a type of <em>DelayedArray</em><a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a>.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx</span>)
<span class="co">#&gt; [1] "HDF5Matrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span>
<span class="fu">is</span>(<span class="no">tenx</span>, <span class="st">"DelayedArray"</span>)
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>The data contained in an <em>HDF5Matrix</em> is actually stored on disk in a <a href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">Hierarchical Data Format (<strong>HDF5</strong>)</a> file. Consequently, the <code>tenx</code> object takes up relatively little space in memory<a href="#fn8" class="footnote-ref" id="fnref8"><sup>8</sup></a>.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span>(<span class="no">tenx</span>), <span class="kw">units</span> <span class="kw">=</span> <span class="st">"auto"</span>)
<span class="co">#&gt; 2.3 Kb</span></pre></body></html></div>
<p>We can learn more about the internals of the <code>tenx</code> object using the <code><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed()</a></code> function.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span>(<span class="no">tenx</span>)
<span class="co">#&gt; An object of class "HDF5ArraySeed"</span>
<span class="co">#&gt; Slot "filepath":</span>
<span class="co">#&gt; [1] "/tmp/RtmpQlTtWh/BiocFileCache/43131dcac8_1040"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "name":</span>
<span class="co">#&gt; [1] "counts"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "type":</span>
<span class="co">#&gt; [1] NA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "dim":</span>
<span class="co">#&gt; [1]   27998 1306127</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "chunkdim":</span>
<span class="co">#&gt; [1] 100 100</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "first_val":</span>
<span class="co">#&gt; [1] 0</span></pre></body></html></div>
</div>
</div>
<div id="three-examples-of-computing-on-a-delayedarray" class="section level3">
<h3 class="hasAnchor">
<a href="#three-examples-of-computing-on-a-delayedarray" class="anchor"></a>Three examples of computing on a DelayedArray</h3>
<p>We will now play around with computing on the counts matrix. To make things slightly easier, we will first subset the data to 1000 samples.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">tenx_subset</span> <span class="kw">&lt;-</span> <span class="no">tenx</span>[, <span class="fl">1</span>:<span class="fl">1000</span>]</pre></body></html></div>
<div id="library-sizes" class="section level4">
<h4 class="hasAnchor">
<a href="#library-sizes" class="anchor"></a>Library sizes</h4>
<p>Firstly, let’s compute the library sizes for this subset of samples. We can do this using <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code>.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">lib_sizes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">tenx_subset</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">lib_sizes</span>)
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt;    1453    2755    3898    4689    5548   34233</span></pre></body></html></div>
</div>
<div id="proportion-of-cells-with-non-zero-expression-for-each-gene" class="section level4">
<h4 class="hasAnchor">
<a href="#proportion-of-cells-with-non-zero-expression-for-each-gene" class="anchor"></a>Proportion of cells with non-zero expression for each gene</h4>
<p>Secondly, suppose we want to know for each gene the proportion of cells with non-zero expression. We can do this using <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowSums()</a></code> in conjunction with some standard R commands (logical comparisons and division).</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">prop_non_zero</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowSums</a></span>(<span class="no">tenx_subset</span> <span class="kw">&gt;</span> <span class="fl">0</span>) /  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_subset</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">prop_non_zero</span>)
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt; 0.00000 0.00000 0.00300 0.07235 0.07400 1.00000</span></pre></body></html></div>
</div>
<div id="median-expression-of-each-gene" class="section level4">
<h4 class="hasAnchor">
<a href="#median-expression-of-each-gene" class="anchor"></a>Median expression of each gene</h4>
<p>Finally, suppose we want to know the median expression of each gene. Here, we will quantify expression as counts per million (CPM) using library size normalization.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">cpm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="fl">1e6</span> * <span class="no">tenx_subset</span>) / <span class="no">lib_sizes</span>)
<span class="no">cpm</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "double":</span>
<span class="co">#&gt;              [,1]     [,2]     [,3] ...   [,999]  [,1000]</span>
<span class="co">#&gt;     [1,]        0        0        0   .        0        0</span>
<span class="co">#&gt;     [2,]        0        0        0   .        0        0</span>
<span class="co">#&gt;     [3,]        0        0        0   .        0        0</span>
<span class="co">#&gt;     [4,]        0        0        0   .        0        0</span>
<span class="co">#&gt;     [5,]        0        0        0   .        0        0</span>
<span class="co">#&gt;      ...        .        .        .   .        .        .</span>
<span class="co">#&gt; [27994,]   0.0000   0.0000   0.0000   .   0.0000   0.0000</span>
<span class="co">#&gt; [27995,] 247.1577   0.0000   0.0000   . 277.1619   0.0000</span>
<span class="co">#&gt; [27996,]   0.0000   0.0000   0.0000   .   0.0000   0.0000</span>
<span class="co">#&gt; [27997,]   0.0000   0.0000   0.0000   .   0.0000   0.0000</span>
<span class="co">#&gt; [27998,]   0.0000   0.0000   0.0000   .   0.0000   0.0000</span></pre></body></html></div>
<p>We can then compute the median expression of each gene using <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colMedians.html">DelayedMatrixStats::rowMedians()</a></code>.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DelayedMatrixStats</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'DelayedMatrixStats'</span>
<span class="co">#&gt; The following object is masked from 'package:Biobase':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     rowMedians</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colAlls, colAnyMissings, colAnyNAs, colAnys, colAvgsPerRowSet,</span>
<span class="co">#&gt;     colCollapse, colCounts, colCummaxs, colCummins, colCumprods,</span>
<span class="co">#&gt;     colCumsums, colDiffs, colIQRDiffs, colIQRs, colLogSumExps,</span>
<span class="co">#&gt;     colMadDiffs, colMads, colMeans2, colMedians, colOrderStats,</span>
<span class="co">#&gt;     colProds, colQuantiles, colRanks, colSdDiffs, colSds, colSums2,</span>
<span class="co">#&gt;     colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyMissings, rowAnyNAs, rowAnys,</span>
<span class="co">#&gt;     rowAvgsPerColSet, rowCollapse, rowCounts, rowCummaxs, rowCummins,</span>
<span class="co">#&gt;     rowCumprods, rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs,</span>
<span class="co">#&gt;     rowLogSumExps, rowMadDiffs, rowMads, rowMeans2, rowMedians,</span>
<span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanks, rowSdDiffs,</span>
<span class="co">#&gt;     rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span>
<span class="no">median_expression</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colMedians.html">rowMedians</a></span>(<span class="no">cpm</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">median_expression</span>)
<span class="co">#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span>
<span class="co">#&gt;     0.00     0.00     0.00    18.73     0.00 22069.23</span></pre></body></html></div>
</div>
<div id="summary" class="section level4">
<h4 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h4>
<p>These 3 examples highlight the power of the DelayedArray framework. Recall that the data in these examples live on disk in an HDF5 file, yet we interacted with <code>tenx_subset</code> and computed on it much as we would if the data were in-memory as an ordinary matrix. Also note that all 3 examples returned ordinary R vectors.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">lib_sizes</span>)
<span class="co">#&gt; [1] "numeric"</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">prop_non_zero</span>)
<span class="co">#&gt; [1] "numeric"</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">median_expression</span>)
<span class="co">#&gt; [1] "numeric"</span></pre></body></html></div>
<p>To do so, we made (implicit) use of the three fundamental concepts of the DelayedArray framework:</p>
<ol style="list-style-type: decimal">
<li>Delayed operations</li>
<li>Block processing</li>
<li>Realization</li>
</ol>
<p>We’ll now discuss each of these in turn.</p>
</div>
</div>
<div id="delayed-operations" class="section level3">
<h3 class="hasAnchor">
<a href="#delayed-operations" class="anchor"></a>Delayed operations</h3>
<p>Taking a careful look at <code>tenx_subset</code>, we see that it is a <em>DelayedMatrix</em> rather than an <em>HDF5Matrix</em>.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">tenx_subset</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span></pre></body></html></div>
<p>The subsetting operation has ‘degraded’ the <code>tenx_subset</code> object to a <em>DelayedMatrix</em>.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu">is</span>(<span class="no">tenx_subset</span>, <span class="st">"HDF5Matrix"</span>)
<span class="co">#&gt; [1] FALSE</span>
<span class="fu">is</span>(<span class="no">tenx_subset</span>, <span class="st">"DelayedMatrix"</span>)
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>The <code><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree()</a></code> function can help us see what changed when we subsetted the data.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx</span>)
<span class="co">#&gt; 27998x1306127 integer: HDF5Matrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx_subset</span>)
<span class="co">#&gt; 27998x1000 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
<p>The subsetting operation has been registered in what is termed a ‘delayed operation’. Registering a delayed operation does not modify the underlying data. Instead, the operation is recorded and only performed when the <em>DelayedArray</em> object is ‘realized’. Realization of a <em>DelayedArray</em> triggers the execution of the delayed operations carried by the object and returns the result as an ordinary <em>array</em>.</p>
<p>This allows us to chain together multiple operations and only perform them as required. Here is a contrived example.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="co"># Add 1 to every element (a delayed op).</span>
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="no">tenx_subset</span> + <span class="fl">1L</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">x</span>)
<span class="co">#&gt; 27998x1000 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;       └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>

<span class="co"># Compute log of every element (another delayed op).</span>
<span class="no">lx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="no">x</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">lx</span>)
<span class="co">#&gt; 27998x1000 double: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 double: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;       └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>

<span class="co"># Transpose the result (another delayed op).</span>
<span class="no">tlx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="no">lx</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tlx</span>)
<span class="co">#&gt; 1000x27998 double: DelayedMatrix object</span>
<span class="co">#&gt; └─ 1000x27998 double: Unary iso op stack</span>
<span class="co">#&gt;    └─ 1000x27998 integer: Aperm (perm=c(2,1))</span>
<span class="co">#&gt;       └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;          └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>

<span class="co"># Realize a subset of the data as an ordinary matrix.</span>
<span class="fu"><a href="https://rdrr.io/r/base/array.html">as.array</a></span>(<span class="no">tlx</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">10</span>])
<span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7]     [,8]      [,9] [,10]</span>
<span class="co">#&gt; [1,]    0    0    0    0    0    0    0 0.000000 0.0000000     0</span>
<span class="co">#&gt; [2,]    0    0    0    0    0    0    0 0.000000 0.0000000     0</span>
<span class="co">#&gt; [3,]    0    0    0    0    0    0    0 0.000000 0.6931472     0</span>
<span class="co">#&gt; [4,]    0    0    0    0    0    0    0 0.000000 0.0000000     0</span>
<span class="co">#&gt; [5,]    0    0    0    0    0    0    0 1.098612 0.6931472     0</span></pre></body></html></div>
<p>Many common operations can be registered as delayed operations. Here are some examples<a href="#fn9" class="footnote-ref" id="fnref9"><sup>9</sup></a>. Notice that in each case the result is ‘degraded’ to a <em>DelayedMatrix</em><a href="#fn10" class="footnote-ref" id="fnref10"><sup>10</sup></a>.</p>
<div id="delayedsubset" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedsubset" class="anchor"></a>DelayedSubset</h4>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx</span>[, <span class="fl">1</span>:<span class="fl">100</span>]
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 100&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;            [,1]   [,2]   [,3]   [,4] ...  [,97]  [,98]  [,99] [,100]</span>
<span class="co">#&gt;     [1,]      0      0      0      0   .      0      0      1      0</span>
<span class="co">#&gt;     [2,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt;     [3,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt;     [4,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt;     [5,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt;      ...      .      .      .      .   .      .      .      .      .</span>
<span class="co">#&gt; [27994,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt; [27995,]      1      0      0      2   .      0      1      2      0</span>
<span class="co">#&gt; [27996,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt; [27997,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt; [27998,]      0      0      0      0   .      0      0      0      0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x100 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x100 integer: Subset</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayedaperm" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedaperm" class="anchor"></a>DelayedAperm</h4>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="no">tenx</span>)
<span class="no">val</span>
<span class="co">#&gt; &lt;1306127 x 27998&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]     [,2]     [,3]     [,4] ... [,27995] [,27996] [,27997]</span>
<span class="co">#&gt;       [1,]        0        0        0        0   .        1        0        0</span>
<span class="co">#&gt;       [2,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt;       [3,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt;       [4,]        0        0        0        0   .        2        0        0</span>
<span class="co">#&gt;       [5,]        0        0        0        0   .        2        0        0</span>
<span class="co">#&gt;        ...        .        .        .        .   .        .        .        .</span>
<span class="co">#&gt; [1306123,]        0        0        0        0   .        0        2        0</span>
<span class="co">#&gt; [1306124,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt; [1306125,]        0        0        0        0   .        1        0        0</span>
<span class="co">#&gt; [1306126,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt; [1306127,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt;            [,27998]</span>
<span class="co">#&gt;       [1,]        0</span>
<span class="co">#&gt;       [2,]        0</span>
<span class="co">#&gt;       [3,]        0</span>
<span class="co">#&gt;       [4,]        0</span>
<span class="co">#&gt;       [5,]        0</span>
<span class="co">#&gt;        ...        .</span>
<span class="co">#&gt; [1306123,]        0</span>
<span class="co">#&gt; [1306124,]        0</span>
<span class="co">#&gt; [1306125,]        0</span>
<span class="co">#&gt; [1306126,]        0</span>
<span class="co">#&gt; [1306127,]        0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 1306127x27998 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 1306127x27998 integer: Aperm (perm=c(2,1))</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayedunaryisoop" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedunaryisoop" class="anchor"></a>DelayedUnaryIsoOp</h4>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx</span> + <span class="fl">1L</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [2,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [3,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [4,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [5,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt; [27995,]          2          1          1          3   .          1          2</span>
<span class="co">#&gt; [27996,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt; [27997,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt; [27998,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          1          1</span>
<span class="co">#&gt;     [2,]          1          1</span>
<span class="co">#&gt;     [3,]          1          1</span>
<span class="co">#&gt;     [4,]          1          1</span>
<span class="co">#&gt;     [5,]          1          1</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          1          1</span>
<span class="co">#&gt; [27995,]          1          1</span>
<span class="co">#&gt; [27996,]          1          1</span>
<span class="co">#&gt; [27997,]          1          1</span>
<span class="co">#&gt; [27998,]          1          1</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx</span> + <span class="fl">1</span>:<span class="fl">2</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [2,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt;     [3,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [4,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt;     [5,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt; [27995,]          2          1          1          3   .          1          2</span>
<span class="co">#&gt; [27996,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt; [27997,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt; [27998,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          1          1</span>
<span class="co">#&gt;     [2,]          2          2</span>
<span class="co">#&gt;     [3,]          1          1</span>
<span class="co">#&gt;     [4,]          2          2</span>
<span class="co">#&gt;     [5,]          1          1</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          2          2</span>
<span class="co">#&gt; [27995,]          1          1</span>
<span class="co">#&gt; [27996,]          2          2</span>
<span class="co">#&gt; [27997,]          1          1</span>
<span class="co">#&gt; [27998,]          2          2</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Unary iso op with args</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayedsubassign" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedsubassign" class="anchor"></a>DelayedSubassign</h4>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="no">tenx</span>
<span class="no">tmp</span>[<span class="fl">1</span>, ] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tmp</span>), <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">tmp</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          3         10          1          5   .          2          4</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          6          5</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tmp</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Subassign</span>
<span class="co">#&gt;    ├─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="co">#&gt;    └─ right value: 1x1306127 integer: [seed] matrix object</span></pre></body></html></div>
<p><strong>WARNING</strong>: Be careful with delayed subassignment because you can end up with objects that are surprisingly large in-memory. This is because the subassigned values are kept in-memory until the data are <em>realized</em>.</p>
</div>
<div id="delayeddimnames" class="section level4">
<h4 class="hasAnchor">
<a href="#delayeddimnames" class="anchor"></a>DelayedDimnames</h4>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="no">tenx</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">tmp</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"R"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">tmp</span>)))
<span class="no">tmp</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;              [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     R1          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     R2          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     R3          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     R4          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     R5          0          0          0          0   .          0          0</span>
<span class="co">#&gt;    ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; R27994          0          0          0          0   .          0          0</span>
<span class="co">#&gt; R27995          1          0          0          2   .          0          1</span>
<span class="co">#&gt; R27996          0          0          0          0   .          0          0</span>
<span class="co">#&gt; R27997          0          0          0          0   .          0          0</span>
<span class="co">#&gt; R27998          0          0          0          0   .          0          0</span>
<span class="co">#&gt;        [,1306126] [,1306127]</span>
<span class="co">#&gt;     R1          0          0</span>
<span class="co">#&gt;     R2          0          0</span>
<span class="co">#&gt;     R3          0          0</span>
<span class="co">#&gt;     R4          0          0</span>
<span class="co">#&gt;     R5          0          0</span>
<span class="co">#&gt;    ...          .          .</span>
<span class="co">#&gt; R27994          0          0</span>
<span class="co">#&gt; R27995          0          0</span>
<span class="co">#&gt; R27996          0          0</span>
<span class="co">#&gt; R27997          0          0</span>
<span class="co">#&gt; R27998          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tmp</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Set dimnames</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayednaryisoop" class="section level4">
<h4 class="hasAnchor">
<a href="#delayednaryisoop" class="anchor"></a>DelayedNaryIsoOp</h4>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx</span> + <span class="no">tenx</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          2          0          0          4   .          0          2</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: N-ary iso op</span>
<span class="co">#&gt;    ├─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayedabind" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedabind" class="anchor"></a>DelayedAbind</h4>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html">cbind</a></span>(<span class="no">tenx</span>, <span class="no">tenx</span>)
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 2612254&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,2612251] [,2612252]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,2612253] [,2612254]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x2612254 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x2612254 integer: Abind (along=2)</span>
<span class="co">#&gt;    ├─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="no-op" class="section level4">
<h4 class="hasAnchor">
<a href="#no-op" class="anchor"></a>No-op</h4>
<p>The DelayedArray framework is smart enough to recognise that some combinations of operations are ‘no-ops’.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="no">tenx</span>))
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: HDF5Matrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
<p>But it can be fooled.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="co"># This is a no-op but DelayedArray doesn't recognise it as one.</span>
<span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx</span> + <span class="fl">0L</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
</div>
<div id="block-processing" class="section level3">
<h3 class="hasAnchor">
<a href="#block-processing" class="anchor"></a>Block processing</h3>
<p>In <a href="#library-sizes">Library sizes</a>, we computed the column sums of <code>tenx_subset</code> by running <code>colSums(tenx_subset</code>). If you have used the <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code> function in your own R work, then this code likely looks familiar to you. However, recall that the <code>tenx_subset</code> data live on disk in an HDF5 file, so how did <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code> know how to handle this?</p>
<p>The ‘trick’ is that we are using a specialised version of <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code><a href="#fn11" class="footnote-ref" id="fnref11"><sup>11</sup></a> which uses a technique called ‘block processing’ to compute the column sums.</p>
<div id="illustration-of-block-processing" class="section level4">
<h4 class="hasAnchor">
<a href="#illustration-of-block-processing" class="anchor"></a>Illustration of block processing</h4>
<p>Block processing involves 2 steps:</p>
<ol style="list-style-type: decimal">
<li>Load a ‘block’ of the data into memory and compute a statistic(s) on the block.</li>
<li>Combine the block-level statistics in an appropriate way to get the final result.</li>
</ol>
<p>Some examples of block processing are illustrated in the following figures:</p>
<p><img src="images/Slide1.png"><!-- --><img src="images/Slide2.png"><!-- --><img src="images/Slide3.png"><!-- --><img src="images/Slide4.png"><!-- --><img src="images/Slide5.png"><!-- --><img src="images/Slide6.png"><!-- --><img src="images/Slide7.png"><!-- --></p>
<p>For example, to compute the column sums we could define a block to be a column, loop over the blocks (columns), load each block (column) into memory, and compute it’s sum. You may already be thinking:</p>
<ul>
<li>“I have a very ‘tall’ matrix and I can’t load even a single column into memory. Can block processing support this?”</li>
<li>“I have enough RAM to load 100 columns of my matrix into memory. Can block processing support this?”</li>
</ul>
<p>The answer to both these questions is generally “yes”, which we will return to in the <a href="#block-size">Block size</a> section.</p>
</div>
</div>
<div id="more-examples-of-operations-that-use-block-processing" class="section level3">
<h3 class="hasAnchor">
<a href="#more-examples-of-operations-that-use-block-processing" class="anchor"></a>More examples of operations that use block processing</h3>
<p>To more clearly see block processing in action, we’ll turn on verbose progress reporting from the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package.</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="kw pkg">DelayedArray</span><span class="kw ns">:::</span><span class="fu">set_verbose_block_processing</span>(<span class="fl">TRUE</span>)
<span class="co">#&gt; [1] FALSE</span></pre></body></html></div>
<p>Now, let’s re-run <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums(tenx_subset)</a></code>.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="co"># invisible() is used to prevent the result from printing to the screen.</span>
<span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">tenx_subset</span>))
<span class="co">#&gt; Processing block [[1/2, 1/1]] ... OK</span>
<span class="co">#&gt; Processing block [[2/2, 1/1]] ... OK</span></pre></body></html></div>
<p>In this case, <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code> has processed <code>tenx_subset</code> in 2 blocks. The verbose progress report tells us that these blocks are over the <em>rows</em> of <code>tenx_subset</code>.</p>
<p>Let’s take a look at a few more examples of functions implemented with block processing.</p>
<div id="functions-in-delayedarray" class="section level4">
<h4 class="hasAnchor">
<a href="#functions-in-delayedarray" class="anchor"></a>Functions in DelayedArray</h4>
<p>Some of the most useful functions in the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package implement common operations on a <em>DelayedMatrix</em> using block processing. These include the following row and column summarization methods:</p>
<ul>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowSums()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowMeans()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMeans()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowMaxs()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMaxs()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowMins()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMins()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowRanges()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colRanges()</a></code></li>
</ul>
<p>Two useful but lesser known functions use block processing to compute column/row sums of a <em>DelayedMatrix</em> based on a grouping variable:</p>
<ul>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">rowsum()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">colsum()</a></code></li>
</ul>
<p>Matrix multiplication is also implemented using block processing:</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="co"># This is mathematically equivalent to colSums(tenx_subset).</span>
<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">1</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">tenx_subset</span>)) <span class="kw">%*%</span> <span class="no">tenx_subset</span>
<span class="co">#&gt; &lt;1 x 1000&gt; matrix of class DelayedMatrix and type "double":</span>
<span class="co">#&gt;         [,1]    [,2]    [,3] ...  [,999] [,1000]</span>
<span class="co">#&gt; [1,]    4046    2087    4654   .    7216    5553</span></pre></body></html></div>
</div>
<div id="functions-in-delayedmatrixstats" class="section level4">
<h4 class="hasAnchor">
<a href="#functions-in-delayedmatrixstats" class="anchor"></a>Functions in DelayedMatrixStats</h4>
<p>We’ve already seen the <em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em> package in action back when computing the <a href="#median-expression-of-each-gene">Median expression of each gene</a>. <em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em> is a port of the <em><a href="https://CRAN.R-project.org/package=matrixStats">matrixStats</a></em> package’s API for use with <em>DelayedMatrix</em> objects. It provides <a href="https://github.com/PeteHaitch/DelayedMatrixStats#api-coverage">more than 70 functions</a> that apply to rows and columns of <em>DelayedMatrix</em> objects.</p>
</div>
<div id="try-it-yourself" class="section level4">
<h4 class="hasAnchor">
<a href="#try-it-yourself" class="anchor"></a>Try it yourself</h4>
<p>Try out some of the block processing functions from <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> and <em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em> on <code>tenx_subset</code><a href="#fn12" class="footnote-ref" id="fnref12"><sup>12</sup></a>.</p>
</div>
</div>
<div id="general-block-processing" class="section level3">
<h3 class="hasAnchor">
<a href="#general-block-processing" class="anchor"></a>General block processing</h3>
<p>As we have seen, many common row/column summarization methods on a <em>DelayedMatrix</em> have already been implemented in <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> and <em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em>. Nonetheless, there may be times you need to implement your own algorithm using block processing. The documentation on this topic is a little sparse, but some details can be found in <code>help("block_processing", "DelayedArray"),</code> <code><a href="https://rdrr.io/pkg/DelayedArray/man/ArrayGrid-class.html">help("ArrayGrid", "DelayedArray")</a></code>, <code><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">help("blockApply", "DelayedArray")</a></code>, and <code><a href="https://rdrr.io/pkg/DelayedArray/man/AutoGrid.html">help("AutoGrid", "DelayedArray")</a></code> or by reading the source code of the aforementioned packages. Briefly, to perform block processing requires that you:</p>
<ol style="list-style-type: decimal">
<li>Set up an <em>ArrayGrid</em> over the <em>DelayedArray</em> to be processed. This specifies the block structure that will be traversed when processing the <em>DelayedArray</em>.
<ul>
<li>The <code><a href="https://rdrr.io/pkg/DelayedArray/man/AutoGrid.html">defaultAutoGrid()</a></code>, <code><a href="https://rdrr.io/pkg/DelayedArray/man/AutoGrid.html">rowAutoGrid()</a></code>, and <code><a href="https://rdrr.io/pkg/DelayedArray/man/AutoGrid.html">colAutoGrid()</a></code> functions can help set up the <em>ArrayGrid</em>.</li>
</ul>
</li>
<li>Iterate over the <em>DelayedArray</em> via the <em>ArrayGrid</em> to read each block of data into memory as an ordinary (i.e. dense) or sparse array and compute the statistic for that block.
<ul>
<li>The <code><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">blockApply()</a></code> and <code><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">blockReduce()</a></code> functions can help perform the block processing, even incorporating parallelization via the <em><a href="https://bioconductor.org/packages/3.12/BiocParallel">BiocParallel</a></em> package.</li>
</ul>
</li>
<li>Appropriately combine the block-level statistics to get your final result.
<ul>
<li>This is typically up to you as a developer of the function.</li>
</ul>
</li>
</ol>
<div id="example" class="section level4">
<h4 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h4>
<p>Let’s implement a basic version of <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code> where we define each block to be a column</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">basic_colSums</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">x</span>) {
  <span class="co"># 1. Set up the ArrayGrid.</span>
  <span class="no">grid</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/AutoGrid.html">colAutoGrid</a></span>(<span class="no">x</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">1</span>)
  <span class="co"># 2. Load the blocks into memory and compute the block-level statistics.</span>
  <span class="no">block_level_stat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">blockApply</a></span>(<span class="no">x</span>, <span class="no">colSums</span>, <span class="kw">grid</span> <span class="kw">=</span> <span class="no">grid</span>)
  <span class="co"># 3. Combine the block-level statistics.</span>
  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">block_level_stat</span>)
}

<span class="co"># Check basic_colSums() gives the correct result.</span>
<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="fu">basic_colSums</span>(<span class="no">tenx_subset</span>), <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">tenx_subset</span>))
<span class="co">#&gt; Processing block 1/1000 ... OK</span>
<span class="co">#&gt; Processing block 2/1000 ... OK</span>
<span class="co">#&gt; Processing block 3/1000 ... OK</span>
<span class="co">#&gt; Processing block 4/1000 ... OK</span>
<span class="co">#&gt; Processing block 5/1000 ... OK</span>
<span class="co">#&gt; Processing block 6/1000 ... OK</span>
<span class="co">#&gt; Processing block 7/1000 ... OK</span>
<span class="co">#&gt; Processing block 8/1000 ... OK</span>
<span class="co">#&gt; Processing block 9/1000 ... OK</span>
<span class="co">#&gt; Processing block 10/1000 ... OK</span>
<span class="co">#&gt; Processing block 11/1000 ... OK</span>
<span class="co">#&gt; Processing block 12/1000 ... OK</span>
<span class="co">#&gt; Processing block 13/1000 ... OK</span>
<span class="co">#&gt; Processing block 14/1000 ... OK</span>
<span class="co">#&gt; Processing block 15/1000 ... OK</span>
<span class="co">#&gt; Processing block 16/1000 ... OK</span>
<span class="co">#&gt; Processing block 17/1000 ... OK</span>
<span class="co">#&gt; Processing block 18/1000 ... OK</span>
<span class="co">#&gt; Processing block 19/1000 ... OK</span>
<span class="co">#&gt; Processing block 20/1000 ... OK</span>
<span class="co">#&gt; Processing block 21/1000 ... OK</span>
<span class="co">#&gt; Processing block 22/1000 ... OK</span>
<span class="co">#&gt; Processing block 23/1000 ... OK</span>
<span class="co">#&gt; Processing block 24/1000 ... OK</span>
<span class="co">#&gt; Processing block 25/1000 ... OK</span>
<span class="co">#&gt; Processing block 26/1000 ... OK</span>
<span class="co">#&gt; Processing block 27/1000 ... OK</span>
<span class="co">#&gt; Processing block 28/1000 ... OK</span>
<span class="co">#&gt; Processing block 29/1000 ... OK</span>
<span class="co">#&gt; Processing block 30/1000 ... OK</span>
<span class="co">#&gt; Processing block 31/1000 ... OK</span>
<span class="co">#&gt; Processing block 32/1000 ... OK</span>
<span class="co">#&gt; Processing block 33/1000 ... OK</span>
<span class="co">#&gt; Processing block 34/1000 ... OK</span>
<span class="co">#&gt; Processing block 35/1000 ... OK</span>
<span class="co">#&gt; Processing block 36/1000 ... OK</span>
<span class="co">#&gt; Processing block 37/1000 ... OK</span>
<span class="co">#&gt; Processing block 38/1000 ... OK</span>
<span class="co">#&gt; Processing block 39/1000 ... OK</span>
<span class="co">#&gt; Processing block 40/1000 ... OK</span>
<span class="co">#&gt; Processing block 41/1000 ... OK</span>
<span class="co">#&gt; Processing block 42/1000 ... OK</span>
<span class="co">#&gt; Processing block 43/1000 ... OK</span>
<span class="co">#&gt; Processing block 44/1000 ... OK</span>
<span class="co">#&gt; Processing block 45/1000 ... OK</span>
<span class="co">#&gt; Processing block 46/1000 ... OK</span>
<span class="co">#&gt; Processing block 47/1000 ... OK</span>
<span class="co">#&gt; Processing block 48/1000 ... OK</span>
<span class="co">#&gt; Processing block 49/1000 ... OK</span>
<span class="co">#&gt; Processing block 50/1000 ... OK</span>
<span class="co">#&gt; Processing block 51/1000 ... OK</span>
<span class="co">#&gt; Processing block 52/1000 ... OK</span>
<span class="co">#&gt; Processing block 53/1000 ... OK</span>
<span class="co">#&gt; Processing block 54/1000 ... OK</span>
<span class="co">#&gt; Processing block 55/1000 ... OK</span>
<span class="co">#&gt; Processing block 56/1000 ... OK</span>
<span class="co">#&gt; Processing block 57/1000 ... OK</span>
<span class="co">#&gt; Processing block 58/1000 ... OK</span>
<span class="co">#&gt; Processing block 59/1000 ... OK</span>
<span class="co">#&gt; Processing block 60/1000 ... OK</span>
<span class="co">#&gt; Processing block 61/1000 ... OK</span>
<span class="co">#&gt; Processing block 62/1000 ... OK</span>
<span class="co">#&gt; Processing block 63/1000 ... OK</span>
<span class="co">#&gt; Processing block 64/1000 ... OK</span>
<span class="co">#&gt; Processing block 65/1000 ... OK</span>
<span class="co">#&gt; Processing block 66/1000 ... OK</span>
<span class="co">#&gt; Processing block 67/1000 ... OK</span>
<span class="co">#&gt; Processing block 68/1000 ... OK</span>
<span class="co">#&gt; Processing block 69/1000 ... OK</span>
<span class="co">#&gt; Processing block 70/1000 ... OK</span>
<span class="co">#&gt; Processing block 71/1000 ... OK</span>
<span class="co">#&gt; Processing block 72/1000 ... OK</span>
<span class="co">#&gt; Processing block 73/1000 ... OK</span>
<span class="co">#&gt; Processing block 74/1000 ... OK</span>
<span class="co">#&gt; Processing block 75/1000 ... OK</span>
<span class="co">#&gt; Processing block 76/1000 ... OK</span>
<span class="co">#&gt; Processing block 77/1000 ... OK</span>
<span class="co">#&gt; Processing block 78/1000 ... OK</span>
<span class="co">#&gt; Processing block 79/1000 ... OK</span>
<span class="co">#&gt; Processing block 80/1000 ... OK</span>
<span class="co">#&gt; Processing block 81/1000 ... OK</span>
<span class="co">#&gt; Processing block 82/1000 ... OK</span>
<span class="co">#&gt; Processing block 83/1000 ... OK</span>
<span class="co">#&gt; Processing block 84/1000 ... OK</span>
<span class="co">#&gt; Processing block 85/1000 ... OK</span>
<span class="co">#&gt; Processing block 86/1000 ... OK</span>
<span class="co">#&gt; Processing block 87/1000 ... OK</span>
<span class="co">#&gt; Processing block 88/1000 ... OK</span>
<span class="co">#&gt; Processing block 89/1000 ... OK</span>
<span class="co">#&gt; Processing block 90/1000 ... OK</span>
<span class="co">#&gt; Processing block 91/1000 ... OK</span>
<span class="co">#&gt; Processing block 92/1000 ... OK</span>
<span class="co">#&gt; Processing block 93/1000 ... OK</span>
<span class="co">#&gt; Processing block 94/1000 ... OK</span>
<span class="co">#&gt; Processing block 95/1000 ... OK</span>
<span class="co">#&gt; Processing block 96/1000 ... OK</span>
<span class="co">#&gt; Processing block 97/1000 ... OK</span>
<span class="co">#&gt; Processing block 98/1000 ... OK</span>
<span class="co">#&gt; Processing block 99/1000 ... OK</span>
<span class="co">#&gt; Processing block 100/1000 ... OK</span>
<span class="co">#&gt; Processing block 101/1000 ... OK</span>
<span class="co">#&gt; Processing block 102/1000 ... OK</span>
<span class="co">#&gt; Processing block 103/1000 ... OK</span>
<span class="co">#&gt; Processing block 104/1000 ... OK</span>
<span class="co">#&gt; Processing block 105/1000 ... OK</span>
<span class="co">#&gt; Processing block 106/1000 ... OK</span>
<span class="co">#&gt; Processing block 107/1000 ... OK</span>
<span class="co">#&gt; Processing block 108/1000 ... OK</span>
<span class="co">#&gt; Processing block 109/1000 ... OK</span>
<span class="co">#&gt; Processing block 110/1000 ... OK</span>
<span class="co">#&gt; Processing block 111/1000 ... OK</span>
<span class="co">#&gt; Processing block 112/1000 ... OK</span>
<span class="co">#&gt; Processing block 113/1000 ... OK</span>
<span class="co">#&gt; Processing block 114/1000 ... OK</span>
<span class="co">#&gt; Processing block 115/1000 ... OK</span>
<span class="co">#&gt; Processing block 116/1000 ... OK</span>
<span class="co">#&gt; Processing block 117/1000 ... OK</span>
<span class="co">#&gt; Processing block 118/1000 ... OK</span>
<span class="co">#&gt; Processing block 119/1000 ... OK</span>
<span class="co">#&gt; Processing block 120/1000 ... OK</span>
<span class="co">#&gt; Processing block 121/1000 ... OK</span>
<span class="co">#&gt; Processing block 122/1000 ... OK</span>
<span class="co">#&gt; Processing block 123/1000 ... OK</span>
<span class="co">#&gt; Processing block 124/1000 ... OK</span>
<span class="co">#&gt; Processing block 125/1000 ... OK</span>
<span class="co">#&gt; Processing block 126/1000 ... OK</span>
<span class="co">#&gt; Processing block 127/1000 ... OK</span>
<span class="co">#&gt; Processing block 128/1000 ... OK</span>
<span class="co">#&gt; Processing block 129/1000 ... OK</span>
<span class="co">#&gt; Processing block 130/1000 ... OK</span>
<span class="co">#&gt; Processing block 131/1000 ... OK</span>
<span class="co">#&gt; Processing block 132/1000 ... OK</span>
<span class="co">#&gt; Processing block 133/1000 ... OK</span>
<span class="co">#&gt; Processing block 134/1000 ... OK</span>
<span class="co">#&gt; Processing block 135/1000 ... OK</span>
<span class="co">#&gt; Processing block 136/1000 ... OK</span>
<span class="co">#&gt; Processing block 137/1000 ... OK</span>
<span class="co">#&gt; Processing block 138/1000 ... OK</span>
<span class="co">#&gt; Processing block 139/1000 ... OK</span>
<span class="co">#&gt; Processing block 140/1000 ... OK</span>
<span class="co">#&gt; Processing block 141/1000 ... OK</span>
<span class="co">#&gt; Processing block 142/1000 ... OK</span>
<span class="co">#&gt; Processing block 143/1000 ... OK</span>
<span class="co">#&gt; Processing block 144/1000 ... OK</span>
<span class="co">#&gt; Processing block 145/1000 ... OK</span>
<span class="co">#&gt; Processing block 146/1000 ... OK</span>
<span class="co">#&gt; Processing block 147/1000 ... OK</span>
<span class="co">#&gt; Processing block 148/1000 ... OK</span>
<span class="co">#&gt; Processing block 149/1000 ... OK</span>
<span class="co">#&gt; Processing block 150/1000 ... OK</span>
<span class="co">#&gt; Processing block 151/1000 ... OK</span>
<span class="co">#&gt; Processing block 152/1000 ... OK</span>
<span class="co">#&gt; Processing block 153/1000 ... OK</span>
<span class="co">#&gt; Processing block 154/1000 ... OK</span>
<span class="co">#&gt; Processing block 155/1000 ... OK</span>
<span class="co">#&gt; Processing block 156/1000 ... OK</span>
<span class="co">#&gt; Processing block 157/1000 ... OK</span>
<span class="co">#&gt; Processing block 158/1000 ... OK</span>
<span class="co">#&gt; Processing block 159/1000 ... OK</span>
<span class="co">#&gt; Processing block 160/1000 ... OK</span>
<span class="co">#&gt; Processing block 161/1000 ... OK</span>
<span class="co">#&gt; Processing block 162/1000 ... OK</span>
<span class="co">#&gt; Processing block 163/1000 ... OK</span>
<span class="co">#&gt; Processing block 164/1000 ... OK</span>
<span class="co">#&gt; Processing block 165/1000 ... OK</span>
<span class="co">#&gt; Processing block 166/1000 ... OK</span>
<span class="co">#&gt; Processing block 167/1000 ... OK</span>
<span class="co">#&gt; Processing block 168/1000 ... OK</span>
<span class="co">#&gt; Processing block 169/1000 ... OK</span>
<span class="co">#&gt; Processing block 170/1000 ... OK</span>
<span class="co">#&gt; Processing block 171/1000 ... OK</span>
<span class="co">#&gt; Processing block 172/1000 ... OK</span>
<span class="co">#&gt; Processing block 173/1000 ... OK</span>
<span class="co">#&gt; Processing block 174/1000 ... OK</span>
<span class="co">#&gt; Processing block 175/1000 ... OK</span>
<span class="co">#&gt; Processing block 176/1000 ... OK</span>
<span class="co">#&gt; Processing block 177/1000 ... OK</span>
<span class="co">#&gt; Processing block 178/1000 ... OK</span>
<span class="co">#&gt; Processing block 179/1000 ... OK</span>
<span class="co">#&gt; Processing block 180/1000 ... OK</span>
<span class="co">#&gt; Processing block 181/1000 ... OK</span>
<span class="co">#&gt; Processing block 182/1000 ... OK</span>
<span class="co">#&gt; Processing block 183/1000 ... OK</span>
<span class="co">#&gt; Processing block 184/1000 ... OK</span>
<span class="co">#&gt; Processing block 185/1000 ... OK</span>
<span class="co">#&gt; Processing block 186/1000 ... OK</span>
<span class="co">#&gt; Processing block 187/1000 ... OK</span>
<span class="co">#&gt; Processing block 188/1000 ... OK</span>
<span class="co">#&gt; Processing block 189/1000 ... OK</span>
<span class="co">#&gt; Processing block 190/1000 ... OK</span>
<span class="co">#&gt; Processing block 191/1000 ... OK</span>
<span class="co">#&gt; Processing block 192/1000 ... OK</span>
<span class="co">#&gt; Processing block 193/1000 ... OK</span>
<span class="co">#&gt; Processing block 194/1000 ... OK</span>
<span class="co">#&gt; Processing block 195/1000 ... OK</span>
<span class="co">#&gt; Processing block 196/1000 ... OK</span>
<span class="co">#&gt; Processing block 197/1000 ... OK</span>
<span class="co">#&gt; Processing block 198/1000 ... OK</span>
<span class="co">#&gt; Processing block 199/1000 ... OK</span>
<span class="co">#&gt; Processing block 200/1000 ... OK</span>
<span class="co">#&gt; Processing block 201/1000 ... OK</span>
<span class="co">#&gt; Processing block 202/1000 ... OK</span>
<span class="co">#&gt; Processing block 203/1000 ... OK</span>
<span class="co">#&gt; Processing block 204/1000 ... OK</span>
<span class="co">#&gt; Processing block 205/1000 ... OK</span>
<span class="co">#&gt; Processing block 206/1000 ... OK</span>
<span class="co">#&gt; Processing block 207/1000 ... OK</span>
<span class="co">#&gt; Processing block 208/1000 ... OK</span>
<span class="co">#&gt; Processing block 209/1000 ... OK</span>
<span class="co">#&gt; Processing block 210/1000 ... OK</span>
<span class="co">#&gt; Processing block 211/1000 ... OK</span>
<span class="co">#&gt; Processing block 212/1000 ... OK</span>
<span class="co">#&gt; Processing block 213/1000 ... OK</span>
<span class="co">#&gt; Processing block 214/1000 ... OK</span>
<span class="co">#&gt; Processing block 215/1000 ... OK</span>
<span class="co">#&gt; Processing block 216/1000 ... OK</span>
<span class="co">#&gt; Processing block 217/1000 ... OK</span>
<span class="co">#&gt; Processing block 218/1000 ... OK</span>
<span class="co">#&gt; Processing block 219/1000 ... OK</span>
<span class="co">#&gt; Processing block 220/1000 ... OK</span>
<span class="co">#&gt; Processing block 221/1000 ... OK</span>
<span class="co">#&gt; Processing block 222/1000 ... OK</span>
<span class="co">#&gt; Processing block 223/1000 ... OK</span>
<span class="co">#&gt; Processing block 224/1000 ... OK</span>
<span class="co">#&gt; Processing block 225/1000 ... OK</span>
<span class="co">#&gt; Processing block 226/1000 ... OK</span>
<span class="co">#&gt; Processing block 227/1000 ... OK</span>
<span class="co">#&gt; Processing block 228/1000 ... OK</span>
<span class="co">#&gt; Processing block 229/1000 ... OK</span>
<span class="co">#&gt; Processing block 230/1000 ... OK</span>
<span class="co">#&gt; Processing block 231/1000 ... OK</span>
<span class="co">#&gt; Processing block 232/1000 ... OK</span>
<span class="co">#&gt; Processing block 233/1000 ... OK</span>
<span class="co">#&gt; Processing block 234/1000 ... OK</span>
<span class="co">#&gt; Processing block 235/1000 ... OK</span>
<span class="co">#&gt; Processing block 236/1000 ... OK</span>
<span class="co">#&gt; Processing block 237/1000 ... OK</span>
<span class="co">#&gt; Processing block 238/1000 ... OK</span>
<span class="co">#&gt; Processing block 239/1000 ... OK</span>
<span class="co">#&gt; Processing block 240/1000 ... OK</span>
<span class="co">#&gt; Processing block 241/1000 ... OK</span>
<span class="co">#&gt; Processing block 242/1000 ... OK</span>
<span class="co">#&gt; Processing block 243/1000 ... OK</span>
<span class="co">#&gt; Processing block 244/1000 ... OK</span>
<span class="co">#&gt; Processing block 245/1000 ... OK</span>
<span class="co">#&gt; Processing block 246/1000 ... OK</span>
<span class="co">#&gt; Processing block 247/1000 ... OK</span>
<span class="co">#&gt; Processing block 248/1000 ... OK</span>
<span class="co">#&gt; Processing block 249/1000 ... OK</span>
<span class="co">#&gt; Processing block 250/1000 ... OK</span>
<span class="co">#&gt; Processing block 251/1000 ... OK</span>
<span class="co">#&gt; Processing block 252/1000 ... OK</span>
<span class="co">#&gt; Processing block 253/1000 ... OK</span>
<span class="co">#&gt; Processing block 254/1000 ... OK</span>
<span class="co">#&gt; Processing block 255/1000 ... OK</span>
<span class="co">#&gt; Processing block 256/1000 ... OK</span>
<span class="co">#&gt; Processing block 257/1000 ... OK</span>
<span class="co">#&gt; Processing block 258/1000 ... OK</span>
<span class="co">#&gt; Processing block 259/1000 ... OK</span>
<span class="co">#&gt; Processing block 260/1000 ... OK</span>
<span class="co">#&gt; Processing block 261/1000 ... OK</span>
<span class="co">#&gt; Processing block 262/1000 ... OK</span>
<span class="co">#&gt; Processing block 263/1000 ... OK</span>
<span class="co">#&gt; Processing block 264/1000 ... OK</span>
<span class="co">#&gt; Processing block 265/1000 ... OK</span>
<span class="co">#&gt; Processing block 266/1000 ... OK</span>
<span class="co">#&gt; Processing block 267/1000 ... OK</span>
<span class="co">#&gt; Processing block 268/1000 ... OK</span>
<span class="co">#&gt; Processing block 269/1000 ... OK</span>
<span class="co">#&gt; Processing block 270/1000 ... OK</span>
<span class="co">#&gt; Processing block 271/1000 ... OK</span>
<span class="co">#&gt; Processing block 272/1000 ... OK</span>
<span class="co">#&gt; Processing block 273/1000 ... OK</span>
<span class="co">#&gt; Processing block 274/1000 ... OK</span>
<span class="co">#&gt; Processing block 275/1000 ... OK</span>
<span class="co">#&gt; Processing block 276/1000 ... OK</span>
<span class="co">#&gt; Processing block 277/1000 ... OK</span>
<span class="co">#&gt; Processing block 278/1000 ... OK</span>
<span class="co">#&gt; Processing block 279/1000 ... OK</span>
<span class="co">#&gt; Processing block 280/1000 ... OK</span>
<span class="co">#&gt; Processing block 281/1000 ... OK</span>
<span class="co">#&gt; Processing block 282/1000 ... OK</span>
<span class="co">#&gt; Processing block 283/1000 ... OK</span>
<span class="co">#&gt; Processing block 284/1000 ... OK</span>
<span class="co">#&gt; Processing block 285/1000 ... OK</span>
<span class="co">#&gt; Processing block 286/1000 ... OK</span>
<span class="co">#&gt; Processing block 287/1000 ... OK</span>
<span class="co">#&gt; Processing block 288/1000 ... OK</span>
<span class="co">#&gt; Processing block 289/1000 ... OK</span>
<span class="co">#&gt; Processing block 290/1000 ... OK</span>
<span class="co">#&gt; Processing block 291/1000 ... OK</span>
<span class="co">#&gt; Processing block 292/1000 ... OK</span>
<span class="co">#&gt; Processing block 293/1000 ... OK</span>
<span class="co">#&gt; Processing block 294/1000 ... OK</span>
<span class="co">#&gt; Processing block 295/1000 ... OK</span>
<span class="co">#&gt; Processing block 296/1000 ... OK</span>
<span class="co">#&gt; Processing block 297/1000 ... OK</span>
<span class="co">#&gt; Processing block 298/1000 ... OK</span>
<span class="co">#&gt; Processing block 299/1000 ... OK</span>
<span class="co">#&gt; Processing block 300/1000 ... OK</span>
<span class="co">#&gt; Processing block 301/1000 ... OK</span>
<span class="co">#&gt; Processing block 302/1000 ... OK</span>
<span class="co">#&gt; Processing block 303/1000 ... OK</span>
<span class="co">#&gt; Processing block 304/1000 ... OK</span>
<span class="co">#&gt; Processing block 305/1000 ... OK</span>
<span class="co">#&gt; Processing block 306/1000 ... OK</span>
<span class="co">#&gt; Processing block 307/1000 ... OK</span>
<span class="co">#&gt; Processing block 308/1000 ... OK</span>
<span class="co">#&gt; Processing block 309/1000 ... OK</span>
<span class="co">#&gt; Processing block 310/1000 ... OK</span>
<span class="co">#&gt; Processing block 311/1000 ... OK</span>
<span class="co">#&gt; Processing block 312/1000 ... OK</span>
<span class="co">#&gt; Processing block 313/1000 ... OK</span>
<span class="co">#&gt; Processing block 314/1000 ... OK</span>
<span class="co">#&gt; Processing block 315/1000 ... OK</span>
<span class="co">#&gt; Processing block 316/1000 ... OK</span>
<span class="co">#&gt; Processing block 317/1000 ... OK</span>
<span class="co">#&gt; Processing block 318/1000 ... OK</span>
<span class="co">#&gt; Processing block 319/1000 ... OK</span>
<span class="co">#&gt; Processing block 320/1000 ... OK</span>
<span class="co">#&gt; Processing block 321/1000 ... OK</span>
<span class="co">#&gt; Processing block 322/1000 ... OK</span>
<span class="co">#&gt; Processing block 323/1000 ... OK</span>
<span class="co">#&gt; Processing block 324/1000 ... OK</span>
<span class="co">#&gt; Processing block 325/1000 ... OK</span>
<span class="co">#&gt; Processing block 326/1000 ... OK</span>
<span class="co">#&gt; Processing block 327/1000 ... OK</span>
<span class="co">#&gt; Processing block 328/1000 ... OK</span>
<span class="co">#&gt; Processing block 329/1000 ... OK</span>
<span class="co">#&gt; Processing block 330/1000 ... OK</span>
<span class="co">#&gt; Processing block 331/1000 ... OK</span>
<span class="co">#&gt; Processing block 332/1000 ... OK</span>
<span class="co">#&gt; Processing block 333/1000 ... OK</span>
<span class="co">#&gt; Processing block 334/1000 ... OK</span>
<span class="co">#&gt; Processing block 335/1000 ... OK</span>
<span class="co">#&gt; Processing block 336/1000 ... OK</span>
<span class="co">#&gt; Processing block 337/1000 ... OK</span>
<span class="co">#&gt; Processing block 338/1000 ... OK</span>
<span class="co">#&gt; Processing block 339/1000 ... OK</span>
<span class="co">#&gt; Processing block 340/1000 ... OK</span>
<span class="co">#&gt; Processing block 341/1000 ... OK</span>
<span class="co">#&gt; Processing block 342/1000 ... OK</span>
<span class="co">#&gt; Processing block 343/1000 ... OK</span>
<span class="co">#&gt; Processing block 344/1000 ... OK</span>
<span class="co">#&gt; Processing block 345/1000 ... OK</span>
<span class="co">#&gt; Processing block 346/1000 ... OK</span>
<span class="co">#&gt; Processing block 347/1000 ... OK</span>
<span class="co">#&gt; Processing block 348/1000 ... OK</span>
<span class="co">#&gt; Processing block 349/1000 ... OK</span>
<span class="co">#&gt; Processing block 350/1000 ... OK</span>
<span class="co">#&gt; Processing block 351/1000 ... OK</span>
<span class="co">#&gt; Processing block 352/1000 ... OK</span>
<span class="co">#&gt; Processing block 353/1000 ... OK</span>
<span class="co">#&gt; Processing block 354/1000 ... OK</span>
<span class="co">#&gt; Processing block 355/1000 ... OK</span>
<span class="co">#&gt; Processing block 356/1000 ... OK</span>
<span class="co">#&gt; Processing block 357/1000 ... OK</span>
<span class="co">#&gt; Processing block 358/1000 ... OK</span>
<span class="co">#&gt; Processing block 359/1000 ... OK</span>
<span class="co">#&gt; Processing block 360/1000 ... OK</span>
<span class="co">#&gt; Processing block 361/1000 ... OK</span>
<span class="co">#&gt; Processing block 362/1000 ... OK</span>
<span class="co">#&gt; Processing block 363/1000 ... OK</span>
<span class="co">#&gt; Processing block 364/1000 ... OK</span>
<span class="co">#&gt; Processing block 365/1000 ... OK</span>
<span class="co">#&gt; Processing block 366/1000 ... OK</span>
<span class="co">#&gt; Processing block 367/1000 ... OK</span>
<span class="co">#&gt; Processing block 368/1000 ... OK</span>
<span class="co">#&gt; Processing block 369/1000 ... OK</span>
<span class="co">#&gt; Processing block 370/1000 ... OK</span>
<span class="co">#&gt; Processing block 371/1000 ... OK</span>
<span class="co">#&gt; Processing block 372/1000 ... OK</span>
<span class="co">#&gt; Processing block 373/1000 ... OK</span>
<span class="co">#&gt; Processing block 374/1000 ... OK</span>
<span class="co">#&gt; Processing block 375/1000 ... OK</span>
<span class="co">#&gt; Processing block 376/1000 ... OK</span>
<span class="co">#&gt; Processing block 377/1000 ... OK</span>
<span class="co">#&gt; Processing block 378/1000 ... OK</span>
<span class="co">#&gt; Processing block 379/1000 ... OK</span>
<span class="co">#&gt; Processing block 380/1000 ... OK</span>
<span class="co">#&gt; Processing block 381/1000 ... OK</span>
<span class="co">#&gt; Processing block 382/1000 ... OK</span>
<span class="co">#&gt; Processing block 383/1000 ... OK</span>
<span class="co">#&gt; Processing block 384/1000 ... OK</span>
<span class="co">#&gt; Processing block 385/1000 ... OK</span>
<span class="co">#&gt; Processing block 386/1000 ... OK</span>
<span class="co">#&gt; Processing block 387/1000 ... OK</span>
<span class="co">#&gt; Processing block 388/1000 ... OK</span>
<span class="co">#&gt; Processing block 389/1000 ... OK</span>
<span class="co">#&gt; Processing block 390/1000 ... OK</span>
<span class="co">#&gt; Processing block 391/1000 ... OK</span>
<span class="co">#&gt; Processing block 392/1000 ... OK</span>
<span class="co">#&gt; Processing block 393/1000 ... OK</span>
<span class="co">#&gt; Processing block 394/1000 ... OK</span>
<span class="co">#&gt; Processing block 395/1000 ... OK</span>
<span class="co">#&gt; Processing block 396/1000 ... OK</span>
<span class="co">#&gt; Processing block 397/1000 ... OK</span>
<span class="co">#&gt; Processing block 398/1000 ... OK</span>
<span class="co">#&gt; Processing block 399/1000 ... OK</span>
<span class="co">#&gt; Processing block 400/1000 ... OK</span>
<span class="co">#&gt; Processing block 401/1000 ... OK</span>
<span class="co">#&gt; Processing block 402/1000 ... OK</span>
<span class="co">#&gt; Processing block 403/1000 ... OK</span>
<span class="co">#&gt; Processing block 404/1000 ... OK</span>
<span class="co">#&gt; Processing block 405/1000 ... OK</span>
<span class="co">#&gt; Processing block 406/1000 ... OK</span>
<span class="co">#&gt; Processing block 407/1000 ... OK</span>
<span class="co">#&gt; Processing block 408/1000 ... OK</span>
<span class="co">#&gt; Processing block 409/1000 ... OK</span>
<span class="co">#&gt; Processing block 410/1000 ... OK</span>
<span class="co">#&gt; Processing block 411/1000 ... OK</span>
<span class="co">#&gt; Processing block 412/1000 ... OK</span>
<span class="co">#&gt; Processing block 413/1000 ... OK</span>
<span class="co">#&gt; Processing block 414/1000 ... OK</span>
<span class="co">#&gt; Processing block 415/1000 ... OK</span>
<span class="co">#&gt; Processing block 416/1000 ... OK</span>
<span class="co">#&gt; Processing block 417/1000 ... OK</span>
<span class="co">#&gt; Processing block 418/1000 ... OK</span>
<span class="co">#&gt; Processing block 419/1000 ... OK</span>
<span class="co">#&gt; Processing block 420/1000 ... OK</span>
<span class="co">#&gt; Processing block 421/1000 ... OK</span>
<span class="co">#&gt; Processing block 422/1000 ... OK</span>
<span class="co">#&gt; Processing block 423/1000 ... OK</span>
<span class="co">#&gt; Processing block 424/1000 ... OK</span>
<span class="co">#&gt; Processing block 425/1000 ... OK</span>
<span class="co">#&gt; Processing block 426/1000 ... OK</span>
<span class="co">#&gt; Processing block 427/1000 ... OK</span>
<span class="co">#&gt; Processing block 428/1000 ... OK</span>
<span class="co">#&gt; Processing block 429/1000 ... OK</span>
<span class="co">#&gt; Processing block 430/1000 ... OK</span>
<span class="co">#&gt; Processing block 431/1000 ... OK</span>
<span class="co">#&gt; Processing block 432/1000 ... OK</span>
<span class="co">#&gt; Processing block 433/1000 ... OK</span>
<span class="co">#&gt; Processing block 434/1000 ... OK</span>
<span class="co">#&gt; Processing block 435/1000 ... OK</span>
<span class="co">#&gt; Processing block 436/1000 ... OK</span>
<span class="co">#&gt; Processing block 437/1000 ... OK</span>
<span class="co">#&gt; Processing block 438/1000 ... OK</span>
<span class="co">#&gt; Processing block 439/1000 ... OK</span>
<span class="co">#&gt; Processing block 440/1000 ... OK</span>
<span class="co">#&gt; Processing block 441/1000 ... OK</span>
<span class="co">#&gt; Processing block 442/1000 ... OK</span>
<span class="co">#&gt; Processing block 443/1000 ... OK</span>
<span class="co">#&gt; Processing block 444/1000 ... OK</span>
<span class="co">#&gt; Processing block 445/1000 ... OK</span>
<span class="co">#&gt; Processing block 446/1000 ... OK</span>
<span class="co">#&gt; Processing block 447/1000 ... OK</span>
<span class="co">#&gt; Processing block 448/1000 ... OK</span>
<span class="co">#&gt; Processing block 449/1000 ... OK</span>
<span class="co">#&gt; Processing block 450/1000 ... OK</span>
<span class="co">#&gt; Processing block 451/1000 ... OK</span>
<span class="co">#&gt; Processing block 452/1000 ... OK</span>
<span class="co">#&gt; Processing block 453/1000 ... OK</span>
<span class="co">#&gt; Processing block 454/1000 ... OK</span>
<span class="co">#&gt; Processing block 455/1000 ... OK</span>
<span class="co">#&gt; Processing block 456/1000 ... OK</span>
<span class="co">#&gt; Processing block 457/1000 ... OK</span>
<span class="co">#&gt; Processing block 458/1000 ... OK</span>
<span class="co">#&gt; Processing block 459/1000 ... OK</span>
<span class="co">#&gt; Processing block 460/1000 ... OK</span>
<span class="co">#&gt; Processing block 461/1000 ... OK</span>
<span class="co">#&gt; Processing block 462/1000 ... OK</span>
<span class="co">#&gt; Processing block 463/1000 ... OK</span>
<span class="co">#&gt; Processing block 464/1000 ... OK</span>
<span class="co">#&gt; Processing block 465/1000 ... OK</span>
<span class="co">#&gt; Processing block 466/1000 ... OK</span>
<span class="co">#&gt; Processing block 467/1000 ... OK</span>
<span class="co">#&gt; Processing block 468/1000 ... OK</span>
<span class="co">#&gt; Processing block 469/1000 ... OK</span>
<span class="co">#&gt; Processing block 470/1000 ... OK</span>
<span class="co">#&gt; Processing block 471/1000 ... OK</span>
<span class="co">#&gt; Processing block 472/1000 ... OK</span>
<span class="co">#&gt; Processing block 473/1000 ... OK</span>
<span class="co">#&gt; Processing block 474/1000 ... OK</span>
<span class="co">#&gt; Processing block 475/1000 ... OK</span>
<span class="co">#&gt; Processing block 476/1000 ... OK</span>
<span class="co">#&gt; Processing block 477/1000 ... OK</span>
<span class="co">#&gt; Processing block 478/1000 ... OK</span>
<span class="co">#&gt; Processing block 479/1000 ... OK</span>
<span class="co">#&gt; Processing block 480/1000 ... OK</span>
<span class="co">#&gt; Processing block 481/1000 ... OK</span>
<span class="co">#&gt; Processing block 482/1000 ... OK</span>
<span class="co">#&gt; Processing block 483/1000 ... OK</span>
<span class="co">#&gt; Processing block 484/1000 ... OK</span>
<span class="co">#&gt; Processing block 485/1000 ... OK</span>
<span class="co">#&gt; Processing block 486/1000 ... OK</span>
<span class="co">#&gt; Processing block 487/1000 ... OK</span>
<span class="co">#&gt; Processing block 488/1000 ... OK</span>
<span class="co">#&gt; Processing block 489/1000 ... OK</span>
<span class="co">#&gt; Processing block 490/1000 ... OK</span>
<span class="co">#&gt; Processing block 491/1000 ... OK</span>
<span class="co">#&gt; Processing block 492/1000 ... OK</span>
<span class="co">#&gt; Processing block 493/1000 ... OK</span>
<span class="co">#&gt; Processing block 494/1000 ... OK</span>
<span class="co">#&gt; Processing block 495/1000 ... OK</span>
<span class="co">#&gt; Processing block 496/1000 ... OK</span>
<span class="co">#&gt; Processing block 497/1000 ... OK</span>
<span class="co">#&gt; Processing block 498/1000 ... OK</span>
<span class="co">#&gt; Processing block 499/1000 ... OK</span>
<span class="co">#&gt; Processing block 500/1000 ... OK</span>
<span class="co">#&gt; Processing block 501/1000 ... OK</span>
<span class="co">#&gt; Processing block 502/1000 ... OK</span>
<span class="co">#&gt; Processing block 503/1000 ... OK</span>
<span class="co">#&gt; Processing block 504/1000 ... OK</span>
<span class="co">#&gt; Processing block 505/1000 ... OK</span>
<span class="co">#&gt; Processing block 506/1000 ... OK</span>
<span class="co">#&gt; Processing block 507/1000 ... OK</span>
<span class="co">#&gt; Processing block 508/1000 ... OK</span>
<span class="co">#&gt; Processing block 509/1000 ... OK</span>
<span class="co">#&gt; Processing block 510/1000 ... OK</span>
<span class="co">#&gt; Processing block 511/1000 ... OK</span>
<span class="co">#&gt; Processing block 512/1000 ... OK</span>
<span class="co">#&gt; Processing block 513/1000 ... OK</span>
<span class="co">#&gt; Processing block 514/1000 ... OK</span>
<span class="co">#&gt; Processing block 515/1000 ... OK</span>
<span class="co">#&gt; Processing block 516/1000 ... OK</span>
<span class="co">#&gt; Processing block 517/1000 ... OK</span>
<span class="co">#&gt; Processing block 518/1000 ... OK</span>
<span class="co">#&gt; Processing block 519/1000 ... OK</span>
<span class="co">#&gt; Processing block 520/1000 ... OK</span>
<span class="co">#&gt; Processing block 521/1000 ... OK</span>
<span class="co">#&gt; Processing block 522/1000 ... OK</span>
<span class="co">#&gt; Processing block 523/1000 ... OK</span>
<span class="co">#&gt; Processing block 524/1000 ... OK</span>
<span class="co">#&gt; Processing block 525/1000 ... OK</span>
<span class="co">#&gt; Processing block 526/1000 ... OK</span>
<span class="co">#&gt; Processing block 527/1000 ... OK</span>
<span class="co">#&gt; Processing block 528/1000 ... OK</span>
<span class="co">#&gt; Processing block 529/1000 ... OK</span>
<span class="co">#&gt; Processing block 530/1000 ... OK</span>
<span class="co">#&gt; Processing block 531/1000 ... OK</span>
<span class="co">#&gt; Processing block 532/1000 ... OK</span>
<span class="co">#&gt; Processing block 533/1000 ... OK</span>
<span class="co">#&gt; Processing block 534/1000 ... OK</span>
<span class="co">#&gt; Processing block 535/1000 ... OK</span>
<span class="co">#&gt; Processing block 536/1000 ... OK</span>
<span class="co">#&gt; Processing block 537/1000 ... OK</span>
<span class="co">#&gt; Processing block 538/1000 ... OK</span>
<span class="co">#&gt; Processing block 539/1000 ... OK</span>
<span class="co">#&gt; Processing block 540/1000 ... OK</span>
<span class="co">#&gt; Processing block 541/1000 ... OK</span>
<span class="co">#&gt; Processing block 542/1000 ... OK</span>
<span class="co">#&gt; Processing block 543/1000 ... OK</span>
<span class="co">#&gt; Processing block 544/1000 ... OK</span>
<span class="co">#&gt; Processing block 545/1000 ... OK</span>
<span class="co">#&gt; Processing block 546/1000 ... OK</span>
<span class="co">#&gt; Processing block 547/1000 ... OK</span>
<span class="co">#&gt; Processing block 548/1000 ... OK</span>
<span class="co">#&gt; Processing block 549/1000 ... OK</span>
<span class="co">#&gt; Processing block 550/1000 ... OK</span>
<span class="co">#&gt; Processing block 551/1000 ... OK</span>
<span class="co">#&gt; Processing block 552/1000 ... OK</span>
<span class="co">#&gt; Processing block 553/1000 ... OK</span>
<span class="co">#&gt; Processing block 554/1000 ... OK</span>
<span class="co">#&gt; Processing block 555/1000 ... OK</span>
<span class="co">#&gt; Processing block 556/1000 ... OK</span>
<span class="co">#&gt; Processing block 557/1000 ... OK</span>
<span class="co">#&gt; Processing block 558/1000 ... OK</span>
<span class="co">#&gt; Processing block 559/1000 ... OK</span>
<span class="co">#&gt; Processing block 560/1000 ... OK</span>
<span class="co">#&gt; Processing block 561/1000 ... OK</span>
<span class="co">#&gt; Processing block 562/1000 ... OK</span>
<span class="co">#&gt; Processing block 563/1000 ... OK</span>
<span class="co">#&gt; Processing block 564/1000 ... OK</span>
<span class="co">#&gt; Processing block 565/1000 ... OK</span>
<span class="co">#&gt; Processing block 566/1000 ... OK</span>
<span class="co">#&gt; Processing block 567/1000 ... OK</span>
<span class="co">#&gt; Processing block 568/1000 ... OK</span>
<span class="co">#&gt; Processing block 569/1000 ... OK</span>
<span class="co">#&gt; Processing block 570/1000 ... OK</span>
<span class="co">#&gt; Processing block 571/1000 ... OK</span>
<span class="co">#&gt; Processing block 572/1000 ... OK</span>
<span class="co">#&gt; Processing block 573/1000 ... OK</span>
<span class="co">#&gt; Processing block 574/1000 ... OK</span>
<span class="co">#&gt; Processing block 575/1000 ... OK</span>
<span class="co">#&gt; Processing block 576/1000 ... OK</span>
<span class="co">#&gt; Processing block 577/1000 ... OK</span>
<span class="co">#&gt; Processing block 578/1000 ... OK</span>
<span class="co">#&gt; Processing block 579/1000 ... OK</span>
<span class="co">#&gt; Processing block 580/1000 ... OK</span>
<span class="co">#&gt; Processing block 581/1000 ... OK</span>
<span class="co">#&gt; Processing block 582/1000 ... OK</span>
<span class="co">#&gt; Processing block 583/1000 ... OK</span>
<span class="co">#&gt; Processing block 584/1000 ... OK</span>
<span class="co">#&gt; Processing block 585/1000 ... OK</span>
<span class="co">#&gt; Processing block 586/1000 ... OK</span>
<span class="co">#&gt; Processing block 587/1000 ... OK</span>
<span class="co">#&gt; Processing block 588/1000 ... OK</span>
<span class="co">#&gt; Processing block 589/1000 ... OK</span>
<span class="co">#&gt; Processing block 590/1000 ... OK</span>
<span class="co">#&gt; Processing block 591/1000 ... OK</span>
<span class="co">#&gt; Processing block 592/1000 ... OK</span>
<span class="co">#&gt; Processing block 593/1000 ... OK</span>
<span class="co">#&gt; Processing block 594/1000 ... OK</span>
<span class="co">#&gt; Processing block 595/1000 ... OK</span>
<span class="co">#&gt; Processing block 596/1000 ... OK</span>
<span class="co">#&gt; Processing block 597/1000 ... OK</span>
<span class="co">#&gt; Processing block 598/1000 ... OK</span>
<span class="co">#&gt; Processing block 599/1000 ... OK</span>
<span class="co">#&gt; Processing block 600/1000 ... OK</span>
<span class="co">#&gt; Processing block 601/1000 ... OK</span>
<span class="co">#&gt; Processing block 602/1000 ... OK</span>
<span class="co">#&gt; Processing block 603/1000 ... OK</span>
<span class="co">#&gt; Processing block 604/1000 ... OK</span>
<span class="co">#&gt; Processing block 605/1000 ... OK</span>
<span class="co">#&gt; Processing block 606/1000 ... OK</span>
<span class="co">#&gt; Processing block 607/1000 ... OK</span>
<span class="co">#&gt; Processing block 608/1000 ... OK</span>
<span class="co">#&gt; Processing block 609/1000 ... OK</span>
<span class="co">#&gt; Processing block 610/1000 ... OK</span>
<span class="co">#&gt; Processing block 611/1000 ... OK</span>
<span class="co">#&gt; Processing block 612/1000 ... OK</span>
<span class="co">#&gt; Processing block 613/1000 ... OK</span>
<span class="co">#&gt; Processing block 614/1000 ... OK</span>
<span class="co">#&gt; Processing block 615/1000 ... OK</span>
<span class="co">#&gt; Processing block 616/1000 ... OK</span>
<span class="co">#&gt; Processing block 617/1000 ... OK</span>
<span class="co">#&gt; Processing block 618/1000 ... OK</span>
<span class="co">#&gt; Processing block 619/1000 ... OK</span>
<span class="co">#&gt; Processing block 620/1000 ... OK</span>
<span class="co">#&gt; Processing block 621/1000 ... OK</span>
<span class="co">#&gt; Processing block 622/1000 ... OK</span>
<span class="co">#&gt; Processing block 623/1000 ... OK</span>
<span class="co">#&gt; Processing block 624/1000 ... OK</span>
<span class="co">#&gt; Processing block 625/1000 ... OK</span>
<span class="co">#&gt; Processing block 626/1000 ... OK</span>
<span class="co">#&gt; Processing block 627/1000 ... OK</span>
<span class="co">#&gt; Processing block 628/1000 ... OK</span>
<span class="co">#&gt; Processing block 629/1000 ... OK</span>
<span class="co">#&gt; Processing block 630/1000 ... OK</span>
<span class="co">#&gt; Processing block 631/1000 ... OK</span>
<span class="co">#&gt; Processing block 632/1000 ... OK</span>
<span class="co">#&gt; Processing block 633/1000 ... OK</span>
<span class="co">#&gt; Processing block 634/1000 ... OK</span>
<span class="co">#&gt; Processing block 635/1000 ... OK</span>
<span class="co">#&gt; Processing block 636/1000 ... OK</span>
<span class="co">#&gt; Processing block 637/1000 ... OK</span>
<span class="co">#&gt; Processing block 638/1000 ... OK</span>
<span class="co">#&gt; Processing block 639/1000 ... OK</span>
<span class="co">#&gt; Processing block 640/1000 ... OK</span>
<span class="co">#&gt; Processing block 641/1000 ... OK</span>
<span class="co">#&gt; Processing block 642/1000 ... OK</span>
<span class="co">#&gt; Processing block 643/1000 ... OK</span>
<span class="co">#&gt; Processing block 644/1000 ... OK</span>
<span class="co">#&gt; Processing block 645/1000 ... OK</span>
<span class="co">#&gt; Processing block 646/1000 ... OK</span>
<span class="co">#&gt; Processing block 647/1000 ... OK</span>
<span class="co">#&gt; Processing block 648/1000 ... OK</span>
<span class="co">#&gt; Processing block 649/1000 ... OK</span>
<span class="co">#&gt; Processing block 650/1000 ... OK</span>
<span class="co">#&gt; Processing block 651/1000 ... OK</span>
<span class="co">#&gt; Processing block 652/1000 ... OK</span>
<span class="co">#&gt; Processing block 653/1000 ... OK</span>
<span class="co">#&gt; Processing block 654/1000 ... OK</span>
<span class="co">#&gt; Processing block 655/1000 ... OK</span>
<span class="co">#&gt; Processing block 656/1000 ... OK</span>
<span class="co">#&gt; Processing block 657/1000 ... OK</span>
<span class="co">#&gt; Processing block 658/1000 ... OK</span>
<span class="co">#&gt; Processing block 659/1000 ... OK</span>
<span class="co">#&gt; Processing block 660/1000 ... OK</span>
<span class="co">#&gt; Processing block 661/1000 ... OK</span>
<span class="co">#&gt; Processing block 662/1000 ... OK</span>
<span class="co">#&gt; Processing block 663/1000 ... OK</span>
<span class="co">#&gt; Processing block 664/1000 ... OK</span>
<span class="co">#&gt; Processing block 665/1000 ... OK</span>
<span class="co">#&gt; Processing block 666/1000 ... OK</span>
<span class="co">#&gt; Processing block 667/1000 ... OK</span>
<span class="co">#&gt; Processing block 668/1000 ... OK</span>
<span class="co">#&gt; Processing block 669/1000 ... OK</span>
<span class="co">#&gt; Processing block 670/1000 ... OK</span>
<span class="co">#&gt; Processing block 671/1000 ... OK</span>
<span class="co">#&gt; Processing block 672/1000 ... OK</span>
<span class="co">#&gt; Processing block 673/1000 ... OK</span>
<span class="co">#&gt; Processing block 674/1000 ... OK</span>
<span class="co">#&gt; Processing block 675/1000 ... OK</span>
<span class="co">#&gt; Processing block 676/1000 ... OK</span>
<span class="co">#&gt; Processing block 677/1000 ... OK</span>
<span class="co">#&gt; Processing block 678/1000 ... OK</span>
<span class="co">#&gt; Processing block 679/1000 ... OK</span>
<span class="co">#&gt; Processing block 680/1000 ... OK</span>
<span class="co">#&gt; Processing block 681/1000 ... OK</span>
<span class="co">#&gt; Processing block 682/1000 ... OK</span>
<span class="co">#&gt; Processing block 683/1000 ... OK</span>
<span class="co">#&gt; Processing block 684/1000 ... OK</span>
<span class="co">#&gt; Processing block 685/1000 ... OK</span>
<span class="co">#&gt; Processing block 686/1000 ... OK</span>
<span class="co">#&gt; Processing block 687/1000 ... OK</span>
<span class="co">#&gt; Processing block 688/1000 ... OK</span>
<span class="co">#&gt; Processing block 689/1000 ... OK</span>
<span class="co">#&gt; Processing block 690/1000 ... OK</span>
<span class="co">#&gt; Processing block 691/1000 ... OK</span>
<span class="co">#&gt; Processing block 692/1000 ... OK</span>
<span class="co">#&gt; Processing block 693/1000 ... OK</span>
<span class="co">#&gt; Processing block 694/1000 ... OK</span>
<span class="co">#&gt; Processing block 695/1000 ... OK</span>
<span class="co">#&gt; Processing block 696/1000 ... OK</span>
<span class="co">#&gt; Processing block 697/1000 ... OK</span>
<span class="co">#&gt; Processing block 698/1000 ... OK</span>
<span class="co">#&gt; Processing block 699/1000 ... OK</span>
<span class="co">#&gt; Processing block 700/1000 ... OK</span>
<span class="co">#&gt; Processing block 701/1000 ... OK</span>
<span class="co">#&gt; Processing block 702/1000 ... OK</span>
<span class="co">#&gt; Processing block 703/1000 ... OK</span>
<span class="co">#&gt; Processing block 704/1000 ... OK</span>
<span class="co">#&gt; Processing block 705/1000 ... OK</span>
<span class="co">#&gt; Processing block 706/1000 ... OK</span>
<span class="co">#&gt; Processing block 707/1000 ... OK</span>
<span class="co">#&gt; Processing block 708/1000 ... OK</span>
<span class="co">#&gt; Processing block 709/1000 ... OK</span>
<span class="co">#&gt; Processing block 710/1000 ... OK</span>
<span class="co">#&gt; Processing block 711/1000 ... OK</span>
<span class="co">#&gt; Processing block 712/1000 ... OK</span>
<span class="co">#&gt; Processing block 713/1000 ... OK</span>
<span class="co">#&gt; Processing block 714/1000 ... OK</span>
<span class="co">#&gt; Processing block 715/1000 ... OK</span>
<span class="co">#&gt; Processing block 716/1000 ... OK</span>
<span class="co">#&gt; Processing block 717/1000 ... OK</span>
<span class="co">#&gt; Processing block 718/1000 ... OK</span>
<span class="co">#&gt; Processing block 719/1000 ... OK</span>
<span class="co">#&gt; Processing block 720/1000 ... OK</span>
<span class="co">#&gt; Processing block 721/1000 ... OK</span>
<span class="co">#&gt; Processing block 722/1000 ... OK</span>
<span class="co">#&gt; Processing block 723/1000 ... OK</span>
<span class="co">#&gt; Processing block 724/1000 ... OK</span>
<span class="co">#&gt; Processing block 725/1000 ... OK</span>
<span class="co">#&gt; Processing block 726/1000 ... OK</span>
<span class="co">#&gt; Processing block 727/1000 ... OK</span>
<span class="co">#&gt; Processing block 728/1000 ... OK</span>
<span class="co">#&gt; Processing block 729/1000 ... OK</span>
<span class="co">#&gt; Processing block 730/1000 ... OK</span>
<span class="co">#&gt; Processing block 731/1000 ... OK</span>
<span class="co">#&gt; Processing block 732/1000 ... OK</span>
<span class="co">#&gt; Processing block 733/1000 ... OK</span>
<span class="co">#&gt; Processing block 734/1000 ... OK</span>
<span class="co">#&gt; Processing block 735/1000 ... OK</span>
<span class="co">#&gt; Processing block 736/1000 ... OK</span>
<span class="co">#&gt; Processing block 737/1000 ... OK</span>
<span class="co">#&gt; Processing block 738/1000 ... OK</span>
<span class="co">#&gt; Processing block 739/1000 ... OK</span>
<span class="co">#&gt; Processing block 740/1000 ... OK</span>
<span class="co">#&gt; Processing block 741/1000 ... OK</span>
<span class="co">#&gt; Processing block 742/1000 ... OK</span>
<span class="co">#&gt; Processing block 743/1000 ... OK</span>
<span class="co">#&gt; Processing block 744/1000 ... OK</span>
<span class="co">#&gt; Processing block 745/1000 ... OK</span>
<span class="co">#&gt; Processing block 746/1000 ... OK</span>
<span class="co">#&gt; Processing block 747/1000 ... OK</span>
<span class="co">#&gt; Processing block 748/1000 ... OK</span>
<span class="co">#&gt; Processing block 749/1000 ... OK</span>
<span class="co">#&gt; Processing block 750/1000 ... OK</span>
<span class="co">#&gt; Processing block 751/1000 ... OK</span>
<span class="co">#&gt; Processing block 752/1000 ... OK</span>
<span class="co">#&gt; Processing block 753/1000 ... OK</span>
<span class="co">#&gt; Processing block 754/1000 ... OK</span>
<span class="co">#&gt; Processing block 755/1000 ... OK</span>
<span class="co">#&gt; Processing block 756/1000 ... OK</span>
<span class="co">#&gt; Processing block 757/1000 ... OK</span>
<span class="co">#&gt; Processing block 758/1000 ... OK</span>
<span class="co">#&gt; Processing block 759/1000 ... OK</span>
<span class="co">#&gt; Processing block 760/1000 ... OK</span>
<span class="co">#&gt; Processing block 761/1000 ... OK</span>
<span class="co">#&gt; Processing block 762/1000 ... OK</span>
<span class="co">#&gt; Processing block 763/1000 ... OK</span>
<span class="co">#&gt; Processing block 764/1000 ... OK</span>
<span class="co">#&gt; Processing block 765/1000 ... OK</span>
<span class="co">#&gt; Processing block 766/1000 ... OK</span>
<span class="co">#&gt; Processing block 767/1000 ... OK</span>
<span class="co">#&gt; Processing block 768/1000 ... OK</span>
<span class="co">#&gt; Processing block 769/1000 ... OK</span>
<span class="co">#&gt; Processing block 770/1000 ... OK</span>
<span class="co">#&gt; Processing block 771/1000 ... OK</span>
<span class="co">#&gt; Processing block 772/1000 ... OK</span>
<span class="co">#&gt; Processing block 773/1000 ... OK</span>
<span class="co">#&gt; Processing block 774/1000 ... OK</span>
<span class="co">#&gt; Processing block 775/1000 ... OK</span>
<span class="co">#&gt; Processing block 776/1000 ... OK</span>
<span class="co">#&gt; Processing block 777/1000 ... OK</span>
<span class="co">#&gt; Processing block 778/1000 ... OK</span>
<span class="co">#&gt; Processing block 779/1000 ... OK</span>
<span class="co">#&gt; Processing block 780/1000 ... OK</span>
<span class="co">#&gt; Processing block 781/1000 ... OK</span>
<span class="co">#&gt; Processing block 782/1000 ... OK</span>
<span class="co">#&gt; Processing block 783/1000 ... OK</span>
<span class="co">#&gt; Processing block 784/1000 ... OK</span>
<span class="co">#&gt; Processing block 785/1000 ... OK</span>
<span class="co">#&gt; Processing block 786/1000 ... OK</span>
<span class="co">#&gt; Processing block 787/1000 ... OK</span>
<span class="co">#&gt; Processing block 788/1000 ... OK</span>
<span class="co">#&gt; Processing block 789/1000 ... OK</span>
<span class="co">#&gt; Processing block 790/1000 ... OK</span>
<span class="co">#&gt; Processing block 791/1000 ... OK</span>
<span class="co">#&gt; Processing block 792/1000 ... OK</span>
<span class="co">#&gt; Processing block 793/1000 ... OK</span>
<span class="co">#&gt; Processing block 794/1000 ... OK</span>
<span class="co">#&gt; Processing block 795/1000 ... OK</span>
<span class="co">#&gt; Processing block 796/1000 ... OK</span>
<span class="co">#&gt; Processing block 797/1000 ... OK</span>
<span class="co">#&gt; Processing block 798/1000 ... OK</span>
<span class="co">#&gt; Processing block 799/1000 ... OK</span>
<span class="co">#&gt; Processing block 800/1000 ... OK</span>
<span class="co">#&gt; Processing block 801/1000 ... OK</span>
<span class="co">#&gt; Processing block 802/1000 ... OK</span>
<span class="co">#&gt; Processing block 803/1000 ... OK</span>
<span class="co">#&gt; Processing block 804/1000 ... OK</span>
<span class="co">#&gt; Processing block 805/1000 ... OK</span>
<span class="co">#&gt; Processing block 806/1000 ... OK</span>
<span class="co">#&gt; Processing block 807/1000 ... OK</span>
<span class="co">#&gt; Processing block 808/1000 ... OK</span>
<span class="co">#&gt; Processing block 809/1000 ... OK</span>
<span class="co">#&gt; Processing block 810/1000 ... OK</span>
<span class="co">#&gt; Processing block 811/1000 ... OK</span>
<span class="co">#&gt; Processing block 812/1000 ... OK</span>
<span class="co">#&gt; Processing block 813/1000 ... OK</span>
<span class="co">#&gt; Processing block 814/1000 ... OK</span>
<span class="co">#&gt; Processing block 815/1000 ... OK</span>
<span class="co">#&gt; Processing block 816/1000 ... OK</span>
<span class="co">#&gt; Processing block 817/1000 ... OK</span>
<span class="co">#&gt; Processing block 818/1000 ... OK</span>
<span class="co">#&gt; Processing block 819/1000 ... OK</span>
<span class="co">#&gt; Processing block 820/1000 ... OK</span>
<span class="co">#&gt; Processing block 821/1000 ... OK</span>
<span class="co">#&gt; Processing block 822/1000 ... OK</span>
<span class="co">#&gt; Processing block 823/1000 ... OK</span>
<span class="co">#&gt; Processing block 824/1000 ... OK</span>
<span class="co">#&gt; Processing block 825/1000 ... OK</span>
<span class="co">#&gt; Processing block 826/1000 ... OK</span>
<span class="co">#&gt; Processing block 827/1000 ... OK</span>
<span class="co">#&gt; Processing block 828/1000 ... OK</span>
<span class="co">#&gt; Processing block 829/1000 ... OK</span>
<span class="co">#&gt; Processing block 830/1000 ... OK</span>
<span class="co">#&gt; Processing block 831/1000 ... OK</span>
<span class="co">#&gt; Processing block 832/1000 ... OK</span>
<span class="co">#&gt; Processing block 833/1000 ... OK</span>
<span class="co">#&gt; Processing block 834/1000 ... OK</span>
<span class="co">#&gt; Processing block 835/1000 ... OK</span>
<span class="co">#&gt; Processing block 836/1000 ... OK</span>
<span class="co">#&gt; Processing block 837/1000 ... OK</span>
<span class="co">#&gt; Processing block 838/1000 ... OK</span>
<span class="co">#&gt; Processing block 839/1000 ... OK</span>
<span class="co">#&gt; Processing block 840/1000 ... OK</span>
<span class="co">#&gt; Processing block 841/1000 ... OK</span>
<span class="co">#&gt; Processing block 842/1000 ... OK</span>
<span class="co">#&gt; Processing block 843/1000 ... OK</span>
<span class="co">#&gt; Processing block 844/1000 ... OK</span>
<span class="co">#&gt; Processing block 845/1000 ... OK</span>
<span class="co">#&gt; Processing block 846/1000 ... OK</span>
<span class="co">#&gt; Processing block 847/1000 ... OK</span>
<span class="co">#&gt; Processing block 848/1000 ... OK</span>
<span class="co">#&gt; Processing block 849/1000 ... OK</span>
<span class="co">#&gt; Processing block 850/1000 ... OK</span>
<span class="co">#&gt; Processing block 851/1000 ... OK</span>
<span class="co">#&gt; Processing block 852/1000 ... OK</span>
<span class="co">#&gt; Processing block 853/1000 ... OK</span>
<span class="co">#&gt; Processing block 854/1000 ... OK</span>
<span class="co">#&gt; Processing block 855/1000 ... OK</span>
<span class="co">#&gt; Processing block 856/1000 ... OK</span>
<span class="co">#&gt; Processing block 857/1000 ... OK</span>
<span class="co">#&gt; Processing block 858/1000 ... OK</span>
<span class="co">#&gt; Processing block 859/1000 ... OK</span>
<span class="co">#&gt; Processing block 860/1000 ... OK</span>
<span class="co">#&gt; Processing block 861/1000 ... OK</span>
<span class="co">#&gt; Processing block 862/1000 ... OK</span>
<span class="co">#&gt; Processing block 863/1000 ... OK</span>
<span class="co">#&gt; Processing block 864/1000 ... OK</span>
<span class="co">#&gt; Processing block 865/1000 ... OK</span>
<span class="co">#&gt; Processing block 866/1000 ... OK</span>
<span class="co">#&gt; Processing block 867/1000 ... OK</span>
<span class="co">#&gt; Processing block 868/1000 ... OK</span>
<span class="co">#&gt; Processing block 869/1000 ... OK</span>
<span class="co">#&gt; Processing block 870/1000 ... OK</span>
<span class="co">#&gt; Processing block 871/1000 ... OK</span>
<span class="co">#&gt; Processing block 872/1000 ... OK</span>
<span class="co">#&gt; Processing block 873/1000 ... OK</span>
<span class="co">#&gt; Processing block 874/1000 ... OK</span>
<span class="co">#&gt; Processing block 875/1000 ... OK</span>
<span class="co">#&gt; Processing block 876/1000 ... OK</span>
<span class="co">#&gt; Processing block 877/1000 ... OK</span>
<span class="co">#&gt; Processing block 878/1000 ... OK</span>
<span class="co">#&gt; Processing block 879/1000 ... OK</span>
<span class="co">#&gt; Processing block 880/1000 ... OK</span>
<span class="co">#&gt; Processing block 881/1000 ... OK</span>
<span class="co">#&gt; Processing block 882/1000 ... OK</span>
<span class="co">#&gt; Processing block 883/1000 ... OK</span>
<span class="co">#&gt; Processing block 884/1000 ... OK</span>
<span class="co">#&gt; Processing block 885/1000 ... OK</span>
<span class="co">#&gt; Processing block 886/1000 ... OK</span>
<span class="co">#&gt; Processing block 887/1000 ... OK</span>
<span class="co">#&gt; Processing block 888/1000 ... OK</span>
<span class="co">#&gt; Processing block 889/1000 ... OK</span>
<span class="co">#&gt; Processing block 890/1000 ... OK</span>
<span class="co">#&gt; Processing block 891/1000 ... OK</span>
<span class="co">#&gt; Processing block 892/1000 ... OK</span>
<span class="co">#&gt; Processing block 893/1000 ... OK</span>
<span class="co">#&gt; Processing block 894/1000 ... OK</span>
<span class="co">#&gt; Processing block 895/1000 ... OK</span>
<span class="co">#&gt; Processing block 896/1000 ... OK</span>
<span class="co">#&gt; Processing block 897/1000 ... OK</span>
<span class="co">#&gt; Processing block 898/1000 ... OK</span>
<span class="co">#&gt; Processing block 899/1000 ... OK</span>
<span class="co">#&gt; Processing block 900/1000 ... OK</span>
<span class="co">#&gt; Processing block 901/1000 ... OK</span>
<span class="co">#&gt; Processing block 902/1000 ... OK</span>
<span class="co">#&gt; Processing block 903/1000 ... OK</span>
<span class="co">#&gt; Processing block 904/1000 ... OK</span>
<span class="co">#&gt; Processing block 905/1000 ... OK</span>
<span class="co">#&gt; Processing block 906/1000 ... OK</span>
<span class="co">#&gt; Processing block 907/1000 ... OK</span>
<span class="co">#&gt; Processing block 908/1000 ... OK</span>
<span class="co">#&gt; Processing block 909/1000 ... OK</span>
<span class="co">#&gt; Processing block 910/1000 ... OK</span>
<span class="co">#&gt; Processing block 911/1000 ... OK</span>
<span class="co">#&gt; Processing block 912/1000 ... OK</span>
<span class="co">#&gt; Processing block 913/1000 ... OK</span>
<span class="co">#&gt; Processing block 914/1000 ... OK</span>
<span class="co">#&gt; Processing block 915/1000 ... OK</span>
<span class="co">#&gt; Processing block 916/1000 ... OK</span>
<span class="co">#&gt; Processing block 917/1000 ... OK</span>
<span class="co">#&gt; Processing block 918/1000 ... OK</span>
<span class="co">#&gt; Processing block 919/1000 ... OK</span>
<span class="co">#&gt; Processing block 920/1000 ... OK</span>
<span class="co">#&gt; Processing block 921/1000 ... OK</span>
<span class="co">#&gt; Processing block 922/1000 ... OK</span>
<span class="co">#&gt; Processing block 923/1000 ... OK</span>
<span class="co">#&gt; Processing block 924/1000 ... OK</span>
<span class="co">#&gt; Processing block 925/1000 ... OK</span>
<span class="co">#&gt; Processing block 926/1000 ... OK</span>
<span class="co">#&gt; Processing block 927/1000 ... OK</span>
<span class="co">#&gt; Processing block 928/1000 ... OK</span>
<span class="co">#&gt; Processing block 929/1000 ... OK</span>
<span class="co">#&gt; Processing block 930/1000 ... OK</span>
<span class="co">#&gt; Processing block 931/1000 ... OK</span>
<span class="co">#&gt; Processing block 932/1000 ... OK</span>
<span class="co">#&gt; Processing block 933/1000 ... OK</span>
<span class="co">#&gt; Processing block 934/1000 ... OK</span>
<span class="co">#&gt; Processing block 935/1000 ... OK</span>
<span class="co">#&gt; Processing block 936/1000 ... OK</span>
<span class="co">#&gt; Processing block 937/1000 ... OK</span>
<span class="co">#&gt; Processing block 938/1000 ... OK</span>
<span class="co">#&gt; Processing block 939/1000 ... OK</span>
<span class="co">#&gt; Processing block 940/1000 ... OK</span>
<span class="co">#&gt; Processing block 941/1000 ... OK</span>
<span class="co">#&gt; Processing block 942/1000 ... OK</span>
<span class="co">#&gt; Processing block 943/1000 ... OK</span>
<span class="co">#&gt; Processing block 944/1000 ... OK</span>
<span class="co">#&gt; Processing block 945/1000 ... OK</span>
<span class="co">#&gt; Processing block 946/1000 ... OK</span>
<span class="co">#&gt; Processing block 947/1000 ... OK</span>
<span class="co">#&gt; Processing block 948/1000 ... OK</span>
<span class="co">#&gt; Processing block 949/1000 ... OK</span>
<span class="co">#&gt; Processing block 950/1000 ... OK</span>
<span class="co">#&gt; Processing block 951/1000 ... OK</span>
<span class="co">#&gt; Processing block 952/1000 ... OK</span>
<span class="co">#&gt; Processing block 953/1000 ... OK</span>
<span class="co">#&gt; Processing block 954/1000 ... OK</span>
<span class="co">#&gt; Processing block 955/1000 ... OK</span>
<span class="co">#&gt; Processing block 956/1000 ... OK</span>
<span class="co">#&gt; Processing block 957/1000 ... OK</span>
<span class="co">#&gt; Processing block 958/1000 ... OK</span>
<span class="co">#&gt; Processing block 959/1000 ... OK</span>
<span class="co">#&gt; Processing block 960/1000 ... OK</span>
<span class="co">#&gt; Processing block 961/1000 ... OK</span>
<span class="co">#&gt; Processing block 962/1000 ... OK</span>
<span class="co">#&gt; Processing block 963/1000 ... OK</span>
<span class="co">#&gt; Processing block 964/1000 ... OK</span>
<span class="co">#&gt; Processing block 965/1000 ... OK</span>
<span class="co">#&gt; Processing block 966/1000 ... OK</span>
<span class="co">#&gt; Processing block 967/1000 ... OK</span>
<span class="co">#&gt; Processing block 968/1000 ... OK</span>
<span class="co">#&gt; Processing block 969/1000 ... OK</span>
<span class="co">#&gt; Processing block 970/1000 ... OK</span>
<span class="co">#&gt; Processing block 971/1000 ... OK</span>
<span class="co">#&gt; Processing block 972/1000 ... OK</span>
<span class="co">#&gt; Processing block 973/1000 ... OK</span>
<span class="co">#&gt; Processing block 974/1000 ... OK</span>
<span class="co">#&gt; Processing block 975/1000 ... OK</span>
<span class="co">#&gt; Processing block 976/1000 ... OK</span>
<span class="co">#&gt; Processing block 977/1000 ... OK</span>
<span class="co">#&gt; Processing block 978/1000 ... OK</span>
<span class="co">#&gt; Processing block 979/1000 ... OK</span>
<span class="co">#&gt; Processing block 980/1000 ... OK</span>
<span class="co">#&gt; Processing block 981/1000 ... OK</span>
<span class="co">#&gt; Processing block 982/1000 ... OK</span>
<span class="co">#&gt; Processing block 983/1000 ... OK</span>
<span class="co">#&gt; Processing block 984/1000 ... OK</span>
<span class="co">#&gt; Processing block 985/1000 ... OK</span>
<span class="co">#&gt; Processing block 986/1000 ... OK</span>
<span class="co">#&gt; Processing block 987/1000 ... OK</span>
<span class="co">#&gt; Processing block 988/1000 ... OK</span>
<span class="co">#&gt; Processing block 989/1000 ... OK</span>
<span class="co">#&gt; Processing block 990/1000 ... OK</span>
<span class="co">#&gt; Processing block 991/1000 ... OK</span>
<span class="co">#&gt; Processing block 992/1000 ... OK</span>
<span class="co">#&gt; Processing block 993/1000 ... OK</span>
<span class="co">#&gt; Processing block 994/1000 ... OK</span>
<span class="co">#&gt; Processing block 995/1000 ... OK</span>
<span class="co">#&gt; Processing block 996/1000 ... OK</span>
<span class="co">#&gt; Processing block 997/1000 ... OK</span>
<span class="co">#&gt; Processing block 998/1000 ... OK</span>
<span class="co">#&gt; Processing block 999/1000 ... OK</span>
<span class="co">#&gt; Processing block 1000/1000 ... OK</span>
<span class="co">#&gt; Processing block [[1/2, 1/1]] ... OK</span>
<span class="co">#&gt; Processing block [[2/2, 1/1]] ... OK</span>
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
</div>
<div id="try-it-yourself-1" class="section level4">
<h4 class="hasAnchor">
<a href="#try-it-yourself-1" class="anchor"></a>Try it yourself</h4>
<ul>
<li>Try modifying <code>basic_colSums()</code> to define each block as a group of 100 columns.</li>
<li>Try modifying <code>basic_colSums()</code> to define each block using the <code><a href="https://rdrr.io/pkg/DelayedArray/man/AutoGrid.html">defaultAutoGrid()</a></code>.</li>
</ul>
</div>
</div>
<div id="realization" class="section level3">
<h3 class="hasAnchor">
<a href="#realization" class="anchor"></a>Realization</h3>
<p>To <em>realize</em> a <em>DelayedArray</em> object is to trigger execution of the delayed operations carried by the object and return the result as an ordinary (i.e. dense) or sparse array.</p>
<div id="realizing-in-memory" class="section level4">
<h4 class="hasAnchor">
<a href="#realizing-in-memory" class="anchor"></a>Realizing in-memory</h4>
<p>We can realize a <em>DelayedArray</em> in memory as an ordinary (i.e. dense) or sparse array.</p>
<p>To realize a <em>DelayedArray</em> as an ordinary array, we can call <code><a href="https://rdrr.io/r/base/array.html">as.array()</a></code> on it.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">tenx_subset_realized</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/array.html">as.array</a></span>(<span class="no">tenx_subset</span>)
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_subset_realized</span>)
<span class="co">#&gt; [1] "matrix" "array"</span></pre></body></html></div>
<p>You may have noticed that the <code>tenx_subset</code> data contains a lot of zero values. We might therefore opt to realize the data as a sparse matrix, specifically a sparse matrix from the <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> package.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">tenx_subset_sparse</span> <span class="kw">&lt;-</span> <span class="fu">as</span>(<span class="no">tenx_subset</span>, <span class="st">"sparseMatrix"</span>)
<span class="co">#&gt; Processing block 1/2 ... OK</span>
<span class="co">#&gt; Processing block 2/2 ... OK</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_subset_sparse</span>)
<span class="co">#&gt; [1] "dgCMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "Matrix"</span></pre></body></html></div>
<p>Realizing as a sparse matrix is particularly useful when the data are stored in a sparse <em>DelayedArray</em> object<a href="#fn13" class="footnote-ref" id="fnref13"><sup>13</sup></a>, such as the <em>TENxMatrix</em> object described below.</p>
</div>
<div id="realizing-to-disk" class="section level4">
<h4 class="hasAnchor">
<a href="#realizing-to-disk" class="anchor"></a>Realizing to disk</h4>
<p><a href="#realizing-in-memory">Realizing in-memory</a> realizes the entire object in memory, which could require too much RAM if the object is large<a href="#fn14" class="footnote-ref" id="fnref14"><sup>14</sup></a>. Therefore, a large <em>DelayedArray</em> object may preferably by realized to disk.</p>
<p>Here we will demonstrate by realizing to an HDF5 file, but we could also realize to another on-disk backend such as a TileDB array (see <a href="#realization-backends">Realization backends</a>). Realizing to an HDF5 file requires that the <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> package is installed.</p>
<p>We can realize to disk as a dense array in an HDF5 file (<em>HDF5Array</em>) or as a sparse array (<em>TENxMatrix</em>).</p>
<p>To realize a <em>DelayedArray</em> as a dense array in an HDF5 file, we can call <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array()</a></code> on it.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">tenx_subset_hdf5</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array</a></span>(<span class="no">tenx_subset</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span></pre></body></html></div>
<p>To realize a <em>DelayedArray</em> as a sparse array in an HDF5 file, we can call <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeTENxMatrix.html">writeTENxMatrix()</a></code> on it.</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="no">tenx_subset_sparse_hdf5</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/writeTENxMatrix.html">writeTENxMatrix</a></span>(<span class="no">tenx_subset</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span></pre></body></html></div>
<p>Notice that the process of realization used block processing, which avoids loading the entire dataset entire memory and that the results of these realizations are an <em>HDF5Matrix</em> and a <em>TENxMatrix</em>, respectively<a href="#fn15" class="footnote-ref" id="fnref15"><sup>15</sup></a>.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_subset_hdf5</span>)
<span class="co">#&gt; [1] "HDF5Matrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_subset_sparse_hdf5</span>)
<span class="co">#&gt; [1] "TENxMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span></pre></body></html></div>
<p>Furthermore, neither <code>tenx_subset_hdf5</code> nor <code>tenx_subset_sparse_hdf5</code> carry around the delayed operations of <code>tenx_subset</code> because these have been <em>realized</em> before writing the data</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="co"># Compare the trees of delayed operations.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx_subset_hdf5</span>)
<span class="co">#&gt; 27998x1000 integer: HDF5Matrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer: [seed] HDF5ArraySeed object</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx_subset_sparse_hdf5</span>)
<span class="co">#&gt; 27998x1000 integer, sparse: TENxMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer, sparse: [seed] TENxMatrixSeed object</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx_subset</span>)
<span class="co">#&gt; 27998x1000 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
<p>Used like this, <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array()</a></code> and <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeTENxMatrix.html">writeTENxMatrix()</a></code> will write their results to a file in the <em>HDF5 dump directory</em>, a dumping ground for automatically created HDF5 datasets. We can see a log of the operations that have written to the HDF5 dump directory using <code><a href="https://rdrr.io/pkg/HDF5Array/man/dump-management.html">showHDF5DumpLog()</a></code>.</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/dump-management.html">showHDF5DumpLog</a></span>()
<span class="co">#&gt; [2020-07-28 08:22:35] #1 In file '/tmp/RtmpQlTtWh/HDF5Array_dump/auto00001.h5': creation of dataset '/HDF5ArrayAUTO00001' (27998x1000:integer, chunkdims=5291x188, level=6)</span>
<span class="co">#&gt; [2020-07-28 08:22:38] #2 In file '/tmp/RtmpQlTtWh/HDF5Array_dump/auto00002.h5': creation of dataset '/HDF5ArrayAUTO00002/shape' (2:integer, chunkdims=2, level=0)</span>
<span class="co">#&gt; [2020-07-28 08:22:38] #3 In file '/tmp/RtmpQlTtWh/HDF5Array_dump/auto00002.h5': creation of dataset '/HDF5ArrayAUTO00002/data' (0:integer, chunkdims=16384, level=6)</span>
<span class="co">#&gt; [2020-07-28 08:22:38] #4 In file '/tmp/RtmpQlTtWh/HDF5Array_dump/auto00002.h5': creation of dataset '/HDF5ArrayAUTO00002/indices' (0:integer, chunkdims=16384, level=6)</span>
<span class="co">#&gt; [2020-07-28 08:22:38] #5 In file '/tmp/RtmpQlTtWh/HDF5Array_dump/auto00002.h5': creation of dataset '/HDF5ArrayAUTO00002/indptr' (0:integer, chunkdims=4096, level=0)</span></pre></body></html></div>
<p>Often, however, we will want full control of where and how the data are written to the HDF5 file<a href="#fn16" class="footnote-ref" id="fnref16"><sup>16</sup></a> and the <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array()</a></code> and <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeTENxMatrix.html">writeTENxMatrix()</a></code> functions give you full control over this and more.</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="co"># Write the data to a user-specified HDF5 file using maximum compression and </span>
<span class="co"># 'chunking' along the columns.</span>
<span class="no">my_hdf5_file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span>(<span class="kw">fileext</span> <span class="kw">=</span> <span class="st">".h5"</span>)
<span class="no">tenx_subset_my_file_hdf5</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array</a></span>(
  <span class="no">tenx_subset</span>,
  <span class="kw">filepath</span> <span class="kw">=</span> <span class="no">my_hdf5_file</span>,
  <span class="kw">chunkdim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">tenx_subset</span>), <span class="fl">1</span>),
  <span class="kw">level</span> <span class="kw">=</span> <span class="fl">9</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>

<span class="co"># Compare `tenx_subset_hdf5` to `tenx_subset_my_file_hdf5`</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span>(<span class="no">tenx_subset_hdf5</span>)
<span class="co">#&gt; An object of class "HDF5ArraySeed"</span>
<span class="co">#&gt; Slot "filepath":</span>
<span class="co">#&gt; [1] "/tmp/RtmpQlTtWh/HDF5Array_dump/auto00001.h5"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "name":</span>
<span class="co">#&gt; [1] "/HDF5ArrayAUTO00001"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "type":</span>
<span class="co">#&gt; [1] "integer"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "dim":</span>
<span class="co">#&gt; [1] 27998  1000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "chunkdim":</span>
<span class="co">#&gt; [1] 5291  188</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "first_val":</span>
<span class="co">#&gt; [1] 0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span>(<span class="no">tenx_subset_my_file_hdf5</span>)
<span class="co">#&gt; An object of class "HDF5ArraySeed"</span>
<span class="co">#&gt; Slot "filepath":</span>
<span class="co">#&gt; [1] "/tmp/RtmpQlTtWh/file431ef09529.h5"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "name":</span>
<span class="co">#&gt; [1] "/HDF5ArrayAUTO00003"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "type":</span>
<span class="co">#&gt; [1] "integer"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "dim":</span>
<span class="co">#&gt; [1] 27998  1000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "chunkdim":</span>
<span class="co">#&gt; [1] 27998     1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "first_val":</span>
<span class="co">#&gt; [1] 0</span></pre></body></html></div>
</div>
<div id="realization-backends" class="section level4">
<h4 class="hasAnchor">
<a href="#realization-backends" class="anchor"></a>Realization backends</h4>
<p>We’ve now seen that we can realize to an HDF5 file. This is called the HDF5Array ‘realization backend’ and is implemented in the <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> package. There are a few other realization backends to be aware of.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">supportedRealizationBackends</a></span>()
<span class="co">#&gt;        BACKEND      package  </span>
<span class="co">#&gt; 1     RleArray DelayedArray  </span>
<span class="co">#&gt; 2    HDF5Array    HDF5Array  </span>
<span class="co">#&gt; 3   TENxMatrix    HDF5Array</span></pre></body></html></div>
<p>The <em><a href="https://github.com/LTLA/TileDBArray">TileDBArray</a></em> provides a TileDB realization backend, but this packages is not yet available from Bioconductor.</p>
<p>There is also the <code>NULL</code> backend, which means the data are realized in memory as an ordinary <em>array</em> and then wrapped in a <em>DelayedArray</em>. This is the default realization backend upon loading/attaching the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package.</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">getRealizationBackend</a></span>()
<span class="co">#&gt; NULL</span></pre></body></html></div>
<p>The default realization backend can be altered with <code><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend()</a></code>.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="st">"HDF5Array"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">getRealizationBackend</a></span>()
<span class="co">#&gt; [1] "HDF5Array"</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="kw">NULL</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">getRealizationBackend</a></span>()
<span class="co">#&gt; NULL</span></pre></body></html></div>
<p>It can be important to know what your current realization backend is because it will be used implicitly by some functions. For example, matrix multiplication that involves a <em>DelayedMatrix</em> uses the current realization backend.</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="st">"HDF5Array"</span>)
<span class="no">tenx_subset</span> <span class="kw">%*%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">1</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_subset</span>))
<span class="co">#&gt; Realizing block 1/1 ... OK, writing it ... OK</span>
<span class="co">#&gt; &lt;27998 x 1&gt; matrix of class HDF5Matrix and type "double":</span>
<span class="co">#&gt;          [,1]</span>
<span class="co">#&gt;     [1,]   18</span>
<span class="co">#&gt;     [2,]    0</span>
<span class="co">#&gt;     [3,]    0</span>
<span class="co">#&gt;     [4,]    0</span>
<span class="co">#&gt;     [5,]    0</span>
<span class="co">#&gt;      ...    .</span>
<span class="co">#&gt; [27994,]    0</span>
<span class="co">#&gt; [27995,] 1122</span>
<span class="co">#&gt; [27996,]   93</span>
<span class="co">#&gt; [27997,]    0</span>
<span class="co">#&gt; [27998,]    1</span>

<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="kw">NULL</span>)
<span class="no">tenx_subset</span> <span class="kw">%*%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">1</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_subset</span>))
<span class="co">#&gt; &lt;27998 x 1&gt; matrix of class DelayedMatrix and type "double":</span>
<span class="co">#&gt;          [,1]</span>
<span class="co">#&gt;     [1,]   18</span>
<span class="co">#&gt;     [2,]    0</span>
<span class="co">#&gt;     [3,]    0</span>
<span class="co">#&gt;     [4,]    0</span>
<span class="co">#&gt;     [5,]    0</span>
<span class="co">#&gt;      ...    .</span>
<span class="co">#&gt; [27994,]    0</span>
<span class="co">#&gt; [27995,] 1122</span>
<span class="co">#&gt; [27996,]   93</span>
<span class="co">#&gt; [27997,]    0</span>
<span class="co">#&gt; [27998,]    1</span></pre></body></html></div>
</div>
<div id="the-realize-function" class="section level4">
<h4 class="hasAnchor">
<a href="#the-realize-function" class="anchor"></a>The <code>realize()</code> function</h4>
<p>We’ve seen that we can realize to the HDF5Array backend using <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array(tenx_subset)</a></code>. We could also use coercion by calling <code>as(tenx_subset, "HDF5Array")</code>. A third way of realizing a <em>DelayedArray</em> to an HDF5 file is with the <code><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize()</a></code> function.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize</a></span>(<span class="no">tenx_subset</span>, <span class="kw">BACKEND</span> <span class="kw">=</span> <span class="st">"HDF5Array"</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span></pre></body></html></div>
<p>So why might you use <code><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize()</a></code> instead of these other options? Because it allows us to easily switch out the realization backend and will defer to the current realization backend if none is supplied.</p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="co"># Realize to the current realization backend.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">getRealizationBackend</a></span>()
<span class="co">#&gt; NULL</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize</a></span>(<span class="no">tenx_subset</span>)
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span>

<span class="co"># # Realize as an RleArray.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="st">"RleArray"</span>)
<span class="no">tenx_subset_rlearray</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize</a></span>(<span class="no">tenx_subset</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>
<span class="no">tenx_subset_rlearray</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class RleMatrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span>

<span class="co"># Switch back to the default backend.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="kw">NULL</span>)</pre></body></html></div>
<p>This is probably most useful when writing package code so that you can allow the user control over the realization backend.</p>
</div>
</div>
</div>
<div id="workflow-tips-for-delayedarray-backed-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-tips-for-delayedarray-backed-analyses" class="anchor"></a>Workflow tips for DelayedArray-backed analyses</h2>
<p>We’ll conclude with some miscellaneous tips I’ve collected over the past few years of using DelayedArray-backed workflows.</p>
<p>To demonstrate, we’ll create a <em>SingleCellExperiment</em> object containing <code>tenx_subset</code> (a subset of the <code>tenx</code> counts data).</p>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html">SingleCellExperiment</a></span>(<span class="kw">assays</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">tenx_subset</span>))</pre></body></html></div>
<p>We term this an HDF5-backed SummarizedExperiment because:</p>
<ol style="list-style-type: decimal">
<li>A <em>SingleCellExperiment</em> is (a derivative of) a <em>SummarizedExperiment</em>.</li>
<li>The assay data are stored in an HDF5 file.</li>
</ol>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="no">sce</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 27998 1000 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames: NULL</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames: NULL</span>
<span class="co">#&gt; colData names(0):</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span>
<span class="co"># We'll discuss the use of `withDimnames = FALSE` shortly.</span>
<span class="fu">assay</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span></pre></body></html></div>
<p>To make the example a little bit more interesting, we’ll also normalize the data.</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scuttle</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html">computeLibraryFactors</a></span>(<span class="no">sce</span>)
<span class="co">#&gt; Processing block [[1/2, 1/1]] ... OK</span>
<span class="co">#&gt; Processing block [[2/2, 1/1]] ... OK</span>
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html">logNormCounts</a></span>(<span class="no">sce</span>)</pre></body></html></div>
<p>The resulting <em>SingleCellExperiment</em> object contains two assays - <code>counts</code> and <code>logcounts</code> - both of which are <em>DelayedMatrix</em> objects.</p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="fu">assays</span>(<span class="no">sce</span>)
<span class="co">#&gt; List of length 2</span>
<span class="co">#&gt; names(2): counts logcounts</span>

<span class="fu">assay</span>(<span class="no">sce</span>, <span class="st">"counts"</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span>
<span class="fu">assay</span>(<span class="no">sce</span>, <span class="st">"logcounts"</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "double":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3] ...   [,999]  [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0   .        0        0</span>
<span class="co">#&gt;     [2,]       0       0       0   .        0        0</span>
<span class="co">#&gt;     [3,]       0       0       0   .        0        0</span>
<span class="co">#&gt;     [4,]       0       0       0   .        0        0</span>
<span class="co">#&gt;     [5,]       0       0       0   .        0        0</span>
<span class="co">#&gt;      ...       .       .       .   .        .        .</span>
<span class="co">#&gt; [27994,] 0.00000 0.00000 0.00000   . 0.000000 0.000000</span>
<span class="co">#&gt; [27995,] 1.11025 0.00000 0.00000   . 1.201326 0.000000</span>
<span class="co">#&gt; [27996,] 0.00000 0.00000 0.00000   . 0.000000 0.000000</span>
<span class="co">#&gt; [27997,] 0.00000 0.00000 0.00000   . 0.000000 0.000000</span>
<span class="co">#&gt; [27998,] 0.00000 0.00000 0.00000   . 0.000000 0.000000</span></pre></body></html></div>
<div id="saving-and-loading-hdf5-backed-summarizedexperiment-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#saving-and-loading-hdf5-backed-summarizedexperiment-objects" class="anchor"></a>Saving and loading HDF5-backed SummarizedExperiment objects</h3>
<div id="short-version" class="section level4">
<h4 class="hasAnchor">
<a href="#short-version" class="anchor"></a>Short version</h4>
<p>Use <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code>, <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">quickResaveHDF5SummarizedExperiment()</a></code>, and <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">loadHDF5SummarizedExperiment()</a></code> rather than <code><a href="https://rdrr.io/r/base/readRDS.html">saveRDS()</a></code> and <code><a href="https://rdrr.io/r/base/readRDS.html">readRDS()</a></code> or <code><a href="https://rdrr.io/r/base/save.html">save()</a></code> and <code><a href="https://rdrr.io/r/base/load.html">load()</a></code> when saving/loading HDF5-backed <em>SummarizedExperiment</em> objects.</p>
<p>Here is an example:</p>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="co"># Specify the directory where you want to save the object.</span>
<span class="co"># Here we use a temporary directory.</span>
<span class="no">dir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>(), <span class="st">"my_h5_se"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment</a></span>(<span class="no">sce</span>, <span class="no">dir</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">#&gt; Start writing assay 1/2 to HDF5 file:</span>
<span class="co">#&gt;   /tmp/RtmpQlTtWh/my_h5_se/assays.h5</span>
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Finished writing assay 1/2 to HDF5 file:</span>
<span class="co">#&gt;   /tmp/RtmpQlTtWh/my_h5_se/assays.h5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Start writing assay 2/2 to HDF5 file:</span>
<span class="co">#&gt;   /tmp/RtmpQlTtWh/my_h5_se/assays.h5</span>
<span class="co">#&gt; Realizing block 1/3 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/3 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 3/3 ... OK, writing it ... OK</span>
<span class="co">#&gt; Finished writing assay 2/2 to HDF5 file:</span>
<span class="co">#&gt;   /tmp/RtmpQlTtWh/my_h5_se/assays.h5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Serialize SingleCellExperiment object to RDS file:</span>
<span class="co">#&gt;   /tmp/RtmpQlTtWh/my_h5_se/se.rds</span>

<span class="co"># Load the saved object.</span>
<span class="no">saved_sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">loadHDF5SummarizedExperiment</a></span>(<span class="no">dir</span>)</pre></body></html></div>
<p>Note that this directory is relocatable i.e. it can be moved (or copied) to a different place, on the same or a different computer, before calling <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">loadHDF5SummarizedExperiment()</a></code> on it. For convenient sharing with collaborators, it is suggested to turn it into a tarball (with Unix command tar), or zip file, before the transfer<a href="#fn17" class="footnote-ref" id="fnref17"><sup>17</sup></a>.</p>
<p>Calling <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code> will realize any delayed operations prior to saving the assay data in an HDF5 file, as illustrated below.</p>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="co"># Compare the trees of delayed operations.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>))
<span class="co">#&gt; 27998x1000 double: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 double: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1000 double: Aperm (perm=c(2,1))</span>
<span class="co">#&gt;       └─ 1000x27998 double: Unary iso op with args</span>
<span class="co">#&gt;          └─ 1000x27998 integer: Aperm (perm=c(2,1))</span>
<span class="co">#&gt;             └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;                └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span>(<span class="no">saved_sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>))
<span class="co">#&gt; 27998x1000 double: HDF5Matrix object</span>
<span class="co">#&gt; └─ 27998x1000 double: [seed] HDF5ArraySeed object</span></pre></body></html></div>
<p>Finally please note that, depending on the size of the data to write to disk and the performance of the disk, <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code> can take a long time to complete<a href="#fn18" class="footnote-ref" id="fnref18"><sup>18</sup></a>.</p>
<p>The <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">quickResaveHDF5SummarizedExperiment()</a></code> function can be useful if you have updated a <em>SummarizedExperiment</em> already created by an earlier call to <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code>. For example, suppose you created an HDF5-backed <em>SummarizedExperiment</em>, do some pre-processing of the data, and want to save the result.</p>
<p>Here is the object before the pre-processing:</p>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="no">saved_sce</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 27998 1000 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(2): counts logcounts</span>
<span class="co">#&gt; rownames: NULL</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames: NULL</span>
<span class="co">#&gt; colData names(1): sizeFactor</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></pre></body></html></div>
<p>Now, let’s mock up pre-processing by adding sample metadata to the <em>colData</em>, and excluding certain features (rows) and samples (columns).</p>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="co"># Mock adding sample metadata.</span>
<span class="no">saved_sce</span>$<span class="no">sample</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"S1"</span>, <span class="fl">400</span>), <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"S2"</span>, <span class="fl">600</span>))

<span class="co"># Mock excluding certain features and samples.</span>
<span class="no">keep_feature</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">rbinom</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">sce</span>), <span class="fl">1</span>, <span class="fl">0.95</span>)
<span class="no">keep_sample</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">rbinom</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">sce</span>), <span class="fl">1</span>, <span class="fl">0.8</span>)
<span class="no">saved_sce</span> <span class="kw">&lt;-</span> <span class="no">saved_sce</span>[<span class="no">keep_feature</span>, <span class="no">keep_sample</span>]</pre></body></html></div>
<p>This is the object after the pre-processing:</p>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="no">saved_sce</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 26540 764 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(2): counts logcounts</span>
<span class="co">#&gt; rownames: NULL</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames: NULL</span>
<span class="co">#&gt; colData names(2): sizeFactor sample</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></pre></body></html></div>
<p>We can use <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">quickResaveHDF5SummarizedExperiment()</a></code> quickly re-save the pre-processed object. This is generally much faster than the initial call to <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code> because it does not re-write the assay data to an HDF5 file.</p>
<div class="sourceCode" id="cb55"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">quickResaveHDF5SummarizedExperiment</a></span>(<span class="no">saved_sce</span>)</pre></body></html></div>
<p>Notice that the pre-processing is preserved when we re-load the re-saved HDF5-backed <em>SummarizedExperiment</em>.</p>
<div class="sourceCode" id="cb56"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">loadHDF5SummarizedExperiment</a></span>(<span class="no">dir</span>)
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 26540 764 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(2): counts logcounts</span>
<span class="co">#&gt; rownames: NULL</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames: NULL</span>
<span class="co">#&gt; colData names(2): sizeFactor sample</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></pre></body></html></div>
</div>
<div id="long-version" class="section level4">
<h4 class="hasAnchor">
<a href="#long-version" class="anchor"></a>Long version</h4>
<p>A HDF5-backed <em>SummarizedExperiment</em> is a light-weight shell (the <em>SummarizedExperiment</em>) around a large disk-backed data matrix (the <em>HDF5Matrix</em>). The following explanation comes from <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">help("saveHDF5SummarizedExperiment", "HDF5Array")</a></code>:</p>
<p>Roughly speaking, <code><a href="https://rdrr.io/r/base/readRDS.html">saveRDS()</a></code> only serializes the part of an object that resides in memory<a href="#fn19" class="footnote-ref" id="fnref19"><sup>19</sup></a>. For most objects in R, that’s the whole object, so <code><a href="https://rdrr.io/r/base/readRDS.html">saveRDS()</a></code> does the job.</p>
<p>However some objects are pointing to on-disk data. For example:</p>
<ul>
<li>A <em>TxDb</em> object from the <em><a href="https://bioconductor.org/packages/3.12/GenomicFeatures">GenomicFeatures</a></em> points to an SQLite database</li>
<li>An <em>HDF5Array</em> object points to a dataset in an HDF5 file</li>
<li>A <em>SummarizedExperiment</em> derivative can have one or more of its assays that point to datasets (one per assay) in an HDF5 file.</li>
</ul>
<p>These objects have 2 parts: one part is in memory, and one part is on disk. The 1st part is sometimes called the object shell and is generally thin (i.e. it has a small memory footprint). The 2nd part is the data and is typically big. The object shell and data are linked together via some kind of pointer stored in the shell (e.g. an SQLite connection, or a path to a file, etc.). Note that this is a one way link in the sense that the object shell ‘knows’ where to find the on-disk data but the on-disk data knows nothing about the object shell (and is completely agnostic about what kind of object shell could be pointing to it). Furthermore, at any given time on a given system, there could be more than one object shell pointing to the same on-disk data. These object shells could exist in the same R session or in sessions in other languages (e.g. Python). These various sessions could be run by the same or by different users.</p>
<p>Using <code><a href="https://rdrr.io/r/base/readRDS.html">saveRDS()</a></code> on such object will only serialize the shell part so will produce a small <code>.rds</code> file that contains the serialized object shell but not the object data.</p>
<p>This is problematic because:</p>
<ol style="list-style-type: decimal">
<li>If you later unserialize the object (with <code><a href="https://rdrr.io/r/base/readRDS.html">readRDS()</a></code>) on the same system where you originally serialized it, it is possible that you will get back an object that is fully functional and semantically equivalent to the original object. But here is the catch: this will be the case <strong>ONLY</strong> if the data is still at the original location and has not been modified (i.e. nobody wrote or altered the data in the SQLite database or HDF5 file in the mean time), and if the serialization/unserialization cycle didn’t break the link between the object shell and the data (this serialization/unserialization cycle is known to break open SQLite connections).</li>
<li>After serialization the object shell and data are stored in separate files (in the new <code>.rds</code> file for the shell, still in the original SQLite or HDF5 file for the data), typically in very different places on the file system. But these 2 files are not relocatable, that is, moving or copying them to another system or sending them to collaborators will typically break the link between them. Concretely this means that the object obtained by using <code><a href="https://rdrr.io/r/base/readRDS.html">readRDS()</a></code> on the destination system will be broken.</li>
</ol>
<p><code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code> addresses these issues by saving the object shell and assay data in a folder that is relocatable.</p>
<p>Note that it only works on <em>SummarizedExperiment</em> derivatives. What it does exactly is:</p>
<ol style="list-style-type: decimal">
<li>Write all the assay data to an HDF5 file</li>
<li>Serialize the object shell, which in this case is everything in the object that is not the assay data.</li>
</ol>
<p>The 2 files (HDF5 and <code>.rds</code>) are written to the directory specified by the user. The resulting directory contains a full representation of the object and is relocatable, that is, it can be moved or copied to another place on the system, or to another system (possibly after making a tarball of it), where <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">loadHDF5SummarizedExperiment()</a></code> can then be used to load the object back in R.</p>
<p><code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">quickResaveHDF5SummarizedExperiment()</a></code> preserves the HDF5 file and datasets that the assays in the <em>SummarizedExperiment</em> are already pointing to (and which were created by an earlier call to <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code>). All it does is re-serialize the <em>SummarizedExperiment</em> on top of the <code>.rds</code> file that is associated with this HDF5 file (and which was created by an earlier call to <code>saveHDF5SummarizedExperiment(</code> or <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">quickResaveHDF5SummarizedExperiment()</a></code>). Because the delayed operations possibly carried by the assays in x are not realized, this is very fast.</p>
</div>
</div>
<div id="block-geometry" class="section level3">
<h3 class="hasAnchor">
<a href="#block-geometry" class="anchor"></a>Block geometry</h3>
<p>The block geometry (size and shape) are key determinants of performance when designing/applying functions to <em>DelayedArray</em> objects. For example, functions may be faster if fewer blocks are required (e.g., to minimise reading data from disk) or a function may require a sample’s complete data (i.e. a full column of data) to work at all.</p>
<p>A function should enforce a ‘minimal’ block geometry (i.e. the geometry required to produce a correct result) whilst also allowing the user to alter the block geometry for improved performance on their system and, ideally, offering some sort of automatic block geometry that is reasonably performant in most applications of the function.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> sets some default values that control the geometry of the automatic blocks. These are the automatic block size and the block shape, of which the block size is more relevant to the user.</p>
<div id="block-size" class="section level4">
<h4 class="hasAnchor">
<a href="#block-size" class="anchor"></a>Block size</h4>
<p>The <code><a href="https://rdrr.io/pkg/DelayedArray/man/AutoBlock-global-settings.html">getAutoBlockSize()</a></code> gives the automatic size in bytes of a block used when performing automatic block processing. By default, this is set to 10^{8} meaning each block can use up to 100 Mb of data.</p>
<p>Using fewer, larger blocks generally means faster performance (at the cost of higher peak memory usage). Conversely, using more, smaller blocks generally means slower performance (at the benefit of lower peak memory usage). Therefore, a user may wish to increase/decrease this on machines with sufficient memory by using <code><a href="https://rdrr.io/pkg/DelayedArray/man/AutoBlock-global-settings.html">setAutoBlockSize()</a></code> or by specifying the block size in functions that accept this as an argument.</p>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)))
<span class="co">#&gt; Processing block [[1/2, 1/1]] ... OK</span>
<span class="co">#&gt; Processing block [[2/2, 1/1]] ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.357   0.040   0.398</span>

<span class="co"># Increasing the block size 10-fold.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/AutoBlock-global-settings.html">setAutoBlockSize</a></span>(<span class="fl">1e9</span>)
<span class="co">#&gt; automatic block size set to 1e+09 bytes (was 1e+08)</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)))
<span class="co">#&gt; Processing block [[1/1, 1/1]] ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.327   0.074   0.401</span>

<span class="co"># Decreasing the default block size 100-fold</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/AutoBlock-global-settings.html">setAutoBlockSize</a></span>(<span class="fl">1e6</span>)
<span class="co">#&gt; automatic block size set to 1e+06 bytes (was 1e+09)</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)))
<span class="co">#&gt; Processing block [[1/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[2/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[3/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[4/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[5/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[6/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[7/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[8/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[9/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[10/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[11/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[12/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[13/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[14/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[15/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[16/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[17/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[18/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[19/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[20/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[21/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[22/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[23/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[24/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[25/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[26/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[27/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[28/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[29/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[30/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[31/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[32/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[33/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[34/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[35/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[36/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[37/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[38/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[39/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[40/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[41/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[42/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[43/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[44/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[45/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[46/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[47/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[48/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[49/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[50/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[51/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[52/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[53/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[54/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[55/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[56/56, 1/2]] ... OK</span>
<span class="co">#&gt; Processing block [[1/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[2/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[3/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[4/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[5/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[6/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[7/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[8/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[9/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[10/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[11/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[12/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[13/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[14/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[15/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[16/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[17/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[18/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[19/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[20/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[21/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[22/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[23/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[24/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[25/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[26/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[27/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[28/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[29/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[30/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[31/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[32/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[33/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[34/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[35/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[36/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[37/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[38/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[39/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[40/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[41/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[42/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[43/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[44/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[45/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[46/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[47/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[48/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[49/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[50/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[51/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[52/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[53/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[54/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[55/56, 2/2]] ... OK</span>
<span class="co">#&gt; Processing block [[56/56, 2/2]] ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.311   0.022   1.333</span>

<span class="co"># Reverting to default block size.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/AutoBlock-global-settings.html">setAutoBlockSize</a></span>(<span class="fl">1e8</span>)
<span class="co">#&gt; automatic block size set to 1e+08 bytes (was 1e+06)</span></pre></body></html></div>
</div>
<div id="block-shape" class="section level4">
<h4 class="hasAnchor">
<a href="#block-shape" class="anchor"></a>Block shape</h4>
<p>The choice of block shape is typically be more relevant to the function’s designer rather than its users. It is desirable to have functions that can handle arbitrary block shapes. However, this is not always possible, such as a function that requires that the data be processed by column.</p>
</div>
</div>
<div id="chunking" class="section level3">
<h3 class="hasAnchor">
<a href="#chunking" class="anchor"></a>Chunking</h3>
<p>Data stored on disk, such as in an HDF5 file, are usually ‘chunked’ into sub-matrices (for matrix data) or hyper-cubes (for arrays of higher dimension) to allow for more efficient subset selection. For example, we could choose to chunk an <span class="math inline">\(R \times C\)</span> matrix by column, by row, or into <span class="math inline">\(r \times c\)</span> sub-matrices (<span class="math inline">\(r \leq R, c \leq C\)</span>).</p>
<p>We’ll illustrate chunking with the HDF5 format, but similar concepts exist for other disk-backed data (e.g., <em>TileDBArray</em>, <em><a href="https://bioconductor.org/packages/3.12/matter">matter</a></em>) and even to in-memory data (e.g., <em>RleArray</em>).</p>
<div id="chunk-geometry" class="section level4">
<h4 class="hasAnchor">
<a href="#chunk-geometry" class="anchor"></a>Chunk geometry</h4>
<p>With the <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> package, chunking can be controlled by the <code>chunkdim</code> argument when writing data to disk using the <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array()</a></code> or <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code> functions.</p>
<p>In general, you want your data to be chunked in a manner that supports the type of access patterns you will be subsequently making during your analysis. For example, if you know you only need to access data by column then chunk the data by column. Of course, you often either don’t know in advance what access patterns you need or you need both row and column access. In that case, chunking into sub-matrix/hyper-cubes offers the best tradeoff.</p>
<p>We’ll demonstrate how chunking can affect performance by computing the column sums of column-chunked and row-chunked versions of the same matrix. The row-chunked data are much slower to process than the column-chunked data because it requires many more reads from disk.</p>
<div class="sourceCode" id="cb58"><html><body><pre class="r"><span class="co"># Simulate some data.</span>
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1e8</span>), <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">1e2</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">1e6</span>)

<span class="co"># Turn of verbose progress reporting.</span>
<span class="kw pkg">DelayedArray</span><span class="kw ns">:::</span><span class="fu">set_verbose_block_processing</span>(<span class="fl">FALSE</span>)
<span class="co">#&gt; [1] TRUE</span>

<span class="co"># Create column-chunked and row-chunked HDF5Matrix objects.</span>
<span class="no">x_col</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array</a></span>(<span class="no">x</span>, <span class="kw">chunkdim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">x</span>), <span class="fl">1</span>))
<span class="no">x_row</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array</a></span>(<span class="no">x</span>, <span class="kw">chunkdim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">x</span>)))

<span class="co"># Time computing the column sums.</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">x_col</span>))
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.101   0.069   1.169</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">x_row</span>))
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   7.632   0.953   8.586</span></pre></body></html></div>
</div>
<div id="chunk-compression" class="section level4">
<h4 class="hasAnchor">
<a href="#chunk-compression" class="anchor"></a>Chunk compression</h4>
<p>With the <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> package, the chunks can be compressed before writing to disk. Greater compression will lead to smaller files on disk but will generally require longer to write the files and may require longer to process the files. The level of compression is controlled by the by the <code>level</code> argument when writing data to disk using the <code><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array()</a></code> or <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code> functions.</p>
<p>We’ll demonstrate how compression can affect performance by writing uncompressed and maximally-compressed versions of the same matrix to disk and then computing the column sums. The uncompressed data are much faster to write but it takes roughly the same time to compute the column sums on both the uncompressed and maximally compressed data. These simulated data are not very compressible<a href="#fn20" class="footnote-ref" id="fnref20"><sup>20</sup></a>, but the on-disk savings will be more dramatic for other arrays (e.g., try repeating this with the <code>tenx_subset</code> data).</p>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="co"># Simulate some data.</span>
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1e8</span>), <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">1e2</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">1e6</span>)

<span class="co"># Time creating uncompressed and maximally-compressed HDF5Matrix objects.</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">x_no_compression</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array</a></span>(<span class="no">x</span>, <span class="kw">level</span> <span class="kw">=</span> <span class="fl">0</span>))
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.859   0.310   1.170</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">x_max_compression</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/writeHDF5Array.html">writeHDF5Array</a></span>(<span class="no">x</span>, <span class="kw">level</span> <span class="kw">=</span> <span class="fl">9</span>))
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;  45.450   0.335  47.254</span>

<span class="co"># Size of uncompressed and maximally-compressed HDF5 files on disk.</span>
<span class="fu"><a href="https://rdrr.io/r/base/file.info.html">file.size</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">path</a></span>(<span class="no">x_no_compression</span>))
<span class="co">#&gt; [1] 400040928</span>
<span class="fu"><a href="https://rdrr.io/r/base/file.info.html">file.size</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">path</a></span>(<span class="no">x_max_compression</span>))
<span class="co">#&gt; [1] 337061666</span>

<span class="co"># Time computing the column sums.</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">x_no_compression</span>))
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.777   0.068   0.845</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">x_max_compression</span>))
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.219   0.068   1.287</span></pre></body></html></div>
</div>
</div>
<div id="interaction-of-block-geometry-and-chunk-geometry" class="section level3">
<h3 class="hasAnchor">
<a href="#interaction-of-block-geometry-and-chunk-geometry" class="anchor"></a>Interaction of block geometry and chunk geometry</h3>
<p>Within the <em>DelayedArray</em> framework, the <a href="#block-geometry">Block geometry</a> and <a href="#chunk-geometry">Chunk geometry</a> are similar but distinct concepts. The difference is this:</p>
<blockquote>
<p><strong>The block geometry dictates how the data are accessed, the chunk geometry dictates how the data are stored</strong>.</p>
</blockquote>
<p>As noted in <a href="#chunk-geometry">Chunk geometry</a>, when you don’t know in advance what access patterns you need or you need both row and column access, then chunking into sub-matrix/hyper-cubes is often the best tradeoff.</p>
<p>As a user, increasing the <a href="#block-size">Block size</a> is often the easiest way to achieve faster block processing (at the cost of higher peak memory usage) but it may sometimes be beneficial to re-save your data with a chunk geometry that better matches the block geometry of the function(s) you are calling.</p>
</div>
<div id="avoid-subsetting-too-much-and-random-access-patterns" class="section level3">
<h3 class="hasAnchor">
<a href="#avoid-subsetting-too-much-and-random-access-patterns" class="anchor"></a>Avoid subsetting too much and random access patterns</h3>
<p>Reordering/subsetting the data may degrade the performance of even seemingly simple operations. This is especially true of disk-backed data, where performance is best when reading contiguous chunks of data and worst when having to read data with a random access pattern. This is demonstrated below by computing the column sums of the <code>tenx_subset</code> and a row-scrambled version thereof.</p>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="co"># Compute column sums of tenx_subset.</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">tenx_subset</span>))
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   0.332   0.008   0.340</span>

<span class="co"># Compute column sums of a row-scrambled version of tenx_subset.</span>
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="no">tenx_subset</span>[<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">tenx_subset</span>)), ]
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">y</span>))
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   1.573   0.004   1.577</span></pre></body></html></div>
</div>
<div id="process-save-repeat" class="section level3">
<h3 class="hasAnchor">
<a href="#process-save-repeat" class="anchor"></a>Process, save, repeat</h3>
<p>When analysing large datasets, a workflow that is broken up into stages and saves the intermediate outputs can be help preserve one’s sanity. This is true regardless of whether the DelayedArray framework is used - it sucks having to repeat some long pre-processing computation in order to make a quick plot - but it is especially true for DelayedArray-backed analyses where the accumulation of delayed operations may eventually lead to degraded performance.</p>
<p>This means using <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code>/<code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">quickResaveHDF5SummarizedExperiment()</a></code> and <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">loadHDF5SummarizedExperiment()</a></code> following time consuming processing of the object.</p>
</div>
<div id="pragmatism-rules" class="section level3">
<h3 class="hasAnchor">
<a href="#pragmatism-rules" class="anchor"></a>Pragmatism rules</h3>
<p>The DelayedArray framework coupled with disk-based data can be a powerful way to keep memory usage down, but sometimes you need to apply an algorithm that is too slow or simply doesn’t work except on ordinary in-memory arrays. Pragmatism is required: find a machine with a lot of RAM, load the data into memory, compute the thing you need, and move on with your life.</p>
<p>For example, with whole-genome bisulfite sequencing (WGBS) data there are 3 very large matrices stored in the <em>SummarizedExperiment</em> object. But to make a plot of methylation values along a gene promoter, a common requirement, you only need to load in a small ‘slice’ of one of these matrices. With a HDF5-backed <em>SummarizedExperiment</em> object you can quickly do this.</p>
<p>This brings us to a perhaps under-appreciated advantage of using HDF5-backed <em>SummarizedExperiment</em> objects: loading the saved data with <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">HDF5Array::loadHDF5SummarizedExperiment()</a></code> is often <strong>much</strong> faster than loading the in-memory equivalent <em>SummarizedExperiment</em> with <code><a href="https://rdrr.io/r/base/readRDS.html">readRDS()</a></code>. I made extensive use of this when processing large WGBS datasets as I could quickly load the <em>SummarizedExperiment</em> object to compute summaries of the sample metadata (stored in the <code>colData</code> of the <em>SummarizedExperiment</em> object), a process that used to take tens of minutes to hours because the 3 large matrices also had to be loaded into memory. This has been so useful to me that I now keep even ‘small’ WGBS datasets as HDF5-backed <em>SummarizedExperiment</em> objects.</p>
</div>
<div id="avoid-degrading-to-a-delayedarray" class="section level3">
<h3 class="hasAnchor">
<a href="#avoid-degrading-to-a-delayedarray" class="anchor"></a>Avoid ‘degrading’ to a <em>DelayedArray</em>
</h3>
<p>The DelayedArray framework is implemented using the S4 object oriented system. This can be used to write methods that are optimized for a particular backend. For example, we might write a <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMaxs()</a></code> method that is optimized for the <em>TENxMatrix</em> class by exploiting the sparse storage mode of the underlying data. In order for our hypothetical <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMaxs()</a></code> to ‘know’ that it can use this optimized method, however, it would need the data to be supplied as a <em>TENxMatrix</em> instance.</p>
<p>Unfortunately, it is very easy to ‘degrade’ a specialised <em>DelayedArray</em> derivative to a <em>DelayedArray</em>, as we will demonstrate. Let’s start with a <em>TENxMatrix</em> object:</p>
<div class="sourceCode" id="cb61"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_subset_sparse_hdf5</span>)
<span class="co">#&gt; [1] "TENxMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span></pre></body></html></div>
<p>Some common operations degrade the result to a <em>DelayedMatrix</em>, for example:</p>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="co"># Subsetting.</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_subset_sparse_hdf5</span>[, <span class="fl">1</span>:<span class="fl">10</span>])
<span class="co">#&gt; [1] "DelayedMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "DelayedArray"</span>

<span class="co"># Setting dimnames</span>
<span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx_subset_sparse_hdf5</span>
<span class="fu"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>(<span class="no">val</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"R"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">val</span>))), <span class="kw">NULL</span>)
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">val</span>)
<span class="co">#&gt; [1] "DelayedMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "DelayedArray"</span></pre></body></html></div>
<p>A common scenario where this degrading may occur is when extracting the data from a <em>SummarizedExperiment</em>.</p>
<div class="sourceCode" id="cb63"><html><body><pre class="r"><span class="co"># Construct a SummarizedExperiment with a TENxMatrix as an assay.</span>
<span class="no">se</span> <span class="kw">&lt;-</span> <span class="fu">SummarizedExperiment</span>(
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">tenx_subset_sparse_hdf5</span>))
<span class="co"># Add some column names.</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">se</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"S"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">se</span>)))

<span class="co"># Check the class of the assay.</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="fu">assay</span>(<span class="no">se</span>))
<span class="co">#&gt; [1] "DelayedMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "DelayedArray"</span></pre></body></html></div>
<p>What’s happened here? By default, <code>assay(se)</code> calls <code>assay(se, withDimnames = TRUE)</code> which has the effect of copying the dimnames from the <em>SummarizedExperiment</em> and adding them to the returned assay data. As we saw above, setting the dimnames on a <em>TENxMatrix</em> (or other <em>DelayedMatrix</em> derivative) will degrade it to a <em>DelayedMatrix</em>. Consequently, running <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMaxs(assay(se))</a></code> will not call our (hypothetical) optimized method for <em>TENxMatrix</em> objects and will instead defer to the slower, more general block processing method that is implemented for <em>DelayedMatrix</em> objects.</p>
<p>To avoid this ‘degrading upon assay extraction’, we can should set <code>withDimnames = FALSE</code>.</p>
<div class="sourceCode" id="cb64"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="fu">assay</span>(<span class="no">se</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>))
<span class="co">#&gt; [1] "TENxMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span></pre></body></html></div>
<p>More generally, you may need to avoid degrading a <em>DelayedArray</em> derivative to a <em>DelayedArray</em> in order to use backend-optimized methods.</p>
</div>
<div id="make-use-of-sparsity" class="section level3">
<h3 class="hasAnchor">
<a href="#make-use-of-sparsity" class="anchor"></a>Make use of sparsity</h3>
<p>Some recent additions to the <em>DelayedArray</em> framework allow it to better preserve the sparsity of the data. For example, we can preserve the sparsity of <span class="math inline">\(log(x + 1)\)</span> when applied to the sparse <code>tenx_subset_sparse_hdf5</code> object.</p>
<div class="sourceCode" id="cb65"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/read_block.html">is_sparse</a></span>(<span class="no">tenx_subset_sparse_hdf5</span>) <span class="co"># TRUE</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/read_block.html">is_sparse</a></span>(<span class="no">tenx_subset_sparse_hdf5</span> + <span class="fl">1</span>) <span class="co"># FALSE</span>
<span class="co">#&gt; [1] FALSE</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/read_block.html">is_sparse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="no">tenx_subset_sparse_hdf5</span> + <span class="fl">1</span>)) <span class="co"># TRUE</span>
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
</div>
<div id="parallelization-strategies" class="section level3">
<h3 class="hasAnchor">
<a href="#parallelization-strategies" class="anchor"></a>Parallelization strategies</h3>
<p>Several of the routines we used in today’s workshop have ‘native’ parallelization via the <em><a href="https://bioconductor.org/packages/3.12/BiocParallel">BiocParallel</a></em> package. Look out for the <code>BPPARAM</code> argument in function documentation<a href="#fn21" class="footnote-ref" id="fnref21"><sup>21</sup></a>.</p>
<p>Fair warning: parallelization is tricky and performance may not match expectations.</p>
</div>
<div id="whats-in-my-hdf5-file" class="section level3">
<h3 class="hasAnchor">
<a href="#whats-in-my-hdf5-file" class="anchor"></a>What’s in my HDF5 file?</h3>
<p>I do this quite a bit: I’ve got an HDF5 file but I don’t know (or have forgotten) what’s in it. <code><a href="https://rdrr.io/pkg/rhdf5/man/h5ls.html">rhdf5::h5ls()</a></code> to the rescue.</p>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="kw pkg">rhdf5</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5ls.html">h5ls</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">path</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">sce</span>)))
<span class="co">#&gt;   group   name       otype  dclass             dim</span>
<span class="co">#&gt; 0     / counts H5I_DATASET INTEGER 27998 x 1306127</span></pre></body></html></div>
</div>
</div>
<div id="important-topics-i-didnt-discuss" class="section level2">
<h2 class="hasAnchor">
<a href="#important-topics-i-didnt-discuss" class="anchor"></a>Important topics I didn’t discuss</h2>
<p>This 1-hour workshop can not cover everything about the DelayedArray framework. Some important topics that I omitted and some brief references to further resources are:</p>
<ul>
<li>General construction of a disk-backed <em>DelayedArray</em> (e.g., <em>HDF5Array</em>) from a bunch of files (e.g., CSV files, BAM files, BED files, BigWig files, etc.)
<ul>
<li>This requires the construction of a <code><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">RealizationSink()</a></code>.</li>
<li>
<code>bsseq::read.bismark()</code> and <code>minfi:::read.metharray2()</code> are (early) examples of constructing a HDF5-backed <em>SummarizedExperiment</em> from a bunch of files.</li>
</ul>
</li>
<li>Parallelization with disk-backed data.
<ul>
<li>Parallelization performance depends heavily on (A) the choice of backend, and (B) your computer’s hardware.</li>
<li>Rule of thumb:
<ul>
<li>It’s never as straightforward or as big an improvement as you think/hope.
<ul>
<li>Parallel <em>writing</em> from files (e.g. HDF5 files) is a no go.</li>
<li>Parallel <em>reading</em> is sometimes, maybe, perhaps okay …</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li>Choice of disk-backed backends.
<ul>
<li>
<em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> is the <em>de facto</em> standard in Bioconductor (and HDF5 files are the <em>de facto</em> standard for disk-backed array-like datasets in scientific computing).</li>
<li>
<em><a href="https://github.com/TileDB-Inc/TileDB-R">TileDB-R</a></em> and <em><a href="https://github.com/LTLA/TileDBArray">TileDBArray</a></em> look very interesting but are not available from CRAN/Bioconductor at the time of writing.</li>
<li>The <em><a href="https://bioconductor.org/packages/3.12/matter">matter</a></em> package could be the basis of a (hypothetical) <em>matterArray</em> backend.</li>
</ul>
</li>
</ul>
</div>
<div id="concluding-remarks" class="section level2">
<h2 class="hasAnchor">
<a href="#concluding-remarks" class="anchor"></a>Concluding remarks</h2>
<p>We’ve now spent nearly 1 hour learning about the DelayedArray framework. Now comes the kicker:</p>
<blockquote>
<p><strong>Don’t use a <em>DelayedArray</em> if you can don’t need to!</strong></p>
</blockquote>
<p>I write that as someone who has a professional love for the DelayedArray framework. I’ve spent many, many hours developing software that supports and extends it, leveraging it to analyse large WGBS and scRNA-seq datasets, and writing and presenting talks and workshops to teach it. But the simple fact remains:</p>
<blockquote>
<p><strong>If you can load your data into memory and still compute on it, then you’re always going to have a better time doing it that way.</strong></p>
</blockquote>
<p>Analyses will be faster, simpler, and you will have more software options available to you. But when this isn’t an option then the DelayedArray framework is a powerful set of packages to help you get your work done. For example, I find it pretty remarkable that a first-class single-cell analysis workflow can so seamlessly support the use of both in-memory and disk-backed data.</p>
<p>A huge amount of gratitude is owed to Hervé Pagès, the developer of the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> and <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> packages and the broader framework; Aaron Lun, the main developer many of the key packages that build upon the DelayedArray framework (particularly for analysing scRNA-seq data); and Mike Smith, the maintainer of the <em><a href="https://bioconductor.org/packages/3.12/rhdf5">rhdf5</a></em> package. Thank you!</p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<p>This document was prepared using the following software:</p>
<div class="sourceCode" id="cb67"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.0 (2020-04-24)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.4 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] scuttle_0.99.11             DelayedMatrixStats_1.11.1  </span>
<span class="co">#&gt;  [3] TENxBrainData_1.9.0         SingleCellExperiment_1.11.6</span>
<span class="co">#&gt;  [5] SummarizedExperiment_1.19.6 Biobase_2.49.0             </span>
<span class="co">#&gt;  [7] GenomicRanges_1.41.5        GenomeInfoDb_1.25.8        </span>
<span class="co">#&gt;  [9] ExperimentHub_1.15.0        AnnotationHub_2.21.1       </span>
<span class="co">#&gt; [11] BiocFileCache_1.13.0        dbplyr_1.4.4               </span>
<span class="co">#&gt; [13] HDF5Array_1.17.3            rhdf5_2.33.7               </span>
<span class="co">#&gt; [15] DelayedArray_0.15.7         IRanges_2.23.10            </span>
<span class="co">#&gt; [17] S4Vectors_0.27.12           BiocGenerics_0.35.4        </span>
<span class="co">#&gt; [19] matrixStats_0.56.0          Matrix_1.2-18              </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-6                  fs_1.4.2                     </span>
<span class="co">#&gt;  [3] bit64_0.9-7.1                 BiocPkgTools_1.7.0           </span>
<span class="co">#&gt;  [5] httr_1.4.2                    rprojroot_1.3-2              </span>
<span class="co">#&gt;  [7] gh_1.1.0                      tools_4.0.0                  </span>
<span class="co">#&gt;  [9] backports_1.1.8               R6_2.4.1                     </span>
<span class="co">#&gt; [11] DT_0.14                       DBI_1.1.0                    </span>
<span class="co">#&gt; [13] rhdf5filters_1.1.2            tidyselect_1.1.0             </span>
<span class="co">#&gt; [15] bit_1.1-15.2                  curl_4.3                     </span>
<span class="co">#&gt; [17] compiler_4.0.0                graph_1.67.1                 </span>
<span class="co">#&gt; [19] cli_2.0.2                     rvest_0.3.6                  </span>
<span class="co">#&gt; [21] xml2_1.3.2                    desc_1.2.0                   </span>
<span class="co">#&gt; [23] readr_1.3.1                   RBGL_1.65.0                  </span>
<span class="co">#&gt; [25] rappdirs_0.3.1                pkgdown_1.5.1                </span>
<span class="co">#&gt; [27] stringr_1.4.0                 digest_0.6.25                </span>
<span class="co">#&gt; [29] rmarkdown_2.3                 XVector_0.29.3               </span>
<span class="co">#&gt; [31] pkgconfig_2.0.3               htmltools_0.5.0              </span>
<span class="co">#&gt; [33] fastmap_1.0.1                 htmlwidgets_1.5.1            </span>
<span class="co">#&gt; [35] rlang_0.4.7                   RSQLite_2.2.0                </span>
<span class="co">#&gt; [37] shiny_1.5.0                   generics_0.0.2               </span>
<span class="co">#&gt; [39] jsonlite_1.7.0                BiocParallel_1.23.2          </span>
<span class="co">#&gt; [41] crosstalk_1.1.0.1             dplyr_1.0.0                  </span>
<span class="co">#&gt; [43] RCurl_1.98-1.2                magrittr_1.5                 </span>
<span class="co">#&gt; [45] GenomeInfoDbData_1.2.3        Rcpp_1.0.5                   </span>
<span class="co">#&gt; [47] Rhdf5lib_1.11.3               fansi_0.4.1                  </span>
<span class="co">#&gt; [49] lifecycle_0.2.0               stringi_1.4.6                </span>
<span class="co">#&gt; [51] yaml_2.2.1                    zlibbioc_1.35.0              </span>
<span class="co">#&gt; [53] MASS_7.3-51.5                 biocViews_1.57.3             </span>
<span class="co">#&gt; [55] grid_4.0.0                    blob_1.2.1                   </span>
<span class="co">#&gt; [57] promises_1.1.1                crayon_1.3.4                 </span>
<span class="co">#&gt; [59] lattice_0.20-41               hms_0.5.3                    </span>
<span class="co">#&gt; [61] knitr_1.29                    pillar_1.4.6                 </span>
<span class="co">#&gt; [63] igraph_1.2.5                  RUnit_0.4.32                 </span>
<span class="co">#&gt; [65] XML_3.99-0.5                  glue_1.4.1                   </span>
<span class="co">#&gt; [67] BiocVersion_3.12.0            evaluate_0.14                </span>
<span class="co">#&gt; [69] rex_1.2.0                     BiocManager_1.30.10          </span>
<span class="co">#&gt; [71] vctrs_0.3.2                   httpuv_1.5.4                 </span>
<span class="co">#&gt; [73] purrr_0.3.4                   tidyr_1.1.0                  </span>
<span class="co">#&gt; [75] assertthat_0.2.1              xfun_0.16                    </span>
<span class="co">#&gt; [77] mime_0.9                      xtable_1.8-4                 </span>
<span class="co">#&gt; [79] later_1.1.0.1                 tibble_3.0.3                 </span>
<span class="co">#&gt; [81] AnnotationDbi_1.51.3          memoise_1.1.0                </span>
<span class="co">#&gt; [83] ellipsis_0.3.1                interactiveDisplayBase_1.27.5</span>
<span class="co">#&gt; [85] BiocStyle_2.17.0</span></pre></body></html></div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-rizzardi2019neuronal">
<p>Rizzardi, Lindsay F, Peter F Hickey, Varenka Rodriguez DiBlasi, Rakel Tryggvadóttir, Colin M Callahan, Adrian Idrizi, Kasper D Hansen, and Andrew P Feinberg. 2019. “Neuronal Brain-Region-Specific Dna Methylation and Chromatin Accessibility Are Associated with Neuropsychiatric Trait Heritability.” <em>Nature Neuroscience</em> 22 (2): 307.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Higher-dimensional arrays may be appropriate for some types of assays.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>In R, a <em>matrix</em> is just a 2-dimensional <em>array</em><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>In fact, if you use any package that makes use of the <em>SummarizedExperiment</em> class, then you will almost certainly load the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package during the course of your analysis, whether you know it or not! This is because <em><a href="https://bioconductor.org/packages/3.12/SummarizedExperiment">SummarizedExperiment</a></em> depends upon <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em>.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>As listed in the <code>Depends</code> field of the package <code>DESCRIPTION</code> file.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>This dataset is also available in the <em><a href="https://bioconductor.org/packages/3.12/TENxBrainData">TENxBrainData</a></em> Bioconductor package.<a href="#fnref5" class="footnote-back">↩︎</a></p></li>
<li id="fn6"><p>You may have seen similar pretty printing with other Bioconductor objects such as <em>GRanges</em> and <em>DataFrame</em> or with the <em>data.table</em> and <em>tibble</em> extensions to the <em>data.frame</em>. I can’t say enough how much I appreciate these thoughtful touches when doing interactive data analysis.<a href="#fnref6" class="footnote-back">↩︎</a></p></li>
<li id="fn7"><p>As with a 2-dimensional <em>array</em> in base R being commonly known as a <em>matrix</em>, a 2-dimensional <em>DelayedArray</em> is also known as a <em>DelayedMatrix</em> and a 2-dimensional <em>HDF5Array</em> is also known as a <em>HDF5Matrix</em>.<a href="#fnref7" class="footnote-back">↩︎</a></p></li>
<li id="fn8"><p>Most of the object’s size is due to the dimnames being stored in memory.<a href="#fnref8" class="footnote-back">↩︎</a></p></li>
<li id="fn9"><p>The technical names of each type of delayed operation are not important.<a href="#fnref9" class="footnote-back">↩︎</a></p></li>
<li id="fn10"><p>The <a href="#no-op">No-op</a> example is the obvious exception<a href="#fnref10" class="footnote-back">↩︎</a></p></li>
<li id="fn11"><p>Specifically, <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">DelayedArray::colSums()</a></code>.<a href="#fnref11" class="footnote-back">↩︎</a></p></li>
<li id="fn12"><p>If you try these out on <code>tenx</code> you might be waiting a while, so I don’t recommend this in the workshop.<a href="#fnref12" class="footnote-back">↩︎</a></p></li>
<li id="fn13"><p>A sparse <em>DelayedArray</em> is one for which <code><a href="https://rdrr.io/pkg/DelayedArray/man/read_block.html">is_sparse()</a></code> returns <code>TRUE</code>.<a href="#fnref13" class="footnote-back">↩︎</a></p></li>
<li id="fn14"><p>In the above example it’s safe because <code>tenx_subset</code> only requires around 100 Mb of memory even as an ordinary array.<a href="#fnref14" class="footnote-back">↩︎</a></p></li>
<li id="fn15"><p>The <em>TENxMatrix</em> class follows an HDF5-based sparse matrix representation, instead of the conventional (i.e. dense) HDF5 representation, and is used by 10x Genomics.<a href="#fnref15" class="footnote-back">↩︎</a></p></li>
<li id="fn16"><p>We’ll discuss why you might want this later in the workshop.<a href="#fnref16" class="footnote-back">↩︎</a></p></li>
<li id="fn17"><p>Please keep in mind that <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment()</a></code> and <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">loadHDF5SummarizedExperiment()</a></code> don’t know how to produce/read tarballs or zip files at the moment, so the process of packaging/extracting the tarball or zip file is entirely the user responsibility. This is typically done from outside R.<a href="#fnref17" class="footnote-back">↩︎</a></p></li>
<li id="fn18"><p>Use <code>verbose=TRUE</code> to see its progress.<a href="#fnref18" class="footnote-back">↩︎</a></p></li>
<li id="fn19"><p>The reality is a little bit more nuanced, but discussing the full details is not important here, and would only distract us<a href="#fnref19" class="footnote-back">↩︎</a></p></li>
<li id="fn20"><p>E.g., it’s uniformly distributed with few zeros<a href="#fnref20" class="footnote-back">↩︎</a></p></li>
<li id="fn21"><p>Unfortunately, simply typing the function name at the command prompt or using <code><a href="https://rdrr.io/r/base/args.html">args()</a></code> may not be sufficient to identify these functions. E.g., <code><a href="https://rdrr.io/r/base/args.html">args(denoisePCA)</a></code> doesn’t include <code>BPPARAM</code> but <code>help("denoisePCA", "scran")</code> tells you this is an option (Footnote to the footnote: this is because <code>denoicePCA()</code> is an S4 generic with methods defined only for the first argument, <code>x</code>. The methods include the <code>BPPARAM</code> argument and are documented as such.)<a href="#fnref21" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Peter Hickey.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
