<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Effectively using the DelayedArray framework as a user to support the analysis of large datasets • DelayedArrayWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Effectively using the DelayedArray framework as a user to support the analysis of large datasets">
<meta property="og:description" content="DelayedArrayWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DelayedArrayWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Effectively_using_the_DelayedArray_framework_for_users.html">Effectively using the DelayedArray framework as a user to support the analysis of large datasets</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PeteHaitch/BioC2020_DelayedArray_workshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="Effectively_using_the_DelayedArray_framework_for_users_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="Effectively_using_the_DelayedArray_framework_for_users_files/pagedtable-1.1/js/pagedtable.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Effectively using the DelayedArray framework as a user to support the analysis of large datasets</h1>
                        <h4 class="author">Peter Hickey</h4>
            
            <h4 class="date">Last modified: June 5, 2020; Compiled: June 05, 2020</h4>
      
      
      <div class="hidden name"><code>Effectively_using_the_DelayedArray_framework_for_users.Rmd</code></div>

    </div>

    
    
<div id="effectively-using-the-delayedarray-framework-as-a-user-to-support-the-analysis-of-large-datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#effectively-using-the-delayedarray-framework-as-a-user-to-support-the-analysis-of-large-datasets" class="anchor"></a>Effectively using the DelayedArray framework as a user to support the analysis of large datasets</h1>
<p>BioC 2019: 24-27 June</p>
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop description</h2>
<p>This workshop gives an introductory overview of the DelayedArray framework, which can be used by Bioconductor packages to support the analysis of large datasets. A <em>DelayedArray</em> is like an ordinary array in R, but allows for the data to be in-memory<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, on-disk in a file, or even hosted on a remote server. Participants will learn where they might encounter a <em>DelayedArray</em> in the wild while using Bioconductor and help them understand the fundamental concepts underlying the framework. This workshop will be a mixture of lecture with example code and discussion. Examples will mostly be drawn from the analysis of single-cell RNA-sequencing data.</p>
<div id="instructor" class="section level3">
<h3 class="hasAnchor">
<a href="#instructor" class="anchor"></a>Instructor</h3>
<ul>
<li>Peter Hickey (<a href="mailto:hickey@wehi.edu.au" class="email">hickey@wehi.edu.au</a>)</li>
</ul>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li>Basic knowledge of R syntax.</li>
<li>Familiarity with common operations on matrices in R, such as <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code> and <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMeans()</a></code>.</li>
<li>Some familiarity with S4 objects may be helpful but is not required.</li>
<li>Some familiarity with single-cell RNA-sequencing may be helpful but is not required.</li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h3>
<p>Students will be able to run code interactively during the workshop. There will be opportunities throughout the workshop for questions and discussion.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TENxBrainData">TENxBrainData</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TENxPBMCData">TENxPBMCData</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/BiocSingular">BiocSingular</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/scater">scater</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/scran">scran</a></em></li>
</ul>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introductory material</td>
<td>10m</td>
</tr>
<tr class="even">
<td>First contact</td>
<td>40m</td>
</tr>
<tr class="odd">
<td>Package ecosystem</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Real world encounters analysing scRNA-seq data</td>
<td>25m</td>
</tr>
<tr class="odd">
<td>Workflow tips for DelayedArray-backed analyses</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
<div id="workshop-goals-and-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h3>
<div id="learning-goals" class="section level4">
<h4 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h4>
<ul>
<li>Learn of existing packages and functions that <em>use</em> the DelayedArray framework.</li>
<li>Develop a high-level understanding of classes and packages that <em>implement</em> the DelayedArray framework.</li>
<li>Become familiar with the fundamental concepts of delayed operations, block-processing, and realization.</li>
<li>Reason about potential bottlenecks in algorithms operating on <em>DelayedArray</em> objects.</li>
</ul>
</div>
<div id="learning-objectives" class="section level4">
<h4 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h4>
<ul>
<li>Identify when an object is a <em>DelayedArray</em> or one of its derivatives.</li>
<li>Be able to recognise when it is useful to use a <em>DelayedArray</em> instead of an ordinary array or other array-like data structure.</li>
<li>Learn how to load and save a DelayedArray-backed object.</li>
<li>Learn how the ‘block size’ and ‘chunking’ of the dataset affect performance when operating on <em>DelayedArray</em> objects.</li>
<li>Take away some miscellaneous tips and tricks I’ve learnt over the years when working with DelayedArray-backed objects.</li>
</ul>
</div>
</div>
</div>
<div id="introductory-material" class="section level2">
<h2 class="hasAnchor">
<a href="#introductory-material" class="anchor"></a>Introductory material</h2>
<p>Data from a high-throughput biological assay, such as single-cell RNA-sequencing (scRNA-seq), will often be summarised as a matrix of counts, where rows correspond to features and columns to samples<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. Within <strong>Bioconductor</strong>, the <em>SummarizedExperiment</em> class is the recommended container for such data, offering a rich interface that tightly links assay measurements to data on the features and the samples.</p>
<p>The <em>SummarizedExperiment</em> class is used to store rectangular arrays of experimental results (<em>assays</em>). Here, each <em>assay</em> is drawn as a matrix but higher-dimensional arrays are also supported.</p>
<p><img src="../../../../tmp/Rtmpuyl994/SE.svg"><!-- --></p>
<p>Traditionally, the assay data are stored in-memory as an ordinary <em>array</em> object<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. Storing the data in-memory becomes a real pain with the ever-growing size of ’omics datasets. It is now not uncommon to collect <span class="math inline">\(10,000-100,000,000\)</span> measurements on <span class="math inline">\(100 - 1,000,000\)</span> samples, which would occupy <span class="math inline">\(10-1,000\)</span> gigabytes (Gb) if stored in-memory as ordinary R arrays.</p>
<p>The DelayedArray framework offers a solution to this problem. Wrapping an array-like object (typically an on-disk object) in a <em>DelayedArray</em> object allows one to perform common array operations on it without loading the object in memory. In order to reduce memory usage and optimize performance, operations on the object are either delayed or executed using a block processing mechanism.</p>
<div id="projects-enabled-by-delayedarray" class="section level3">
<h3 class="hasAnchor">
<a href="#projects-enabled-by-delayedarray" class="anchor"></a>Projects enabled by DelayedArray</h3>
<p>The DelayedArray framework enables the analysis of datasets that are too large to be stored or processed in-memory. This has become particularly relevant with the advent of large single-cell RNA-sequencing (scRNA-seq) studies containing tens of thousands to millions of cells.</p>
<p>In my own research I have made extensive use of the DelayedArray framework when analysing whole genome bisulfite sequencing (WGBS) datasets. To give a recent example, we profiled 45 human brain samples using WGBS to measure DNA methylation <span class="citation">(Rizzardi et al. 2019)</span>. For most tissues, we focus on so-called so-called CpG methylation, which requires analysing matrices with roughly 20 million rows (CpG loci) and 45 columns (samples). This sized data is challenging, but well within the realms of high performance computing available at a modern research institute. However, the brain also has extensive non-CpG methylation and there are an order of magnitude more non-CpG loci. This necessitated extensive re-factoring of our software tools and we successfully adopted the DelayedArray framework to enable this research.</p>
<p>In this workshop we will see how the analysis of large scRNA-seq datasets is enabled by the DelayedArray framework.</p>
</div>
</div>
<div id="first-contact" class="section level2">
<h2 class="hasAnchor">
<a href="#first-contact" class="anchor"></a>First contact</h2>
<p>The heart of the DelayedArray framework is implemented in the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package, which we now load.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DelayedArray</span>)
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; Loading required package: matrixStats</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class="co">#&gt;     union, unique, unsplit, which.max, which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'DelayedArray'</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     aperm, apply, rowsum</span></pre></body></html></div>
<p>We will begin with an example using some scRNA-seq data on 1.3 million brain cells from embryonic mice, generated by 10X Genomics. These data are available in the <em><a href="https://bioconductor.org/packages/3.12/TENxBrainData">TENxBrainData</a></em> Bioconductor package.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># NOTE: The TENxBrainData package loads and attaches the HDF5Array package, </span>
<span class="co">#       amongst others.</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TENxBrainData</span>)
<span class="co">#&gt; Loading required package: SingleCellExperiment</span>
<span class="co">#&gt; Loading required package: SummarizedExperiment</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with</span>
<span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biobase'</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyMissing, rowMedians</span>
<span class="co">#&gt; Loading required package: HDF5Array</span>
<span class="co">#&gt; Loading required package: rhdf5</span>

<span class="co"># NOTE: This will download the data and may take a little while on the first </span>
<span class="co">#       run. The result will be cached, however, so subsequent runs avoid </span>
<span class="co">#       re-downloading the data.</span>
<span class="no">tenx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxBrainData/man/TENxBrainData.html">TENxBrainData</a></span>()
<span class="co">#&gt; using temporary cache /tmp/Rtmpuyl994/BiocFileCache</span>
<span class="co">#&gt; snapshotDate(): 2020-06-01</span>
<span class="co">#&gt; see ?TENxBrainData and browseVignettes('TENxBrainData') for documentation</span>
<span class="co">#&gt; downloading 1 resources</span>
<span class="co">#&gt; retrieving 1 resource</span>
<span class="co">#&gt; loading from cache</span></pre></body></html></div>
<p>Let’s take a look at the <code>tenx</code> object.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">tenx</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 27998 1306127 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames: NULL</span>
<span class="co">#&gt; rowData names(2): Ensembl Symbol</span>
<span class="co">#&gt; colnames(1306127): AAACCTGAGATAGGAG-1 AAACCTGAGCGGCTTC-1 ...</span>
<span class="co">#&gt;   TTTGTCAGTTAAAGTG-133 TTTGTCATCTGAAAGA-133</span>
<span class="co">#&gt; colData names(4): Barcode Sequence Library Mouse</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></pre></body></html></div>
<p>The data are stored in a <em>SingleCellExperiment</em> object, an extension of the <em>SummarizedExperiment</em> class. With data from 1.3 million cells, this is roughly 100,000-times more samples than a typical bulk RNA-seq dataset and would occupy 146 GB in-memory if stored as an ordinary <em>matrix</em>.</p>
<p>We might expect that interacting with an object containing this many samples would feel sluggish, but this is not the case. For example, we can efficiently subset the object.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co"># Take a random sample of 1000 genes and 200000 samples</span>
<span class="no">tenx</span>[<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">tenx</span>), <span class="fl">1000</span>), <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx</span>), <span class="fl">200000</span>)]
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 1000 200000 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames: NULL</span>
<span class="co">#&gt; rowData names(2): Ensembl Symbol</span>
<span class="co">#&gt; colnames(200000): GCACATACATTCCTCG-116 GCAGTTAAGACAATAC-63 ...</span>
<span class="co">#&gt;   TTTGTCACAGGGAGAG-23 GATCGATTCGCTAGCG-100</span>
<span class="co">#&gt; colData names(4): Barcode Sequence Library Mouse</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></pre></body></html></div>
<p>Now, this efficiency could all be down to some trickery through the outer <em>SingleCellExperiment</em> class. To make things clearer, we’ll extract the <code>counts</code> assay.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co"># NOTE: We'll discuss the use of `withDimnames = FALSE` later in the workshop.</span>
<span class="no">tenx_counts</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">tenx</span>, <span class="st">"counts"</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p>Now, let’s do something that would ordinarily be a terrible idea, and something that’s frustrated me way too many times: let’s “accidentally” print out the entire counts matrix.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">tenx_counts</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span></pre></body></html></div>
<p>Hallelujah! Unlike what you may have experienced when printing out a large matrix, this didn’t overwhelm the screen with thousands of lines of output nor did it cause the R session to hang indefinitely. In fact, this gives us a rather pretty printing of the counts matrix<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>. No need for panicked mashing of <code>Ctrl-c</code> or <code>Esc</code>.</p>
<p>We can now clearly see that <code>tenx_counts</code> is no ordinary <em>matrix</em>. In fact, it is an <em>HDF5Matrix</em>, which is a type of <em>DelayedArray</em><a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_counts</span>)
<span class="co">#&gt; [1] "HDF5Matrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span>
<span class="fu">is</span>(<span class="no">tenx_counts</span>, <span class="st">"DelayedArray"</span>)
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>The data contained in an <em>HDF5Matrix</em> is actually stored on disk in a <a href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">Hierarchical Data Format (<strong>HDF5</strong>)</a> file. Consequently, the <code>tenx_counts</code> object takes up very little space in memory.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span>(<span class="no">tenx_counts</span>), <span class="kw">units</span> <span class="kw">=</span> <span class="st">"auto"</span>)
<span class="co">#&gt; 2.3 Kb</span></pre></body></html></div>
<p>We can learn more about the internals of the <code>tenx_counts</code> object using the <code><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed()</a></code> function<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a>.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span>(<span class="no">tenx_counts</span>)
<span class="co">#&gt; An object of class "HDF5ArraySeed"</span>
<span class="co">#&gt; Slot "filepath":</span>
<span class="co">#&gt; [1] "/tmp/Rtmpuyl994/BiocFileCache/c5f37e4f4d6b_1040"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "name":</span>
<span class="co">#&gt; [1] "counts"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "type":</span>
<span class="co">#&gt; [1] NA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "dim":</span>
<span class="co">#&gt; [1]   27998 1306127</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "chunkdim":</span>
<span class="co">#&gt; [1] 100 100</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "first_val":</span>
<span class="co">#&gt; [1] 0</span></pre></body></html></div>
<div id="three-examples-of-computing-on-a-delayedarray" class="section level3">
<h3 class="hasAnchor">
<a href="#three-examples-of-computing-on-a-delayedarray" class="anchor"></a>Three examples of computing on a DelayedArray</h3>
<p>We will now play around with computing on the counts matrix. To make things slightly easier, we will first subset the data to 1000 samples.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">tenx_counts_subset</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts</span>[, <span class="fl">1</span>:<span class="fl">1000</span>]</pre></body></html></div>
<div id="library-sizes" class="section level4">
<h4 class="hasAnchor">
<a href="#library-sizes" class="anchor"></a>Library sizes</h4>
<p>Firstly, let’s compute the library sizes for this subset of samples. We can do this using <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code>.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">lib_sizes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">tenx_counts_subset</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">lib_sizes</span>)
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt;    1453    2755    3898    4689    5548   34233</span></pre></body></html></div>
</div>
<div id="proportion-of-cells-with-non-zero-expression-for-each-gene" class="section level4">
<h4 class="hasAnchor">
<a href="#proportion-of-cells-with-non-zero-expression-for-each-gene" class="anchor"></a>Proportion of cells with non-zero expression for each gene</h4>
<p>Secondly, suppose we want to know for each gene the proportion of cells with non-zero expression. We can do this using <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowSums()</a></code> in conjunction with some standard R commands (logical comparisons and division).</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">prop_non_zero</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowSums</a></span>(<span class="no">tenx_counts_subset</span> <span class="kw">&gt;</span> <span class="fl">0</span>) /  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">prop_non_zero</span>)
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt; 0.00000 0.00000 0.00300 0.07235 0.07400 1.00000</span></pre></body></html></div>
</div>
<div id="median-expression-of-each-gene" class="section level4">
<h4 class="hasAnchor">
<a href="#median-expression-of-each-gene" class="anchor"></a>Median expression of each gene</h4>
<p>Finally, suppose we want to know the median expression of each gene. Here, we will quantify expression as counts per million (CPM) using library size normalization.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">cpm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="fl">1e6</span> * <span class="no">tenx_counts_subset</span>) / <span class="no">lib_sizes</span>)
<span class="no">cpm</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "double":</span>
<span class="co">#&gt;              [,1]     [,2]     [,3] ...   [,999]  [,1000]</span>
<span class="co">#&gt;     [1,]        0        0        0   .        0        0</span>
<span class="co">#&gt;     [2,]        0        0        0   .        0        0</span>
<span class="co">#&gt;     [3,]        0        0        0   .        0        0</span>
<span class="co">#&gt;     [4,]        0        0        0   .        0        0</span>
<span class="co">#&gt;     [5,]        0        0        0   .        0        0</span>
<span class="co">#&gt;      ...        .        .        .   .        .        .</span>
<span class="co">#&gt; [27994,]   0.0000   0.0000   0.0000   .   0.0000   0.0000</span>
<span class="co">#&gt; [27995,] 247.1577   0.0000   0.0000   . 277.1619   0.0000</span>
<span class="co">#&gt; [27996,]   0.0000   0.0000   0.0000   .   0.0000   0.0000</span>
<span class="co">#&gt; [27997,]   0.0000   0.0000   0.0000   .   0.0000   0.0000</span>
<span class="co">#&gt; [27998,]   0.0000   0.0000   0.0000   .   0.0000   0.0000</span></pre></body></html></div>
<p>We can then compute the median expression of each gene using <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colMedians.html">DelayedMatrixStats::rowMedians()</a></code>.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DelayedMatrixStats</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'DelayedMatrixStats'</span>
<span class="co">#&gt; The following object is masked from 'package:Biobase':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     rowMedians</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colAlls, colAnyMissings, colAnyNAs, colAnys, colAvgsPerRowSet,</span>
<span class="co">#&gt;     colCollapse, colCounts, colCummaxs, colCummins, colCumprods,</span>
<span class="co">#&gt;     colCumsums, colDiffs, colIQRDiffs, colIQRs, colLogSumExps,</span>
<span class="co">#&gt;     colMadDiffs, colMads, colMeans2, colMedians, colOrderStats,</span>
<span class="co">#&gt;     colProds, colQuantiles, colRanks, colSdDiffs, colSds, colSums2,</span>
<span class="co">#&gt;     colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyMissings, rowAnyNAs, rowAnys,</span>
<span class="co">#&gt;     rowAvgsPerColSet, rowCollapse, rowCounts, rowCummaxs, rowCummins,</span>
<span class="co">#&gt;     rowCumprods, rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs,</span>
<span class="co">#&gt;     rowLogSumExps, rowMadDiffs, rowMads, rowMeans2, rowMedians,</span>
<span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanks, rowSdDiffs,</span>
<span class="co">#&gt;     rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span>
<span class="no">median_expression</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colMedians.html">rowMedians</a></span>(<span class="no">cpm</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">median_expression</span>)
<span class="co">#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span>
<span class="co">#&gt;     0.00     0.00     0.00    18.73     0.00 22069.23</span></pre></body></html></div>
</div>
<div id="summary" class="section level4">
<h4 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h4>
<p>These 3 examples highlight the power of the DelayedArray framework. Recall that the data in these examples live on disk in an HDF5 file, yet we interacted with <code>tenx_counts_subset</code> and computed on it much as we would if the data were in-memory as an ordinary matrix. Also note that all 3 examples returned ordinary R vectors.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">lib_sizes</span>)
<span class="co">#&gt; [1] "numeric"</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">prop_non_zero</span>)
<span class="co">#&gt; [1] "numeric"</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">cpm</span>) <span class="co"># TODO: Not an ordinary vector.</span>
<span class="co">#&gt; [1] "DelayedMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "DelayedArray"</span></pre></body></html></div>
<p>To do so, we made (implicit) use of the three fundamental concepts of the DelayedArray framework:</p>
<ol style="list-style-type: decimal">
<li>Delayed operations</li>
<li>Block-processing</li>
<li>Realization</li>
</ol>
<p>We’ll now discuss each of these in turn.</p>
</div>
</div>
<div id="delayed-operations" class="section level3">
<h3 class="hasAnchor">
<a href="#delayed-operations" class="anchor"></a>Delayed operations</h3>
<p>Taking a careful look at <code>tenx_counts_subset</code>, we see that it is a <em>DelayedMatrix</em> rather than an <em>HDF5Matrix</em>.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">tenx_counts_subset</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span></pre></body></html></div>
<p>The subsetting operation has ‘degraded’ the <code>tenx_counts_subset</code> object to a <em>DelayedMatrix</em>.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu">is</span>(<span class="no">tenx_counts_subset</span>, <span class="st">"HDF5Matrix"</span>)
<span class="co">#&gt; [1] FALSE</span>
<span class="fu">is</span>(<span class="no">tenx_counts_subset</span>, <span class="st">"DelayedMatrix"</span>)
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>The <code><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree()</a></code> function can help us see what changed when we subsetted the data.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx_counts</span>)
<span class="co">#&gt; 27998x1306127 integer: HDF5Matrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx_counts_subset</span>)
<span class="co">#&gt; 27998x1000 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
<p>The subsetting operation has been registered in what is termed a ‘delayed operation’. Registering a delayed operation does not modify the underlying data. Instead, the operation is recorded and only performed when the <em>DelayedArray</em> object is ‘realized’. Realization of a <em>DelayedArray</em> triggers the execution of the delayed operations carried by the object and returns the result as an ordinary <em>array</em>.</p>
<p>This allows us to chain together multiple operations and only perform them as required. Here is a contrived example.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="co"># Add 1 to every element (a delayed op).</span>
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts_subset</span> + <span class="fl">1L</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">x</span>)
<span class="co">#&gt; 27998x1000 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;       └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>

<span class="co"># Compute log of every element (another delayed op).</span>
<span class="no">lx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="no">x</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">lx</span>)
<span class="co">#&gt; 27998x1000 double: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 double: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;       └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>

<span class="co"># Transpose the result (another delayed op).</span>
<span class="no">tlx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="no">lx</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tlx</span>)
<span class="co">#&gt; 1000x27998 double: DelayedMatrix object</span>
<span class="co">#&gt; └─ 1000x27998 double: Unary iso op stack</span>
<span class="co">#&gt;    └─ 1000x27998 integer: Aperm (perm=c(2,1))</span>
<span class="co">#&gt;       └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;          └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>

<span class="co"># Realize a subset of the data as an ordinary matrix.</span>
<span class="fu"><a href="https://rdrr.io/r/base/array.html">as.array</a></span>(<span class="no">tlx</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">10</span>])
<span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7]     [,8]      [,9] [,10]</span>
<span class="co">#&gt; [1,]    0    0    0    0    0    0    0 0.000000 0.0000000     0</span>
<span class="co">#&gt; [2,]    0    0    0    0    0    0    0 0.000000 0.0000000     0</span>
<span class="co">#&gt; [3,]    0    0    0    0    0    0    0 0.000000 0.6931472     0</span>
<span class="co">#&gt; [4,]    0    0    0    0    0    0    0 0.000000 0.0000000     0</span>
<span class="co">#&gt; [5,]    0    0    0    0    0    0    0 1.098612 0.6931472     0</span></pre></body></html></div>
<p>Many common operations can be registered as delayed operations. Here are some examples<a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a>. Notice that in each case the result is ‘degraded’ to a <em>DelayedMatrix</em><a href="#fn8" class="footnote-ref" id="fnref8"><sup>8</sup></a>.</p>
<div id="delayedsubset" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedsubset" class="anchor"></a>DelayedSubset</h4>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts</span>[, <span class="fl">1</span>:<span class="fl">100</span>]
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 100&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;            [,1]   [,2]   [,3]   [,4] ...  [,97]  [,98]  [,99] [,100]</span>
<span class="co">#&gt;     [1,]      0      0      0      0   .      0      0      1      0</span>
<span class="co">#&gt;     [2,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt;     [3,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt;     [4,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt;     [5,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt;      ...      .      .      .      .   .      .      .      .      .</span>
<span class="co">#&gt; [27994,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt; [27995,]      1      0      0      2   .      0      1      2      0</span>
<span class="co">#&gt; [27996,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt; [27997,]      0      0      0      0   .      0      0      0      0</span>
<span class="co">#&gt; [27998,]      0      0      0      0   .      0      0      0      0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x100 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x100 integer: Subset</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayedaperm" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedaperm" class="anchor"></a>DelayedAperm</h4>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="no">tenx_counts</span>)
<span class="no">val</span>
<span class="co">#&gt; &lt;1306127 x 27998&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]     [,2]     [,3]     [,4] ... [,27995] [,27996] [,27997]</span>
<span class="co">#&gt;       [1,]        0        0        0        0   .        1        0        0</span>
<span class="co">#&gt;       [2,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt;       [3,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt;       [4,]        0        0        0        0   .        2        0        0</span>
<span class="co">#&gt;       [5,]        0        0        0        0   .        2        0        0</span>
<span class="co">#&gt;        ...        .        .        .        .   .        .        .        .</span>
<span class="co">#&gt; [1306123,]        0        0        0        0   .        0        2        0</span>
<span class="co">#&gt; [1306124,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt; [1306125,]        0        0        0        0   .        1        0        0</span>
<span class="co">#&gt; [1306126,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt; [1306127,]        0        0        0        0   .        0        0        0</span>
<span class="co">#&gt;            [,27998]</span>
<span class="co">#&gt;       [1,]        0</span>
<span class="co">#&gt;       [2,]        0</span>
<span class="co">#&gt;       [3,]        0</span>
<span class="co">#&gt;       [4,]        0</span>
<span class="co">#&gt;       [5,]        0</span>
<span class="co">#&gt;        ...        .</span>
<span class="co">#&gt; [1306123,]        0</span>
<span class="co">#&gt; [1306124,]        0</span>
<span class="co">#&gt; [1306125,]        0</span>
<span class="co">#&gt; [1306126,]        0</span>
<span class="co">#&gt; [1306127,]        0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 1306127x27998 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 1306127x27998 integer: Aperm (perm=c(2,1))</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayedunaryisoop" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedunaryisoop" class="anchor"></a>DelayedUnaryIsoOp</h4>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts</span> + <span class="fl">1L</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [2,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [3,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [4,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [5,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt; [27995,]          2          1          1          3   .          1          2</span>
<span class="co">#&gt; [27996,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt; [27997,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt; [27998,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          1          1</span>
<span class="co">#&gt;     [2,]          1          1</span>
<span class="co">#&gt;     [3,]          1          1</span>
<span class="co">#&gt;     [4,]          1          1</span>
<span class="co">#&gt;     [5,]          1          1</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          1          1</span>
<span class="co">#&gt; [27995,]          1          1</span>
<span class="co">#&gt; [27996,]          1          1</span>
<span class="co">#&gt; [27997,]          1          1</span>
<span class="co">#&gt; [27998,]          1          1</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts</span> + <span class="fl">1</span>:<span class="fl">2</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [2,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt;     [3,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;     [4,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt;     [5,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt; [27995,]          2          1          1          3   .          1          2</span>
<span class="co">#&gt; [27996,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt; [27997,]          1          1          1          1   .          1          1</span>
<span class="co">#&gt; [27998,]          2          2          2          2   .          2          2</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          1          1</span>
<span class="co">#&gt;     [2,]          2          2</span>
<span class="co">#&gt;     [3,]          1          1</span>
<span class="co">#&gt;     [4,]          2          2</span>
<span class="co">#&gt;     [5,]          1          1</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          2          2</span>
<span class="co">#&gt; [27995,]          1          1</span>
<span class="co">#&gt; [27996,]          2          2</span>
<span class="co">#&gt; [27997,]          1          1</span>
<span class="co">#&gt; [27998,]          2          2</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Unary iso op with args</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayedsubassign" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedsubassign" class="anchor"></a>DelayedSubassign</h4>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="co"># NOTE: Be careful with delayed subassignment. You can end up with objects that </span>
<span class="co">#       are surprisingly large in-memory.</span>
<span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts</span>
<span class="no">tmp</span>[, <span class="fl">1</span>] <span class="kw">&lt;-</span> <span class="fl">100</span>
<span class="no">tmp</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "double":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3] ... [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]        100          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]        100          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]        100          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]        100          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]        100          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]        100          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]        100          0          0   .          0          0</span>
<span class="co">#&gt; [27996,]        100          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]        100          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]        100          0          0   .          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tmp</span>)
<span class="co">#&gt; 27998x1306127 double: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 double: Subassign</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayeddimnames" class="section level4">
<h4 class="hasAnchor">
<a href="#delayeddimnames" class="anchor"></a>DelayedDimnames</h4>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">tmp</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"R"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">tmp</span>)))
<span class="no">tmp</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;              [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     R1          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     R2          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     R3          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     R4          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     R5          0          0          0          0   .          0          0</span>
<span class="co">#&gt;    ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; R27994          0          0          0          0   .          0          0</span>
<span class="co">#&gt; R27995          1          0          0          2   .          0          1</span>
<span class="co">#&gt; R27996          0          0          0          0   .          0          0</span>
<span class="co">#&gt; R27997          0          0          0          0   .          0          0</span>
<span class="co">#&gt; R27998          0          0          0          0   .          0          0</span>
<span class="co">#&gt;        [,1306126] [,1306127]</span>
<span class="co">#&gt;     R1          0          0</span>
<span class="co">#&gt;     R2          0          0</span>
<span class="co">#&gt;     R3          0          0</span>
<span class="co">#&gt;     R4          0          0</span>
<span class="co">#&gt;     R5          0          0</span>
<span class="co">#&gt;    ...          .          .</span>
<span class="co">#&gt; R27994          0          0</span>
<span class="co">#&gt; R27995          0          0</span>
<span class="co">#&gt; R27996          0          0</span>
<span class="co">#&gt; R27997          0          0</span>
<span class="co">#&gt; R27998          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tmp</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Set dimnames</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayednaryisoop" class="section level4">
<h4 class="hasAnchor">
<a href="#delayednaryisoop" class="anchor"></a>DelayedNaryIsoOp</h4>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts</span> + <span class="no">tenx_counts</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          2          0          0          4   .          0          2</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: N-ary iso op</span>
<span class="co">#&gt;    ├─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="delayedabind" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedabind" class="anchor"></a>DelayedAbind</h4>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html">cbind</a></span>(<span class="no">tenx_counts</span>, <span class="no">tenx_counts</span>)
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 2612254&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,2612251] [,2612252]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,2612253] [,2612254]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x2612254 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x2612254 integer: Abind (along=2)</span>
<span class="co">#&gt;    ├─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="no-op" class="section level4">
<h4 class="hasAnchor">
<a href="#no-op" class="anchor"></a>No-op</h4>
<p>The DelayedArray framework is smart enough to recognise that some combinations of operations are ‘no-ops’.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">val</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="no">tenx_counts</span>))
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: HDF5Matrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
<p>But it can be fooled.</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="co"># NOTE: This is a no-op but DelayedArray doesn't recognise it as one.</span>
<span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx_counts</span> + <span class="fl">0L</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: Unary iso op stack</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>

<span class="co"># NOTE: This also a no-op but DelayedArray doesn't recognise it as one.</span>
<span class="no">val</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html">cbind</a></span>(<span class="no">tenx_counts</span>) <span class="co"># TODO: No longer true</span>
<span class="no">val</span>
<span class="co">#&gt; &lt;27998 x 1306127&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;                [,1]       [,2]       [,3]       [,4] ... [,1306124] [,1306125]</span>
<span class="co">#&gt;     [1,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [2,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [3,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [4,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;     [5,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;      ...          .          .          .          .   .          .          .</span>
<span class="co">#&gt; [27994,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27995,]          1          0          0          2   .          0          1</span>
<span class="co">#&gt; [27996,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27997,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt; [27998,]          0          0          0          0   .          0          0</span>
<span class="co">#&gt;          [,1306126] [,1306127]</span>
<span class="co">#&gt;     [1,]          0          0</span>
<span class="co">#&gt;     [2,]          0          0</span>
<span class="co">#&gt;     [3,]          0          0</span>
<span class="co">#&gt;     [4,]          0          0</span>
<span class="co">#&gt;     [5,]          0          0</span>
<span class="co">#&gt;      ...          .          .</span>
<span class="co">#&gt; [27994,]          0          0</span>
<span class="co">#&gt; [27995,]          0          0</span>
<span class="co">#&gt; [27996,]          0          0</span>
<span class="co">#&gt; [27997,]          0          0</span>
<span class="co">#&gt; [27998,]          0          0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">val</span>)
<span class="co">#&gt; 27998x1306127 integer: HDF5Matrix object</span>
<span class="co">#&gt; └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
</div>
<div id="block-processing" class="section level3">
<h3 class="hasAnchor">
<a href="#block-processing" class="anchor"></a>Block-processing</h3>
<p>In <a href="#library-sizes">Library sizes</a>, we needed to compute the column sums of <code>tenx_counts_subset</code>, whose data live on disk in an HDF5 file. One way of achieving this would be to load the entire dataset into memory as an ordinary matrix and then run <code><a href="https://rdrr.io/r/base/colSums.html">base::colSums()</a></code><a href="#fn9" class="footnote-ref" id="fnref9"><sup>9</sup></a>.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">lib_sizes_in_mem</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/array.html">as.array</a></span>(<span class="no">tenx_counts_subset</span>))
<span class="co"># Check we get the same result as before.</span>
<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="no">lib_sizes_in_mem</span>, <span class="no">lib_sizes</span>)
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>Here the data at are small enough to load into memory, but what if that’s not the case<a href="#fn10" class="footnote-ref" id="fnref10"><sup>10</sup></a>? One way you might think to do this is to loop over the columns of the matrix, load that column into memory, and compute it’s sum<a href="#fn11" class="footnote-ref" id="fnref11"><sup>11</sup></a></p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">lib_sizes_loop_over_cols</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span>(<span class="st">"numeric"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))
<span class="kw">for</span> (<span class="no">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))) {
  <span class="co"># A simple progress report.</span>
  <span class="kw">if</span> (<span class="no">j</span> <span class="kw">%%</span> <span class="fl">100</span> <span class="kw">==</span> <span class="fl">0</span>) <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span>(<span class="st">"Processed "</span>, <span class="no">j</span>, <span class="st">"/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))
  <span class="no">lib_sizes_loop_over_cols</span>[<span class="no">j</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/array.html">as.array</a></span>(<span class="no">tenx_counts_subset</span>[, <span class="no">j</span>, <span class="kw">drop</span> <span class="kw">=</span> <span class="fl">FALSE</span>]))
}
<span class="co">#&gt; Processed 100/1000</span>
<span class="co">#&gt; Processed 200/1000</span>
<span class="co">#&gt; Processed 300/1000</span>
<span class="co">#&gt; Processed 400/1000</span>
<span class="co">#&gt; Processed 500/1000</span>
<span class="co">#&gt; Processed 600/1000</span>
<span class="co">#&gt; Processed 700/1000</span>
<span class="co">#&gt; Processed 800/1000</span>
<span class="co">#&gt; Processed 900/1000</span>
<span class="co">#&gt; Processed 1000/1000</span>
<span class="co"># Check we get the same result as before.</span>
<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="no">lib_sizes_loop_over_cols</span>, <span class="no">lib_sizes</span>)
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>But what if you can’t load even a single column into memory<a href="#fn12" class="footnote-ref" id="fnref12"><sup>12</sup></a>? We might loop over the columns of the matrix, partition each column, load each partition, compute its sum, and then compute the sum of the partition sums.</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="co"># NOTE: Here we will partition each column into two equal-sized subsets.</span>
<span class="no">lib_sizes_loop_over_cols_subset</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span>(<span class="st">"numeric"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))
<span class="no">tmp_colsums</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span>(<span class="st">"numeric"</span>, <span class="fl">2</span>)
<span class="no">nr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">tenx_counts_subset</span>)

<span class="kw">for</span> (<span class="no">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))) {
  <span class="co"># A simple progress report.</span>
  <span class="kw">if</span> (<span class="no">j</span> <span class="kw">%%</span> <span class="fl">100</span> <span class="kw">==</span> <span class="fl">0</span>) <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span>(<span class="st">"Processed "</span>, <span class="no">j</span>, <span class="st">"/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))
  <span class="kw">for</span> (<span class="no">i</span> <span class="kw">in</span> <span class="fl">1</span>:<span class="fl">2</span>) {
    <span class="no">i1</span> <span class="kw">&lt;-</span> (<span class="no">i</span> - <span class="fl">1</span>) * <span class="no">nr</span> / <span class="fl">2</span> + <span class="fl">1</span>
    <span class="no">i2</span> <span class="kw">&lt;-</span> <span class="no">i</span> * <span class="no">nr</span> / <span class="fl">2</span>
    <span class="no">tmp_colsums</span>[<span class="no">i</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(
      <span class="fu"><a href="https://rdrr.io/r/base/array.html">as.array</a></span>(<span class="no">tenx_counts_subset</span>[<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="no">i1</span>, <span class="no">i2</span>), <span class="no">j</span>, <span class="kw">drop</span> <span class="kw">=</span> <span class="fl">FALSE</span>]))
  }
  <span class="no">lib_sizes_loop_over_cols_subset</span>[<span class="no">j</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">tmp_colsums</span>)
}
<span class="co">#&gt; Processed 100/1000</span>
<span class="co">#&gt; Processed 200/1000</span>
<span class="co">#&gt; Processed 300/1000</span>
<span class="co">#&gt; Processed 400/1000</span>
<span class="co">#&gt; Processed 500/1000</span>
<span class="co">#&gt; Processed 600/1000</span>
<span class="co">#&gt; Processed 700/1000</span>
<span class="co">#&gt; Processed 800/1000</span>
<span class="co">#&gt; Processed 900/1000</span>
<span class="co">#&gt; Processed 1000/1000</span>
<span class="co"># Check we get the same result as before.</span>
<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="no">lib_sizes_loop_over_cols_subset</span>, <span class="no">lib_sizes</span>)
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>Hopefully you can now begin to see the general pattern, a strategy which the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package calls ‘block-processing’:</p>
<ol style="list-style-type: decimal">
<li>Load a ‘block’ of the data into memory.</li>
<li>Compute a summary statistic.</li>
<li>Combine the block-level statistics in an appropriate way to get the final result.</li>
</ol>
<div id="block-processing-illustrated" class="section level4">
<h4 class="hasAnchor">
<a href="#block-processing-illustrated" class="anchor"></a>Block-processing illustrated</h4>
<p>Some examples of block-processing are illustrated in the following figures:</p>
<p><img src="../../../../tmp/RtmpmQxQz3/temp_libpathc59376543be4/DelayedArrayWorkshop/vignettes/Slide1.png"><!-- --><img src="../../../../tmp/RtmpmQxQz3/temp_libpathc59376543be4/DelayedArrayWorkshop/vignettes/Slide2.png"><!-- --><img src="../../../../tmp/RtmpmQxQz3/temp_libpathc59376543be4/DelayedArrayWorkshop/vignettes/Slide3.png"><!-- --><img src="../../../../tmp/RtmpmQxQz3/temp_libpathc59376543be4/DelayedArrayWorkshop/vignettes/Slide4.png"><!-- --><img src="../../../../tmp/RtmpmQxQz3/temp_libpathc59376543be4/DelayedArrayWorkshop/vignettes/Slide5.png"><!-- --><img src="../../../../tmp/RtmpmQxQz3/temp_libpathc59376543be4/DelayedArrayWorkshop/vignettes/Slide6.png"><!-- --><img src="../../../../tmp/RtmpmQxQz3/temp_libpathc59376543be4/DelayedArrayWorkshop/vignettes/Slide7.png"><!-- --></p>
</div>
<div id="block-processed-column-sums" class="section level4">
<h4 class="hasAnchor">
<a href="#block-processed-column-sums" class="anchor"></a>Block-processed column sums</h4>
<p>When we run <code>colSums(tenx_counts_subset</code>) we are using a block-processed version of column sums, specifically the <code>colSums,DelayedMatrix-method</code> implemented in the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package. We can see this more clearly by turning on verbose progress reporting from the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="co"># Turn on progress reports for DelayedArray's block-processing routines.</span>
<span class="kw pkg">DelayedArray</span><span class="kw ns">:::</span><span class="fu">set_verbose_block_processing</span>(<span class="fl">TRUE</span>)
<span class="co">#&gt; [1] FALSE</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">tenx_counts_subset</span>))
<span class="co">#&gt; Processing block 1/2 ... OK</span>
<span class="co">#&gt; Processing block 2/2 ... OK</span>
<span class="co">#&gt; [1]  4046  2087  4654  3193  8444 11178</span></pre></body></html></div>
</div>
<div id="more-examples-of-block-processed-operations-in-delayedarray" class="section level4">
<h4 class="hasAnchor">
<a href="#more-examples-of-block-processed-operations-in-delayedarray" class="anchor"></a>More examples of block-processed operations in <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em>
</h4>
<p>Some of the most useful functions in the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package implement common operations on a <em>DelayedMatrix</em> using block-processing. These include the following row and column summarization methods:</p>
<ul>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowSums()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowMeans()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMeans()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowMaxs()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMaxs()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowMins()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMins()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">rowRanges()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colRanges()</a></code></li>
</ul>
<p>Two useful but lesser known functions use block-processing to compute column/row sums of a <em>DelayedMatrix</em> based on a grouping variable:</p>
<ul>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">rowsum()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">colsum()</a></code></li>
</ul>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="co"># Compute separate library sizes for mitochondrial and non-mitochondrial genes.</span>
<span class="no">is_mito</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"^mt-"</span>, <span class="fu">rowData</span>(<span class="no">tenx</span>)$<span class="no">Symbol</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">is_mito</span>)
<span class="co">#&gt;    Mode   FALSE    TRUE </span>
<span class="co">#&gt; logical   27985      13</span>
<span class="no">lib_sizes_grouped</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">rowsum</a></span>(<span class="no">tenx_counts_subset</span>, <span class="kw">group</span> <span class="kw">=</span> <span class="no">is_mito</span>)
<span class="co">#&gt; Processing block [[1/2, 1/1]] ... OK</span>
<span class="co">#&gt; Processing block [[2/2, 1/1]] ... OK</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html">t</a></span>(<span class="no">lib_sizes_grouped</span>))
<span class="co">#&gt;      FALSE TRUE</span>
<span class="co">#&gt; [1,]  3888  158</span>
<span class="co">#&gt; [2,]  2014   73</span>
<span class="co">#&gt; [3,]  4572   82</span>
<span class="co">#&gt; [4,]  3115   78</span>
<span class="co">#&gt; [5,]  8273  171</span>
<span class="co">#&gt; [6,] 10777  401</span></pre></body></html></div>
<p>Finally, matrix multiplication is implemented as a block-processed operation.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="co"># This is mathematically equivalent to rowSums(tenx_counts_subset).</span>
<span class="co"># TODO: Is this still using block-processing?</span>
<span class="no">tenx_counts_subset</span> <span class="kw">%*%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">1</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))
<span class="co">#&gt; &lt;27998 x 1&gt; matrix of class DelayedMatrix and type "double":</span>
<span class="co">#&gt;          [,1]</span>
<span class="co">#&gt;     [1,]   18</span>
<span class="co">#&gt;     [2,]    0</span>
<span class="co">#&gt;     [3,]    0</span>
<span class="co">#&gt;     [4,]    0</span>
<span class="co">#&gt;     [5,]    0</span>
<span class="co">#&gt;      ...    .</span>
<span class="co">#&gt; [27994,]    0</span>
<span class="co">#&gt; [27995,] 1122</span>
<span class="co">#&gt; [27996,]   93</span>
<span class="co">#&gt; [27997,]    0</span>
<span class="co">#&gt; [27998,]    1</span></pre></body></html></div>
</div>
<div id="delayedmatrixstats" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedmatrixstats" class="anchor"></a>DelayedMatrixStats</h4>
<p>We’ve already seen the <em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em> package in action back when computing the <a href="#median-expression-of-each-gene">Median expression of each gene</a>. <em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em> is a port of the <em><a href="https://CRAN.R-project.org/package=matrixStats">matrixStats</a></em> package’s API for use with <em>DelayedMatrix</em> objects. It provides <a href="https://github.com/PeteHaitch/DelayedMatrixStats#api-coverage">more than 70 functions</a> that apply to rows and columns of <em>DelayedMatrix</em> objects</p>
</div>
<div id="general-block-processing" class="section level4">
<h4 class="hasAnchor">
<a href="#general-block-processing" class="anchor"></a>General block-processing</h4>
<p>As we have seen, many common block-processing operations on a <em>DelayedMatrix</em> have already been implemented in <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> or are provided by <em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em>. Nonetheless, there may be times you need to implement your own algorithm using block-processing. The documentation on this topic is a little sparse, but some details can be found in <code><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">help("block_processing", "DelayedArray")</a></code> and <code><a href="https://rdrr.io/pkg/DelayedArray/man/ArrayGrid-class.html">help("ArrayGrid", "DelayedArray")</a></code> or by reading the source code of the aforementioned packages. Briefly, to perform block-processing requires that you:</p>
<ol style="list-style-type: decimal">
<li>Set up an <em>ArrayGrid</em> over your <em>DelayedArray</em>. This specifies the ‘block’ structure that will be traversed when processing the <em>DelayedArray</em>.</li>
<li>Iterate over the <em>DelayedArray</em> via the <em>ArrayGrid</em>
<ol style="list-style-type: lower-alpha">
<li>Read each block of data into memory as an ordinary array using <code><a href="https://rdrr.io/pkg/DelayedArray/man/read_block.html">read_block()</a></code>.</li>
<li>Compute the statistic for that block</li>
</ol>
</li>
<li>Appropriately combine the block-level statistics to get your final result.</li>
</ol>
<p>The <code><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">blockApply()</a></code> and <code><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">blockReduce()</a></code> functions can help facilitate steps 1-3 , even incorporating parallelization via the <em><a href="https://bioconductor.org/packages/3.12/BiocParallel">BiocParallel</a></em> package.</p>
</div>
</div>
<div id="realization" class="section level3">
<h3 class="hasAnchor">
<a href="#realization" class="anchor"></a>Realization</h3>
<p>To <em>realize</em> a <em>DelayedArray</em> object is to trigger execution of the delayed operations carried by the object and return the result as an ordinary array, One way to achieve this is to call <code><a href="https://rdrr.io/r/base/array.html">as.array()</a></code> on it.</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="no">tenx_counts_subset_realized</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/array.html">as.array</a></span>(<span class="no">tenx_counts_subset</span>)
<span class="no">tenx_counts_subset_realized</span>[<span class="fl">1</span>:<span class="fl">10</span>, <span class="fl">1</span>:<span class="fl">10</span>]
<span class="co">#&gt;       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span>
<span class="co">#&gt;  [1,]    0    0    0    0    0    0    0    0    0     0</span>
<span class="co">#&gt;  [2,]    0    0    0    0    0    0    0    0    0     0</span>
<span class="co">#&gt;  [3,]    0    0    0    0    0    0    0    0    0     0</span>
<span class="co">#&gt;  [4,]    0    0    0    0    0    0    0    0    0     0</span>
<span class="co">#&gt;  [5,]    0    0    0    0    0    0    0    0    0     0</span>
<span class="co">#&gt;  [6,]    0    0    0    0    0    0    0    0    0     0</span>
<span class="co">#&gt;  [7,]    0    0    0    0    0    0    0    0    0     0</span>
<span class="co">#&gt;  [8,]    0    0    0    0    2    0    2    0    1     1</span>
<span class="co">#&gt;  [9,]    0    0    1    0    1    0    0    0    0     0</span>
<span class="co">#&gt; [10,]    0    0    0    0    0    0    0    0    0     0</span></pre></body></html></div>
<p>However, this realizes the full object at once in memory which could require too much memory if the object is big<a href="#fn13" class="footnote-ref" id="fnref13"><sup>13</sup></a>.</p>
<p>A large <em>DelayedArray</em> object is preferably realized on disk, which is most commonly an HDF5 file.</p>
<div id="realizing-to-an-hdf5-file" class="section level4">
<h4 class="hasAnchor">
<a href="#realizing-to-an-hdf5-file" class="anchor"></a>Realizing to an HDF5 file</h4>
<p>Realizing to an HDF5 file requires that the <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> package is installed<a href="#fn14" class="footnote-ref" id="fnref14"><sup>14</sup></a></p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">tenx_counts_subset_hdf5</span> <span class="kw">&lt;-</span> <span class="fu">writeHDF5Array</span>(<span class="no">tenx_counts_subset</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>
<span class="no">tenx_counts_subset_hdf5</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span>

<span class="co"># Alternatively, we could 'coerce' the result to be an HDF5Array.</span>
<span class="fu">as</span>(<span class="no">tenx_counts_subset</span>, <span class="st">"HDF5Array"</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span></pre></body></html></div>
<p>Notice that this process of realization used block-processing, which avoids loading the entire dataset entire memory. Also notice that the result of realization is here returned as an <em>HDF5Matrix</em>, which no longer carries around the delayed operations.</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_counts_subset</span>)
<span class="co">#&gt; [1] "DelayedMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "DelayedArray"</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_counts_subset_hdf5</span>)
<span class="co">#&gt; [1] "HDF5Matrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span>

<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx_counts_subset</span>)
<span class="co">#&gt; 27998x1000 integer: DelayedMatrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer: Subset</span>
<span class="co">#&gt;    └─ 27998x1306127 integer: [seed] HDF5ArraySeed object</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="no">tenx_counts_subset_hdf5</span>)
<span class="co">#&gt; 27998x1000 integer: HDF5Matrix object</span>
<span class="co">#&gt; └─ 27998x1000 integer: [seed] HDF5ArraySeed object</span></pre></body></html></div>
<p>Used like this, <code>writeHDF5Array()</code> and <code>as(..., "HDF5Array")</code> will write their results to a file in <em>HDF5 dump directory</em>, a dumping ground for automatically created HDF5 datasets. We can see a log of the operations that have written to the HDF5 dump directory using <code>showHDF5DumpLog()</code>.</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu">showHDF5DumpLog</span>()
<span class="co">#&gt; [2020-06-05 09:58:06] #1 In file '/tmp/Rtmpuyl994/HDF5Array_dump/auto00001.h5': creation of dataset '/HDF5ArrayAUTO00001' (27998x1000:integer, chunkdims=5291x188, level=6)</span>
<span class="co">#&gt; [2020-06-05 09:58:08] #2 In file '/tmp/Rtmpuyl994/HDF5Array_dump/auto00002.h5': creation of dataset '/HDF5ArrayAUTO00002' (27998x1000:integer, chunkdims=5291x188, level=6)</span></pre></body></html></div>
<p>Often, however, we will want full control of where and how the data are written to the HDF5 file<a href="#fn15" class="footnote-ref" id="fnref15"><sup>15</sup></a> and the <code>writeHDF5Array()</code> function gives you full control over this and more.</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="co"># Write the data to a user-specified HDF5 file using maximum compression and </span>
<span class="co"># 'chunking' along the columns.</span>
<span class="no">my_hdf5_file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span>(<span class="kw">fileext</span> <span class="kw">=</span> <span class="st">".h5"</span>)
<span class="no">tenx_counts_subset_my_file_hdf5</span> <span class="kw">&lt;-</span> <span class="fu">writeHDF5Array</span>(
  <span class="no">tenx_counts_subset</span>,
  <span class="kw">filepath</span> <span class="kw">=</span> <span class="no">my_hdf5_file</span>,
  <span class="kw">chunkdim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">tenx_counts_subset</span>), <span class="fl">1</span>),
  <span class="kw">level</span> <span class="kw">=</span> <span class="fl">9</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>

<span class="co"># Compare `tenx_counts_subset_hdf5` to `tenx_counts_subset_my_file_hdf5`</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span>(<span class="no">tenx_counts_subset_hdf5</span>)
<span class="co">#&gt; An object of class "HDF5ArraySeed"</span>
<span class="co">#&gt; Slot "filepath":</span>
<span class="co">#&gt; [1] "/tmp/Rtmpuyl994/HDF5Array_dump/auto00001.h5"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "name":</span>
<span class="co">#&gt; [1] "/HDF5ArrayAUTO00001"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "type":</span>
<span class="co">#&gt; [1] "integer"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "dim":</span>
<span class="co">#&gt; [1] 27998  1000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "chunkdim":</span>
<span class="co">#&gt; [1] 5291  188</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "first_val":</span>
<span class="co">#&gt; [1] 0</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span>(<span class="no">tenx_counts_subset_my_file_hdf5</span>)
<span class="co">#&gt; An object of class "HDF5ArraySeed"</span>
<span class="co">#&gt; Slot "filepath":</span>
<span class="co">#&gt; [1] "/tmp/Rtmpuyl994/filec5f3459013f7.h5"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "name":</span>
<span class="co">#&gt; [1] "/HDF5ArrayAUTO00003"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "type":</span>
<span class="co">#&gt; [1] "integer"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "dim":</span>
<span class="co">#&gt; [1] 27998  1000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "chunkdim":</span>
<span class="co">#&gt; [1] 27998     1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "first_val":</span>
<span class="co">#&gt; [1] 0</span></pre></body></html></div>
</div>
<div id="realization-backends" class="section level4">
<h4 class="hasAnchor">
<a href="#realization-backends" class="anchor"></a>Realization backends</h4>
<p>We’ve now seen that we can realize to an HDF5 file. This is called as the HDF5Array ‘realization backend’ and is implemented in the <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em> package. There are a few other realization backends to be aware of.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">supportedRealizationBackends</a></span>()
<span class="co">#&gt;        BACKEND      package  </span>
<span class="co">#&gt; 1     RleArray DelayedArray  </span>
<span class="co">#&gt; 2    HDF5Array    HDF5Array  </span>
<span class="co">#&gt; 3   TENxMatrix    HDF5Array</span></pre></body></html></div>
<p>There is also the <code>NULL</code> backend, which means the data are realized in memory as an ordinary <em>array</em> and then wrapped in a <em>DelayedArray</em>. This is the default realization backend upon loading/attaching the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package.</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">getRealizationBackend</a></span>()
<span class="co">#&gt; NULL</span></pre></body></html></div>
<p>The default realization backend can be altered with <code><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend()</a></code>.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="st">"HDF5Array"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">getRealizationBackend</a></span>()
<span class="co">#&gt; [1] "HDF5Array"</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="kw">NULL</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">getRealizationBackend</a></span>()
<span class="co">#&gt; NULL</span></pre></body></html></div>
<p>It can be important to know what your current realization backend is because it will be used implicitly by some functions. For example, matrix multiplication that involves a <em>DelayedMatrix</em> uses the current realization backend.</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="st">"HDF5Array"</span>)
<span class="no">tenx_counts_subset</span> <span class="kw">%*%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">1</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))
<span class="co">#&gt; Realizing block 1/1 ... OK, writing it ... OK</span>
<span class="co">#&gt; &lt;27998 x 1&gt; matrix of class HDF5Matrix and type "double":</span>
<span class="co">#&gt;          [,1]</span>
<span class="co">#&gt;     [1,]   18</span>
<span class="co">#&gt;     [2,]    0</span>
<span class="co">#&gt;     [3,]    0</span>
<span class="co">#&gt;     [4,]    0</span>
<span class="co">#&gt;     [5,]    0</span>
<span class="co">#&gt;      ...    .</span>
<span class="co">#&gt; [27994,]    0</span>
<span class="co">#&gt; [27995,] 1122</span>
<span class="co">#&gt; [27996,]   93</span>
<span class="co">#&gt; [27997,]    0</span>
<span class="co">#&gt; [27998,]    1</span>

<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="kw">NULL</span>)
<span class="no">tenx_counts_subset</span> <span class="kw">%*%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">1</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_counts_subset</span>))
<span class="co">#&gt; &lt;27998 x 1&gt; matrix of class DelayedMatrix and type "double":</span>
<span class="co">#&gt;          [,1]</span>
<span class="co">#&gt;     [1,]   18</span>
<span class="co">#&gt;     [2,]    0</span>
<span class="co">#&gt;     [3,]    0</span>
<span class="co">#&gt;     [4,]    0</span>
<span class="co">#&gt;     [5,]    0</span>
<span class="co">#&gt;      ...    .</span>
<span class="co">#&gt; [27994,]    0</span>
<span class="co">#&gt; [27995,] 1122</span>
<span class="co">#&gt; [27996,]   93</span>
<span class="co">#&gt; [27997,]    0</span>
<span class="co">#&gt; [27998,]    1</span></pre></body></html></div>
</div>
<div id="the-realize-function" class="section level4">
<h4 class="hasAnchor">
<a href="#the-realize-function" class="anchor"></a>The <code>realize()</code> function</h4>
<p>We’ve seen that we can realize to the HDF5Array backend using <code>writeHDF5Array(tenx_counts_subset)</code> and <code>as(tenx_counts_subset, "HDF5Array")</code>. A third way of realizing a <em>DelayedArray</em> to an HDF5 file is with the <code><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize()</a></code> function.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize</a></span>(<span class="no">tenx_counts_subset</span>, <span class="kw">BACKEND</span> <span class="kw">=</span> <span class="st">"HDF5Array"</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class HDF5Matrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span></pre></body></html></div>
<p>So why might you use <code><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize()</a></code> instead of these other options? Because it allows us to easily switch out the realization backend and will defer to the current realization backend if none is supplied.</p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="co"># Realize to the current realization backend.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">getRealizationBackend</a></span>()
<span class="co">#&gt; NULL</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize</a></span>(<span class="no">tenx_counts_subset</span>)
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class DelayedMatrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span>

<span class="co"># Realize as an RleArray.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="st">"RleArray"</span>)
<span class="no">tenx_counts_subset_rlearray</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize</a></span>(<span class="no">tenx_counts_subset</span>)
<span class="co">#&gt; Realizing block 1/2 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/2 ... OK, writing it ... OK</span>
<span class="no">tenx_counts_subset_rlearray</span>
<span class="co">#&gt; &lt;27998 x 1000&gt; matrix of class RleMatrix and type "integer":</span>
<span class="co">#&gt;             [,1]    [,2]    [,3]    [,4] ...  [,997]  [,998]  [,999] [,1000]</span>
<span class="co">#&gt;     [1,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [2,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [3,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [4,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;     [5,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt;      ...       .       .       .       .   .       .       .       .       .</span>
<span class="co">#&gt; [27994,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27995,]       1       0       0       2   .       5       3       2       0</span>
<span class="co">#&gt; [27996,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27997,]       0       0       0       0   .       0       0       0       0</span>
<span class="co">#&gt; [27998,]       0       0       0       0   .       0       0       0       0</span>

<span class="co"># Switch back to the default backend.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/RealizationSink-class.html">setRealizationBackend</a></span>(<span class="kw">NULL</span>)</pre></body></html></div>
<p>This is probably most useful when writing package code so that you can allow the user control over the realization backend.</p>
</div>
</div>
</div>
<div id="package-ecosystem" class="section level2">
<h2 class="hasAnchor">
<a href="#package-ecosystem" class="anchor"></a>Package ecosystem</h2>
<p>The DelayedArray framework is, unsurprisingly, implemented in the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package. However, there are several other key packages that are an important part of the broader ‘ecosystem’. More importantly, as a user of Bioconductor software, it is increasingly likely that you will encounter <em>DelayedArray</em> objects during a data analysis, especially if you are analysing single-cell data<a href="#fn16" class="footnote-ref" id="fnref16"><sup>16</sup></a>. The following table lists packages that depend upon the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package.</p>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="no">dep_tbl</span> <span class="kw">&lt;-</span> <span class="kw pkg">BiocPkgTools</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocPkgTools/man/buildPkgDependencyDataFrame.html">buildPkgDependencyDataFrame</a></span>()
<span class="no">da_dep_tbl</span> <span class="kw">&lt;-</span> <span class="no">dep_tbl</span>[<span class="no">dep_tbl</span>$<span class="no">dependency</span> <span class="kw">==</span> <span class="st">"DelayedArray"</span>,
                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Package"</span>, <span class="st">"edgetype"</span>)]
<span class="no">da_dep_tbl</span> <span class="kw">&lt;-</span> <span class="no">da_dep_tbl</span>[<span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span>(<span class="no">da_dep_tbl</span>, <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="no">edgetype</span>, <span class="no">Package</span>)), ]
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">da_dep_tbl</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Package"</span>, <span class="st">"Dependency Type"</span>)
<span class="kw pkg">rmarkdown</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/rmarkdown/man/paged_table.html">paged_table</a></span>(<span class="no">da_dep_tbl</span>)</pre></body></html></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Package"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Dependency Type"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"DelayedDataFrame","2":"Depends"},{"1":"DelayedMatrixStats","2":"Depends"},{"1":"GDSArray","2":"Depends"},{"1":"HDF5Array","2":"Depends"},{"1":"restfulSE","2":"Depends"},{"1":"rhdf5client","2":"Depends"},{"1":"singleCellTK","2":"Depends"},{"1":"SummarizedExperiment","2":"Depends"},{"1":"VCFArray","2":"Depends"},{"1":"batchelor","2":"Imports"},{"1":"beachmat","2":"Imports"},{"1":"bigPint","2":"Imports"},{"1":"BiocSingular","2":"Imports"},{"1":"bsseq","2":"Imports"},{"1":"CAGEr","2":"Imports"},{"1":"celaref","2":"Imports"},{"1":"celda","2":"Imports"},{"1":"ChIPpeakAnno","2":"Imports"},{"1":"clusterExperiment","2":"Imports"},{"1":"DEScan2","2":"Imports"},{"1":"DSS","2":"Imports"},{"1":"ELMER","2":"Imports"},{"1":"FRASER","2":"Imports"},{"1":"GenoGAM","2":"Imports"},{"1":"GenomicScores","2":"Imports"},{"1":"glmGamPoi","2":"Imports"},{"1":"gramm4R","2":"Imports"},{"1":"hipathia","2":"Imports"},{"1":"LoomExperiment","2":"Imports"},{"1":"mbkmeans","2":"Imports"},{"1":"methrix","2":"Imports"},{"1":"methylSig","2":"Imports"},{"1":"minfi","2":"Imports"},{"1":"netSmooth","2":"Imports"},{"1":"PCAtools","2":"Imports"},{"1":"RTCGAToolbox","2":"Imports"},{"1":"scater","2":"Imports"},{"1":"scDblFinder","2":"Imports"},{"1":"scMerge","2":"Imports"},{"1":"scmeth","2":"Imports"},{"1":"scran","2":"Imports"},{"1":"scuttle","2":"Imports"},{"1":"signatureSearch","2":"Imports"},{"1":"SingleR","2":"Imports"},{"1":"VariantExperiment","2":"Imports"},{"1":"weitrix","2":"Imports"},{"1":"BiocGenerics","2":"Suggests"},{"1":"gwascat","2":"Suggests"},{"1":"iSEE","2":"Suggests"},{"1":"MAST","2":"Suggests"},{"1":"S4Vectors","2":"Suggests"},{"1":"SQLDataFrame","2":"Suggests"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>We will briefly highlight some of the key packages in this table, broadly categorising these as ‘user-focused’/‘user-facing’ or ‘developer-focused’ packages and those that span the spectrum.</p>
<div id="packages-that-both-users-and-developers-should-probably-know-about" class="section level3">
<h3 class="hasAnchor">
<a href="#packages-that-both-users-and-developers-should-probably-know-about" class="anchor"></a>Packages that both users and developers should probably know about</h3>
<div id="delayedarray" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedarray" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em>
</h4>
<p>Well, duh. Implements the <em>DelayedArray</em> and <em>RleArray</em> classes, along with all the fundamentals the enable the delayed operations, block processing, and realization that underpin the DelayedArray framework.</p>
</div>
<div id="hdf5array" class="section level4">
<h4 class="hasAnchor">
<a href="#hdf5array" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em>
</h4>
<p>Implements the <em>HDF5Array</em> and <em>TENxMatrix</em> classes, two convenient and memory-efficient array-like containers for on-disk representation of HDF5 datasets. <em>HDF5Array</em> is for datasets that use the conventional (i.e. dense) HDF5 representation. <em>TENxMatrix</em> is for datasets that use the HDF5-based sparse matrix representation from 10x Genomics.</p>
</div>
<div id="vcfarray-and-gdsarray" class="section level4">
<h4 class="hasAnchor">
<a href="#vcfarray-and-gdsarray" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/VCFArray">VCFArray</a></em> and <em><a href="https://bioconductor.org/packages/3.12/GDSArray">GDSArray</a></em>
</h4>
<p>Implements the <em>VCFArray</em> and <em>GDSArray</em> classes, types of <em>DelayedArray</em>, to represent VCF files and GDS-files in an array-like representation. VCF and GDS files are widely used to represent genotyping or sequence data.</p>
</div>
<div id="rhdf5client-and-restfulse" class="section level4">
<h4 class="hasAnchor">
<a href="#rhdf5client-and-restfulse" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/rhdf5client">rhdf5client</a></em> and <em><a href="https://bioconductor.org/packages/3.12/restfulSE">restfulSE</a></em>
</h4>
<p>Provide functions and classes to interface with remote data stores by operating on <em>SummarizedExperiment</em>-like objects. These data are HDF5 files living on a remote server running <code>h5serv</code>, a REST-based service for HDF5 data.</p>
</div>
</div>
<div id="user-focuseduser-facing-packages" class="section level3">
<h3 class="hasAnchor">
<a href="#user-focuseduser-facing-packages" class="anchor"></a>User-focused/user-facing packages</h3>
<p>These are the packages that as a user you might directly load/attach with <code>library()</code> as part of a data analysis. Alternatively, these may be loaded/attached as a dependency<a href="#fn17" class="footnote-ref" id="fnref17"><sup>17</sup></a> of another package you load/attach as part of an analysis.</p>
<div id="dropletutils" class="section level4">
<h4 class="hasAnchor">
<a href="#dropletutils" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/DropletUtils">DropletUtils</a></em>
</h4>
<p>Provides a number of utility functions for handling single-cell (RNA-seq) data from droplet technologies such as 10X Genomics. This includes <code>read10xCounts()</code> for data loading from the count matrices produced by 10x Genomics’ <strong>CellRanger</strong> software, which may be stored in an HDF5 file. To do this, it makes use of the <em>TENxMatrix</em> class.</p>
</div>
<div id="loomexperiment" class="section level4">
<h4 class="hasAnchor">
<a href="#loomexperiment" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/LoomExperiment">LoomExperiment</a></em>
</h4>
<p>Provides a means to convert from ‘loom’ files to standard Bioconductor classes and back again. The <a href="http://linnarssonlab.org/loompy/index.html">Loom file format</a> uses HDF5 to store experimental data and is used by some tools and labs producing data using single-cell assays. This includes the <code>import()</code> function for data loading from loom files into an <em>HDF5Matrix</em>.</p>
</div>
<div id="scater" class="section level4">
<h4 class="hasAnchor">
<a href="#scater" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/scater">scater</a></em>
</h4>
<p>A collection of tools for doing various analyses of single-cell RNA-seq gene expression data, with a focus on quality control and visualization. These methods work on <em>DelayedMatrix</em> objects as well as ordinary <em>matrix</em> objects and some sparse matrix objects from the <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> package.</p>
</div>
<div id="batchelor" class="section level4">
<h4 class="hasAnchor">
<a href="#batchelor" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/batchelor">batchelor</a></em>
</h4>
<p>Implements a variety of methods for batch correction of single-cell (RNA sequencing) data, such as<code>multiBatchPCA()</code> and <code>fastMNN()</code>. These methods work on <em>DelayedMatrix</em> objects as well as ordinary <em>matrix</em> objects and some sparse matrix objects from the <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> package.</p>
</div>
<div id="biocsingular" class="section level4">
<h4 class="hasAnchor">
<a href="#biocsingular" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/BiocSingular">BiocSingular</a></em>
</h4>
<p>Implements exact and approximate methods for singular value decomposition and principal components analysis using a framework that allows them to be easily switched within Bioconductor packages or workflows. These methods work on <em>DelayedMatrix</em> objects as well as ordinary <em>matrix</em> objects and some sparse matrix objects from the <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> package.</p>
</div>
<div id="mbkmeans" class="section level4">
<h4 class="hasAnchor">
<a href="#mbkmeans" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/mbkmeans">mbkmeans</a></em>
</h4>
<p>Implements the mini-batch k-means algorithm for large datasets, including support for on-disk data representation (i.e. <em>HDF5Matrix</em> objects).</p>
</div>
<div id="bsseq" class="section level4">
<h4 class="hasAnchor">
<a href="#bsseq" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/bsseq">bsseq</a></em>
</h4>
<p>A collection of tools for analyzing and visualizing bisulfite sequencing data. This was one of the first packages to make use of the DelayedArray framework and it supports these throughout the package. This was needed in order to store and analyse large non-CpG methylation datasets (&gt; 300 million loci, hundreds of samples) using HDF5 files. Disclaimer: I did this re-write of <em><a href="https://bioconductor.org/packages/3.12/bsseq">bsseq</a></em> and learnt a lot along the way.</p>
</div>
<div id="minfi" class="section level4">
<h4 class="hasAnchor">
<a href="#minfi" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/minfi">minfi</a></em>
</h4>
<p>Tools to analyze &amp; visualize Illumina Infinium methylation arrays. This doesn’t have the same level of support for <em>DelayedMatrix</em> objects as <em><a href="https://bioconductor.org/packages/3.12/bsseq">bsseq</a></em>, but perhaps one day. This is needed in order to store and analyse large methylation datasets (&gt; 850,000 loci, tens of thousands of) using HDF5 files. Disclaimer: This was the second package, after <em><a href="https://bioconductor.org/packages/3.12/bsseq">bsseq</a></em>, I started to re-write to support the DelayedArray framework. Here, it is rather more difficult because it is a ‘widely’ used package and has code from lots of different authors with different styles.</p>
</div>
<div id="delayedmatrixstats-1" class="section level4">
<h4 class="hasAnchor">
<a href="#delayedmatrixstats-1" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/DelayedMatrixStats">DelayedMatrixStats</a></em>
</h4>
<p>A port of the <em><a href="https://CRAN.R-project.org/package=matrixStats">matrixStats</a></em> API for use with <em>DelayedMatrix</em> objects. High-performing functions operating on rows and columns of <em>DelayedMatrix</em> objects, e.g. <code>col</code> / <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colMedians.html">rowMedians()</a></code>, <code>col</code> / <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colRanks.html">rowRanks()</a></code>, and <code>col</code> / <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/colMads.html">rowSds()</a></code>. Functions optimized per data type and for subsetted calculations such that both memory usage and processing time is minimized. Disclaimer: I wrote this one, too.</p>
</div>
</div>
<div id="developer-focused-packages" class="section level3">
<h3 class="hasAnchor">
<a href="#developer-focused-packages" class="anchor"></a>Developer-focused packages</h3>
<div id="beachmat" class="section level4">
<h4 class="hasAnchor">
<a href="#beachmat" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.12/beachmat">beachmat</a></em>
</h4>
<p>Provides a consistent C++ class interface for reading from and writing data to a variety of commonly used matrix types. Ordinary matrices and several sparse/dense <em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> classes are directly supported, third-party S4 classes may be supported by external linkage (such as the <em>HDF5Matrix</em> class), while all other matrices are handled by DelayedArray block processing.</p>
</div>
</div>
</div>
<div id="real-world-encounter-with-delayedarray-analysing-scrna-seq-data" class="section level2">
<h2 class="hasAnchor">
<a href="#real-world-encounter-with-delayedarray-analysing-scrna-seq-data" class="anchor"></a>Real world encounter with DelayedArray analysing scRNA-seq data</h2>
<p><strong>TODO:</strong> Remove this section? Didn’t get through it last time and it overlaps with OSCA and other workshops.</p>
<p>We will perform a brief analysis of a peripheral blood mononuclear cell (PBMC) scRNA-seq dataset from 10X Genomics <span class="citation">(Zheng et al. 2017)</span> to illustrate a real world analysis that makes use of the DelayedArray framework<a href="#fn18" class="footnote-ref" id="fnref18"><sup>18</sup></a>.</p>
<p>This dataset contains more than 4000 cells, which isn’t that large in the grand scheme of things, but does allow us to demonstrate some important concepts in a workshop format. It has been pre-packaged and is available from the <em><a href="https://bioconductor.org/packages/3.12/TENxPBMCData">TENxPBMCData</a></em> package as a <em>SingleCellExperiment</em> object with a <em>HDF5Matrix</em> in the counts assay. In order to efficiently analyse this data, we would like to use software that supports the <em>HDF5Matrix</em> class via the DelayedArray framework. Thanks to the tireless efforts of Aaron Lun, this is true of many of the key Bioconductor packages for analysing scRNA-seq data: <em><a href="https://bioconductor.org/packages/3.12/DropletUtils">DropletUtils</a></em>, <em><a href="https://bioconductor.org/packages/3.12/scater">scater</a></em>, <em><a href="https://bioconductor.org/packages/3.12/scran">scran</a></em>, and <em><a href="https://bioconductor.org/packages/3.12/BiocSingular">BiocSingular</a></em>.</p>
<div id="loading-the-data" class="section level4">
<h4 class="hasAnchor">
<a href="#loading-the-data" class="anchor"></a>Loading the data</h4>
<p>A typical analysis of a 10x Genomics dataset would start by using the <code>DropletUtils::read10xCounts()</code> function to load the output of 10x Genomics’ CellRanger software into R. One of the CellRanger outputs is an HDF5 file, which <code>DropletUtils::read10xCounts()</code> can read and return as a <em>TENxMatrix</em>, an HDF5-backed <em>DelayedMatrix</em>. For this workshop, we’ll use one we prepared earlier and simply load the ‘filtered’ dataset from the <em><a href="https://bioconductor.org/packages/3.12/TENxPBMCData">TENxPBMCData</a></em> package which stores the count data in an <em>HDF5Matrix</em>.</p>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TENxPBMCData</span>)
<span class="co"># NOTE: This will download the data and may take a little while on the first </span>
<span class="co">#       run. The result will be cached, however, so subsequent runs avoid </span>
<span class="co">#       re-downloading the data.</span>
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html">TENxPBMCData</a></span>(<span class="st">"pbmc4k"</span>)
<span class="no">sce</span>
<span class="co"># NOTE: The counts data are stored in an HDF5Matrix.</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="fu">assay</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>))</pre></body></html></div>
<p>We’ll also switch to use the gene symbols provided by 10x Genomics rather than referring to genes by their Ensembl IDs.</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">sce</span>) <span class="kw">&lt;-</span> <span class="fu">uniquifyFeatureNames</span>(
  <span class="fu">rowData</span>(<span class="no">sce</span>)$<span class="no">ENSEMBL_ID</span>,
  <span class="fu">rowData</span>(<span class="no">sce</span>)$<span class="no">Symbol_TENx</span>)</pre></body></html></div>
</div>
<div id="quality-control-on-the-cells" class="section level4">
<h4 class="hasAnchor">
<a href="#quality-control-on-the-cells" class="anchor"></a>Quality control on the cells</h4>
<p>After loading the data into R, we compute some quality control metrics of the data using <code>scater::addPerCellQC()</code>. Under the hood, this involves computing column sums, comparisons of values, and subsetting of the HDF5-backed count matrix, all seamlessly handled by this function.</p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="co"># NOTE: Need to know which genes are mitochondrial to perform QC.</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">EnsDb.Hsapiens.v86</span>)
<span class="no">location</span> <span class="kw">&lt;-</span> <span class="fu">mapIds</span>(
  <span class="no">EnsDb.Hsapiens.v86</span>,
  <span class="kw">keys</span> <span class="kw">=</span> <span class="fu">rowData</span>(<span class="no">sce</span>)$<span class="no">ENSEMBL_ID</span>,
  <span class="kw">column</span> <span class="kw">=</span> <span class="st">"SEQNAME"</span>,
  <span class="kw">keytype</span> <span class="kw">=</span> <span class="st">"GENEID"</span>)
<span class="no">is_mito</span> <span class="kw">&lt;-</span> <span class="no">location</span> <span class="kw">==</span> <span class="st">"MT"</span>

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu">addPerCellQC</span>(<span class="no">sce</span>, <span class="kw">feature_controls</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">Mito</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">is_mito</span>)))
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/multiplot.html">multiplot</a></span>(
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html">plotColData</a></span>(<span class="no">sce</span>, <span class="st">"sum"</span>) +
    <span class="fu">ylab</span>(<span class="st">"Log-total UMI count"</span>) +
    <span class="fu">scale_y_log10</span>(),
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html">plotColData</a></span>(<span class="no">sce</span>, <span class="st">"detected"</span>) +
    <span class="fu">ylab</span>(<span class="st">"Log-total number of expressed features"</span>) +
    <span class="fu">scale_y_log10</span>(),
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html">plotColData</a></span>(<span class="no">sce</span>, <span class="st">"subsets_Mito_percent"</span>) +
    <span class="fu">ylab</span>(<span class="st">"Proportion of reads in mitochondrial genes"</span>),
  <span class="kw">cols</span> <span class="kw">=</span> <span class="fl">2</span>)

<span class="co"># We'll discard cells with large mitochondrial proportions, using it as a proxy</span>
<span class="co"># for cell damage.</span>
<span class="no">high_mito</span> <span class="kw">&lt;-</span> <span class="fu">isOutlier</span>(<span class="no">sce</span>$<span class="no">pct_counts_Mito</span>, <span class="kw">nmads</span> <span class="kw">=</span> <span class="fl">3</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"higher"</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="no">sce</span>[, !<span class="no">high_mito</span>]</pre></body></html></div>
</div>
<div id="examining-gene-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#examining-gene-expression" class="anchor"></a>Examining gene expression</h3>
<p>We can make a plot of the distribution of the most highly expressed genes using <code><a href="https://rdrr.io/pkg/scater/man/plotHighestExprs.html">scater::plotHighestExprs()</a></code>. Under the hood, this involves computing row and column sums of the HDF5-backed count matrix. Again, this is seamlessly handled by this function.</p>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotHighestExprs.html">plotHighestExprs</a></span>(<span class="no">sce</span>)</pre></body></html></div>
</div>
<div id="normalization" class="section level3">
<h3 class="hasAnchor">
<a href="#normalization" class="anchor"></a>Normalization</h3>
<p>Let’s move onto something a little more computationally challenging. Proper normalization is essential for all analyses of gene expression data. We apply the deconvolution method of Lun, Bach, and Marioni (2016) to compute size factors for all cells.</p>
<p>For highly heterogeneous datasets, like this sample of PBMCs, it is advisable to perform a rough clustering of the cells to better satisfy the assumptions of this normalization method. Namely, we want to avoid normalizing together cells with a large number of differentially expressed genes between them.</p>
<p>We will use the <code><a href="https://rdrr.io/pkg/scran/man/quickCluster.html">scran::quickCluster()</a></code> function to perform a clustering based on the principal component scores generated from the log-expression matrix. This principal component analysis (PCA) in turn uses an approximate singular value decomposition (SVD) with the augmented implicitly restarted Lanczos bidiagonalization algorithm (<strong>irlba</strong>). If that all sounds rather complicated, then don’t worry: that’s the point of this example! We are able to apply these cutting edge techniques to our HDF5-backed data<a href="#fn19" class="footnote-ref" id="fnref19"><sup>19</sup></a>.</p>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scran</span>)
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1000</span>)
<span class="no">clusters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html">quickCluster</a></span>(
  <span class="co"># NOTE: We're gonna cheat here to save time in the workshop.</span>
  <span class="co">#       We'll load the counts into memory as a sparse matrix and compute the </span>
  <span class="co">#       clustering on that.</span>
  <span class="kw">x</span> <span class="kw">=</span> <span class="fu">as</span>(<span class="fu">counts</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>), <span class="st">"dgCMatrix"</span>),
  <span class="kw">use.ranks</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
  <span class="kw">BSPARAM</span> <span class="kw">=</span> <span class="kw pkg">BiocSingular</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html">IrlbaParam</a></span>())
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html">table</a></span>(<span class="no">clusters</span>)</pre></body></html></div>
<p>Following the clustering, we apply the deconvolution method to compute size factors for all cells <span class="citation">(Lun, Bach, and Marioni 2016)</span>.</p>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html">computeSumFactors</a></span>(<span class="no">sce</span>, <span class="kw">min.mean</span> <span class="kw">=</span> <span class="fl">0.1</span>, <span class="kw">cluster</span> <span class="kw">=</span> <span class="no">clusters</span>)</pre></body></html></div>
<p>Finally, we compute normalized log-expression values from the count data using the size factors we just computed. Normalized expression values are computed by dividing the counts for each cell by the size factor for that cell.</p>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu">normalize</span>(<span class="no">sce</span>)
<span class="co"># NOTE: Normalized counts are available in the 'logcounts' assay.</span>
<span class="fu">logcounts</span>(<span class="no">sce</span>)</pre></body></html></div>
<p>This final step is near instantaneous because it is a delayed operation on the counts matrix.</p>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="fu">logcounts</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>))
<span class="co"># The counts and logcounts assays share the same seed.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span>(<span class="fu">counts</span>(<span class="no">sce</span>))
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span>(<span class="fu">logcounts</span>(<span class="no">sce</span>))</pre></body></html></div>
</div>
<div id="dimensionality-reduction" class="section level3">
<h3 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality reduction</h3>
<p>Skipping over some details, we use functions from the <em><a href="https://bioconductor.org/packages/3.12/scran">scran</a></em> package to decompose the variation in the log-expression matrix to a technical and biological component.</p>
<div class="sourceCode" id="cb55"><html><body><pre class="r"><span class="no">tech_trend</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/defunct.html">makeTechTrend</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">sce</span>)
<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/defunct.html">trendVar</a></span>(<span class="no">sce</span>, <span class="kw">use.spikes</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">loess.args</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">span</span> <span class="kw">=</span> <span class="fl">0.05</span>))
<span class="no">fit</span>$<span class="no">trend</span> <span class="kw">&lt;-</span> <span class="no">tech_trend</span>
<span class="no">dec</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/defunct.html">decomposeVar</a></span>(<span class="kw">fit</span> <span class="kw">=</span> <span class="no">fit</span>)</pre></body></html></div>
<p>We can then perform a principal component analysis (PCA) of the log-expression matrix to create a representation of the data using fewer dimensions. In doing so, we also denoise the log-expression data by removing PCs corresponding to technical noise.</p>
<p>In the interests of time, we are again going to cheat by loading the log-expression matrix into memory as a sparse matrix before doing the PCA. However, the same function works on the HDF5-backed data, albeit more slowly than when applied to an in-memory version of the data.</p>
<div class="sourceCode" id="cb56"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1000</span>)
<span class="no">pcs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html">denoisePCA</a></span>(
  <span class="co"># NOTE: We're gonna cheat here to save time. </span>
  <span class="co">#       We'll load the log-expression matrix into memory as a sparse matrix </span>
  <span class="co">#       and compute the PCA on that.</span>
  <span class="co">#       To run on the HDF5-backed data, simply use `x = logcounts(sce)` </span>
  <span class="co">#       (which will return the PCs an ordinary matrix) or `x = sce` (which </span>
  <span class="co">#       will return a SingleCellExperiment that is a modified version of </span>
  <span class="co">#       `sce` containing the PCs as an ordinary matrix as the "PCA" entry in </span>
  <span class="co">#       the reducedDims slot).</span>
  <span class="kw">x</span> <span class="kw">=</span> <span class="fu">as</span>(<span class="fu">logcounts</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>), <span class="st">"dgCMatrix"</span>),
  <span class="kw">technical</span> <span class="kw">=</span> <span class="no">tech_trend</span>,
  <span class="kw">BSPARAM</span> <span class="kw">=</span> <span class="kw pkg">BiocSingular</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html">IrlbaParam</a></span>())
<span class="co"># NOTE: Store the PCs in the SingleCellExperiment.</span>
<span class="fu">reducedDim</span>(<span class="no">sce</span>, <span class="st">"PCA"</span>) <span class="kw">&lt;-</span> <span class="no">pcs</span></pre></body></html></div>
</div>
<div id="clustering-and-visualization" class="section level3">
<h3 class="hasAnchor">
<a href="#clustering-and-visualization" class="anchor"></a>Clustering and visualization</h3>
<p>Our <em>SingleCellExperiment</em> object now contains:</p>
<ol style="list-style-type: decimal">
<li>The counts matrix: stored on disk in an HDF5 file and wrapped in a <em>DelayedMatrix</em>.</li>
<li>The normalized log-expression matrix: represented by a <em>DelayedMatrix</em> as delayed operations on top of the counts matrix.</li>
<li>The principal components (PCs): stored in-memory as an ordinary matrix</li>
</ol>
<p>The key at this point is that PCs are stored in-memory. Much of the downstream analysis makes use of this representation rather than the counts or log-expression matrices. In particular, we demonstrate that we can perform a clustering and <span class="math inline">\(t\)</span>-SNE representation of the data.</p>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="co"># Clustering</span>
<span class="no">snn_gr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span>(<span class="no">sce</span>, <span class="kw">use.dimred</span> <span class="kw">=</span> <span class="st">"PCA"</span>)
<span class="no">clusters</span> <span class="kw">&lt;-</span> <span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>(<span class="no">snn_gr</span>)
<span class="no">sce</span>$<span class="no">Cluster</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">clusters</span>$<span class="no">membership</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html">table</a></span>(<span class="no">sce</span>$<span class="no">Cluster</span>)

<span class="co"># t-SNE</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1000</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE</a></span>(<span class="no">sce</span>, <span class="kw">use_dimred</span> <span class="kw">=</span> <span class="st">"PCA"</span>, <span class="kw">perplexity</span> <span class="kw">=</span> <span class="fl">30</span>)

<span class="co"># Visualize clustering on t-SNE plot.</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotTSNE</a></span>(<span class="no">sce</span>, <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"Cluster"</span>)</pre></body></html></div>
</div>
<div id="concluding-remarks" class="section level3">
<h3 class="hasAnchor">
<a href="#concluding-remarks" class="anchor"></a>Concluding remarks</h3>
<p>This analysis illustrates that we can analyse scRNA-seq data that is stored on disk using tools that leverage the DelayedArray framework. More importantly, the same functions and workflow apply regardless of how the data are stored.</p>
<p>The biggest discrepancy in performance is when performing the principal component analysis of the data. However, by trading off additional computational time for reduced memory requirements, this is fully achievable and the resulting PCs can be saved for downstream analysis. This puts the DelayedArray-backed analysis on par with a fully in-memory-based analysis and with the advantage of a reduced in-memory footprint.</p>
</div>
</div>
<div id="workflow-tips-for-delayedarray-backed-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-tips-for-delayedarray-backed-analyses" class="anchor"></a>Workflow tips for DelayedArray-backed analyses</h2>
<p>We’ll conclude with some miscellaneous tips I’ve collected over the past few years of using DelayedArray-backed workflows.</p>
<div id="saving-and-loading-hdf5-backed-summarizedexperiment-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#saving-and-loading-hdf5-backed-summarizedexperiment-objects" class="anchor"></a>Saving and loading HDF5-backed SummarizedExperiment objects</h3>
<div class="sourceCode" id="cb58"><html><body><pre class="r"><span class="co"># TODO: Better/simpler example dataset</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TENxPBMCData</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scran</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="co">#&gt; Loading required package: scuttle</span>
<span class="co">#&gt; Loading required package: ggplot2</span>
<span class="co"># NOTE: This will download the data and may take a little while on the first </span>
<span class="co">#       run. The result will be cached, however, so subsequent runs avoid </span>
<span class="co">#       re-downloading the data.</span>
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html">TENxPBMCData</a></span>(<span class="st">"pbmc4k"</span>)
<span class="co">#&gt; snapshotDate(): 2020-06-01</span>
<span class="co">#&gt; see ?TENxPBMCData and browseVignettes('TENxPBMCData') for documentation</span>
<span class="co">#&gt; downloading 1 resources</span>
<span class="co">#&gt; retrieving 1 resource</span>
<span class="co">#&gt; loading from cache</span>
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html">computeSumFactors</a></span>(<span class="no">sce</span>, <span class="kw">min.mean</span> <span class="kw">=</span> <span class="fl">0.1</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu">logNormCounts</span>(<span class="no">sce</span>)</pre></body></html></div>
<div id="short-version" class="section level4">
<h4 class="hasAnchor">
<a href="#short-version" class="anchor"></a>Short version</h4>
<p>Use <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">HDF5Array::saveHDF5SummarizedExperiment()</a></code> and <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">HDF5Array::loadHDF5SummarizedExperiment()</a></code> rather than <code><a href="https://rdrr.io/r/base/readRDS.html">saveRDS()</a></code> and <code><a href="https://rdrr.io/r/base/readRDS.html">readRDS()</a></code> or <code><a href="https://rdrr.io/r/base/save.html">save()</a></code> and <code><a href="https://rdrr.io/r/base/load.html">load()</a></code>. Calling <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">HDF5Array::saveHDF5SummarizedExperiment()</a></code> will realize any delayed operations prior to saving the assay data in an HDF5 file, as illustrated below.</p>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="no">dir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>(), <span class="st">"my_h5_se"</span>)
<span class="fu">saveHDF5SummarizedExperiment</span>(<span class="no">sce</span>, <span class="no">dir</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">#&gt; Start writing assay 1/2 to HDF5 file:</span>
<span class="co">#&gt;   /tmp/Rtmpuyl994/my_h5_se/assays.h5</span>
<span class="co">#&gt; Realizing block 1/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 3/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 4/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 5/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 6/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 7/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 8/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 9/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 10/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 11/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 12/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 13/13 ... OK, writing it ... OK</span>
<span class="co">#&gt; Finished writing assay 1/2 to HDF5 file:</span>
<span class="co">#&gt;   /tmp/Rtmpuyl994/my_h5_se/assays.h5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Start writing assay 2/2 to HDF5 file:</span>
<span class="co">#&gt;   /tmp/Rtmpuyl994/my_h5_se/assays.h5</span>
<span class="co">#&gt; Realizing block 1/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 3/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 4/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 5/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 6/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 7/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 8/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 9/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 10/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 11/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 12/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 13/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 14/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 15/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 16/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 17/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 18/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 19/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 20/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 21/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 22/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 23/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 24/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 25/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 26/26 ... OK, writing it ... OK</span>
<span class="co">#&gt; Finished writing assay 2/2 to HDF5 file:</span>
<span class="co">#&gt;   /tmp/Rtmpuyl994/my_h5_se/assays.h5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Serialize SingleCellExperiment object to RDS file:</span>
<span class="co">#&gt;   /tmp/Rtmpuyl994/my_h5_se/se.rds</span>

<span class="co"># NOTE: There are no delayed operations on the saved version.</span>
<span class="no">saved_sce</span> <span class="kw">&lt;-</span> <span class="fu">loadHDF5SummarizedExperiment</span>(<span class="no">dir</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="fu">logcounts</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>))
<span class="co">#&gt; 33694x4340 double: DelayedMatrix object</span>
<span class="co">#&gt; └─ 33694x4340 double: Set dimnames</span>
<span class="co">#&gt;    └─ 33694x4340 double: Unary iso op stack</span>
<span class="co">#&gt;       └─ 33694x4340 double: Aperm (perm=c(2,1))</span>
<span class="co">#&gt;          └─ 4340x33694 double: Unary iso op with args</span>
<span class="co">#&gt;             └─ 4340x33694 integer: Aperm (perm=c(2,1))</span>
<span class="co">#&gt;                └─ 33694x4340 integer: [seed] HDF5ArraySeed object</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">showtree</a></span>(<span class="fu">logcounts</span>(<span class="no">saved_sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>))
<span class="co">#&gt; 33694x4340 double: HDF5Matrix object</span>
<span class="co">#&gt; └─ 33694x4340 double: [seed] HDF5ArraySeed object</span></pre></body></html></div>
</div>
<div id="long-version" class="section level4">
<h4 class="hasAnchor">
<a href="#long-version" class="anchor"></a>Long version</h4>
<p>A HDF5-backed <em>SummarizedExperiment</em>, like the 10x PBMC dataset we analysed in <a href="#real-world-encounter-with-delayedarray-analysing-scrna-seq-data">Real world encounter with DelayedArray analysing scRNA-seq data</a>, is a light-weight shell (the <em>SummarizedExperiment</em>) around a large disk-backed data matrix (the <em>HDF5Matrix</em>). The following explanation comes from <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">help("saveHDF5SummarizedExperiment", "HDF5Array")</a></code>:</p>
<p>Roughly speaking, <code><a href="https://rdrr.io/r/base/readRDS.html">saveRDS()</a></code> only serializes the part of an object that resides in memory<a href="#fn20" class="footnote-ref" id="fnref20"><sup>20</sup></a>. For most objects in R, that’s the whole object, so <code><a href="https://rdrr.io/r/base/readRDS.html">saveRDS()</a></code> does the job.</p>
<p>However some objects are pointing to on-disk data. For example:</p>
<ul>
<li>A <em>TxDb</em> object from the <em><a href="https://bioconductor.org/packages/3.12/GenomicFeatures">GenomicFeatures</a></em> points to an SQLite db</li>
<li>An <em>HDF5Array</em> object points to a dataset in an HDF5 file</li>
<li>A <em>SummarizedExperiment</em> derivative can have one or more of its assays that point to datasets (one per assay) in an HDF5 file.</li>
</ul>
<p>These objects have 2 parts: one part is in memory, and one part is on disk. The 1st part is sometimes called the object shell and is generally thin (i.e. it has a small memory footprint). The 2nd part is the data and is typically big. The object shell and data are linked together via some kind of pointer stored in the shell (e.g. an SQLite connection, or a path to a file, etc.). Note that this is a one way link in the sense that the object shell “knows” where to find the on-disk data but the on-disk data knows nothing about the object shell (and is completely agnostic about what kind of object shell could be pointing to it). Furthermore, at any given time on a given system, there could be more than one object shell pointing to the same on-disk data. These object shells could exist in the same R session or in sessions in other languages (e.g. Python). These various sessions could be run by the same or by different users.</p>
<p>Using <code><a href="https://rdrr.io/r/base/readRDS.html">saveRDS()</a></code> on such object will only serialize the shell part so will produce a small <code>.rds</code> file that contains the serialized object shell but not the object data.</p>
<p>This is problematic because:</p>
<ol style="list-style-type: decimal">
<li>If you later unserialize the object (with <code><a href="https://rdrr.io/r/base/readRDS.html">readRDS()</a></code>) on the same system where you originally serialized it, it is possible that you will get back an object that is fully functional and semantically equivalent to the original object. But here is the catch: this will be the case <strong>ONLY</strong> if the data is still at the original location and has not been modified (i.e. nobody wrote or altered the data in the SQLite db or HDF5 file in the mean time), and if the serialization/unserialization cycle didn’t break the link between the object shell and the data (this serialization/unserialization cycle is known to break open SQLite connections).</li>
<li>After serialization the object shell and data are stored in separate files (in the new <code>.rds</code> file for the shell, still in the original SQLite or HDF5 file for the data), typically in very different places on the file system. But these 2 files are not relocatable, that is, moving or copying them to another system or sending them to collaborators will typically break the link between them. Concretely this means that the object obtained by using <code><a href="https://rdrr.io/r/base/readRDS.html">readRDS()</a></code> on the destination system will be broken.</li>
</ol>
<p><code>saveHDF5SummarizedExperiment()</code> addresses these issues by saving the object shell and assay data in a folder that is relocatable.</p>
<p>Note that it only works on <em>SummarizedExperiment</em> derivatives. What it does exactly is:</p>
<ol style="list-style-type: decimal">
<li>Write all the assay data to an HDF5 file</li>
<li>Serialize the object shell, which in this case is everything in the object that is not the assay data.</li>
</ol>
<p>The 2 files (HDF5 and <code>.rds</code>) are written to the directory specified by the user. The resulting directory contains a full representation of the object and is relocatable, that is, it can be moved or copied to another place on the system, or to another system (possibly after making a tarball of it), where <code>loadHDF5SummarizedExperiment()</code> can then be used to load the object back in R.</p>
</div>
</div>
<div id="avoid-subsetting-too-much-and-random-access-patterns" class="section level3">
<h3 class="hasAnchor">
<a href="#avoid-subsetting-too-much-and-random-access-patterns" class="anchor"></a>Avoid subsetting too much and random access patterns</h3>
<p>Reordering/subsetting the data may degrade the performance of even seemingly simple operations. This is especially true of HDF5-backed data, where performance is best when reading contiguous chunks of data and worst when having to read data with a random access pattern.</p>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu">counts</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="no">x</span>[<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">x</span>)), <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">x</span>))]
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">x</span>))
<span class="co">#&gt; Processing block 1/9 ... OK</span>
<span class="co">#&gt; Processing block 2/9 ... OK</span>
<span class="co">#&gt; Processing block 3/9 ... OK</span>
<span class="co">#&gt; Processing block 4/9 ... OK</span>
<span class="co">#&gt; Processing block 5/9 ... OK</span>
<span class="co">#&gt; Processing block 6/9 ... OK</span>
<span class="co">#&gt; Processing block 7/9 ... OK</span>
<span class="co">#&gt; Processing block 8/9 ... OK</span>
<span class="co">#&gt; Processing block 9/9 ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   7.437   0.817   8.257</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">y</span>))
<span class="co">#&gt; Processing block 1/6 ... OK</span>
<span class="co">#&gt; Processing block 2/6 ... OK</span>
<span class="co">#&gt; Processing block 3/6 ... OK</span>
<span class="co">#&gt; Processing block 4/6 ... OK</span>
<span class="co">#&gt; Processing block 5/6 ... OK</span>
<span class="co">#&gt; Processing block 6/6 ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;  15.315   1.035  16.353</span></pre></body></html></div>
</div>
<div id="process-save-repeat" class="section level3">
<h3 class="hasAnchor">
<a href="#process-save-repeat" class="anchor"></a>Process, save, repeat</h3>
<p>When analysing large datasets, a workflow that is broken up into stages and saves the intermediate outputs can be help preserve ones sanity. This is true regardless of whether the DelayedArray framework is used: it sucks having to repeat some long pre-processing computation in order to make a quick plot. But it is especially true for DelayedArray-backed analyses where the accumulation of delayed operations will eventually lead to degraded performance. This means using <code>saveHDF5SummarizedExperiment()</code>/<code>loadHDF5SummarizedExperiment()</code>, or <code><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize()</a></code>-ing the result if there is sufficient memory available, following time consuming processing of the object.</p>
</div>
<div id="pragmatism-rules" class="section level3">
<h3 class="hasAnchor">
<a href="#pragmatism-rules" class="anchor"></a>Pragmatism rules</h3>
<p>We’ve seen some examples of pragmatism in this workshop: just load the data into memory, compute the thing you need, and move on. For the workshop, I used it to ‘cheat’ in order to speed things up, but it is frequently a valid strategy when analysing data!</p>
<p>For example, a normalized scRNA-seq dataset carries around two matrices: the raw counts and the normalized expression values. You might have enough RAM to load one of these a time but not both at once. With a HDF5-backed <em>SingleCellExperiment</em> you can easily just load into memory the matrix you actually need at a given step in the analysis.</p>
<p>Another example, with WGBS data you carry around 3 very large matrices. But to make a plot of methylation values along a gene promoter, a common requirement, you only need to load in a small ‘slice’ of one of these matrices. With a HDF5-backed <em>BSseq</em><a href="#fn21" class="footnote-ref" id="fnref21"><sup>21</sup></a> object you can quickly do this.</p>
<p>This brings us to a perhaps underappreciated advantage of using HDF5-backed <em>SummarizedExperiment</em> derivatives, namely that loading the saved data with <code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html">HDF5Array::loadHDF5SummarizedExperiment()</a></code> is really fast. I made extensive use of this when processing large WGBS datasets as I could quickly load the <em>BSseq</em> object to compute summaries of the sample metadata (stored in the <code>colData</code> of the <em>BSseq</em> object), a process that used to take tens of minutes to hours because the 3 large matrices also had to be loaded into memory. This has been so useful to me that I now keep even ‘small’ WGBS datasets as HDF5-backed <em>BSseq</em> objects.</p>
<p>This pragmatism has served me well. Often times I find myself starting with a very large dataset, do my initial processing using the disk-backed data, and run <code>saveHDF5SummarizedExperiment()</code> to produce a ‘pristine’ object. I can then move it to a large-memory machine, do my big computation (e.g., PCA), save the result, and <strong>move on with my life</strong>.</p>
</div>
<div id="avoid-degrading-to-a-delayedarray" class="section level3">
<h3 class="hasAnchor">
<a href="#avoid-degrading-to-a-delayedarray" class="anchor"></a>Avoid ‘degrading’ to a <em>DelayedArray</em>
</h3>
<p>The DelayedArray framework is implemented using the S4 object oriented system. This can be used to write methods that are optimized for a particular backend. For example, we might write a <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMaxs()</a></code> method that is optimized for the <em>TENxMatrix</em> class by exploiting the sparse storage mode of the underlying data. In order for <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMaxs()</a></code> to ‘know’ that it can use this optimized method, however, it needs for the data to be supplied as a <em>TENxMatrix</em> instance.</p>
<p>Unfortunately, it is very easy to ‘degrade’ a specialised <em>DelayedArray</em> derivative to a <em>DelayedArray</em>.</p>
<div class="sourceCode" id="cb61"><html><body><pre class="r"><span class="co"># Let's create a TENxMatrix</span>
<span class="no">tenx_matrix</span> <span class="kw">&lt;-</span> <span class="fu">as</span>(<span class="fu">counts</span>(<span class="no">sce</span>), <span class="st">"TENxMatrix"</span>)
<span class="co">#&gt; Realizing block 1/6 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/6 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 3/6 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 4/6 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 5/6 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 6/6 ... OK, writing it ... OK</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_matrix</span>)
<span class="co">#&gt; [1] "TENxMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span>

<span class="co"># NOTE: All these degrade the result to a DelayedMatrix.</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tenx_matrix</span>[, <span class="fl">1</span>:<span class="fl">10</span>])
<span class="co">#&gt; [1] "DelayedMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "DelayedArray"</span>
<span class="no">val</span> <span class="kw">&lt;-</span> <span class="no">tenx_matrix</span>
<span class="fu"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>(<span class="no">val</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"R"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">val</span>))), <span class="kw">NULL</span>)
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">val</span>)
<span class="co">#&gt; [1] "DelayedMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "DelayedArray"</span></pre></body></html></div>
<p>A common scenario where this degrading may occur is when extracting the data from a <em>SummarizedExperiment</em>.</p>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="fu">SingleCellExperiment</span>(
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">tenx_matrix</span>),
  <span class="kw">colData</span> <span class="kw">=</span> <span class="fu">DataFrame</span>(<span class="kw">row.names</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"S"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_matrix</span>)))))
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="fu">assay</span>(<span class="no">tmp</span>))
<span class="co">#&gt; [1] "DelayedMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "DelayedArray"</span></pre></body></html></div>
<p>What’s happened here? By default, <code>assay(tmp)</code> calls <code>assay(tmp, withDimnames = TRUE)</code> which has the effect of copying the dimnames from the <em>SummarizedExperiment</em> and adding them to the return assay data. As we saw above, setting the dimnames on a <em>TENxMatrix</em> (or other <em>DelayedMatrix</em> derivative) will degrade it to a <em>DelayedMatrix</em>. Consequently, running <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colMaxs(assay(tmp))</a></code> will not call our (hypothetical) optimized method for <em>TENxMatrix</em> objects and will instead defer to the slower, more general block-processing method that is implemented for <em>DelayedMatrix</em> objects.</p>
<p>To avoid this ‘degrading upon assay extraction’, we can should set <code>withDimnames = FALSE</code>.</p>
<div class="sourceCode" id="cb63"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="fu">assay</span>(<span class="no">tmp</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>))
<span class="co">#&gt; [1] "TENxMatrix"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "HDF5Array"</span></pre></body></html></div>
<p>More generally, you may need to avoid degrading a <em>DelayedArray</em> derivative to a <em>DelayedArray</em> in order to use backend-optimized methods.</p>
</div>
<div id="block-size" class="section level3">
<h3 class="hasAnchor">
<a href="#block-size" class="anchor"></a>Block size</h3>
<p>The maximum size of a block used when performing block-processing is given by <code><a href="https://rdrr.io/pkg/DelayedArray/man/blockGrid.html">getAutoBlockSize()</a></code>. By default, this is set to 100000000 meaning each block can use up to 100000000 / <code>1e6</code> = 100 Mb of data.</p>
<p>Using fewer, larger blocks generally means faster performance (at the cost of higher peak memory usage), so we may wish to increase this on machines with sufficient memory by using <code><a href="https://rdrr.io/pkg/DelayedArray/man/blockGrid.html">setAutoBlockSize()</a></code>.</p>
<div class="sourceCode" id="cb64"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="fu">counts</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)))
<span class="co">#&gt; Processing block 1/9 ... OK</span>
<span class="co">#&gt; Processing block 2/9 ... OK</span>
<span class="co">#&gt; Processing block 3/9 ... OK</span>
<span class="co">#&gt; Processing block 4/9 ... OK</span>
<span class="co">#&gt; Processing block 5/9 ... OK</span>
<span class="co">#&gt; Processing block 6/9 ... OK</span>
<span class="co">#&gt; Processing block 7/9 ... OK</span>
<span class="co">#&gt; Processing block 8/9 ... OK</span>
<span class="co">#&gt; Processing block 9/9 ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   6.561   0.800   7.364</span>

<span class="co"># NOTE: Increasing the block size 10-fold.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/blockGrid.html">setAutoBlockSize</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/blockGrid.html">getAutoBlockSize</a></span>() * <span class="fl">10</span>)
<span class="co">#&gt; automatic block size set to 1e+09 bytes (was 1e+08)</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="fu">counts</span>(<span class="no">sce</span>, <span class="kw">withDimnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)))
<span class="co">#&gt; Processing block 1/1 ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   8.488   1.002   9.491</span>

<span class="co"># NOTE: Revert to default block size.</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/blockGrid.html">setAutoBlockSize</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/blockGrid.html">getAutoBlockSize</a></span>() / <span class="fl">10</span>)
<span class="co">#&gt; automatic block size set to 1e+08 bytes (was 1e+09)</span></pre></body></html></div>
</div>
<div id="chunking" class="section level3">
<h3 class="hasAnchor">
<a href="#chunking" class="anchor"></a>Chunking</h3>
<p>Data stored in an HDF5 file are usually ‘chunked’ into sub-matrices (for matrix data) or hyper-cubes (for arrays of higher dimension)<a href="#fn22" class="footnote-ref" id="fnref22"><sup>22</sup></a>. For example, we could choose to chunk an <span class="math inline">\(R \times C\)</span> matrix by column, by row, or into <span class="math inline">\(r \times c\)</span> sub-matrices (<span class="math inline">\(r \leq R, c \leq C\)</span>). This may remind you of the choice of block dimensions used in block-processing. The difference is this:</p>
<blockquote>
<p><strong>Block dimensions dictate how the data are accessed, chunk dimensions dictate how the data are stored</strong>.</p>
</blockquote>
<p>In general, you want your data to be chunked in a manner that supports the type of access patterns you will be making. For example, if you know you only need to access data by column then chunk the data by column. Of course, you often either don’t know in advance what access patterns you need or you need both row and column access. In that case, chunking into sub-matrices offers the best tradeoff.</p>
<p>We’ll demonstrate how chunking can affect performance by comparing computing the column sums of a column-chunked and row-chunked dataset.</p>
<div class="sourceCode" id="cb65"><html><body><pre class="r"><span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1e8</span>), <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">1e2</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">1e6</span>)
<span class="no">x_col</span> <span class="kw">&lt;-</span> <span class="fu">writeHDF5Array</span>(<span class="no">x</span>, <span class="kw">chunkdim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">x</span>), <span class="fl">1</span>))
<span class="co">#&gt; Realizing block 1/4 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/4 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 3/4 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 4/4 ... OK, writing it ... OK</span>
<span class="no">x_row</span> <span class="kw">&lt;-</span> <span class="fu">writeHDF5Array</span>(<span class="no">x</span>, <span class="kw">chunkdim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">x</span>)))
<span class="co">#&gt; Realizing block 1/4 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 2/4 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 3/4 ... OK, writing it ... OK</span>
<span class="co">#&gt; Realizing block 4/4 ... OK, writing it ... OK</span>

<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">x_col</span>))
<span class="co">#&gt; Processing block 1/4 ... OK</span>
<span class="co">#&gt; Processing block 2/4 ... OK</span>
<span class="co">#&gt; Processing block 3/4 ... OK</span>
<span class="co">#&gt; Processing block 4/4 ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   5.113   0.469   5.583</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums</a></span>(<span class="no">x_row</span>))
<span class="co">#&gt; Processing block 1/4 ... OK</span>
<span class="co">#&gt; Processing block 2/4 ... OK</span>
<span class="co">#&gt; Processing block 3/4 ... OK</span>
<span class="co">#&gt; Processing block 4/4 ... OK</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   7.623   1.131   8.757</span></pre></body></html></div>
<p>Chunking can be controlled by the <code>chunkdim</code> argument when writing data to disk using the <code>writeHDF5Array()</code> or <code>saveHDF5SummarizedExperiment()</code> functions.</p>
</div>
<div id="parallelization-strategies" class="section level3">
<h3 class="hasAnchor">
<a href="#parallelization-strategies" class="anchor"></a>Parallelization strategies</h3>
<p>Several of the routines we used in today’s workshop have ‘native’ parallelization via the <em><a href="https://bioconductor.org/packages/3.12/BiocParallel">BiocParallel</a></em> package. Look out for the <code>BPPARAM</code> argument in function documentation<a href="#fn23" class="footnote-ref" id="fnref23"><sup>23</sup></a>.</p>
<p>Fair warning: parallelization is tricky and performance may not match expectations.</p>
</div>
<div id="whats-in-my-hdf5-file" class="section level3">
<h3 class="hasAnchor">
<a href="#whats-in-my-hdf5-file" class="anchor"></a>What’s in my HDF5 file?</h3>
<p>I do this quite a bit: I’ve got an HDF5 file but I don’t know (or have forgotten) what’s in it. <code><a href="https://rdrr.io/pkg/rhdf5/man/h5ls.html">rhdf5::h5ls()</a></code> to the rescue.</p>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="kw pkg">rhdf5</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5ls.html">h5ls</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">path</a></span>(<span class="fu">counts</span>(<span class="no">sce</span>)))
<span class="co">#&gt;   group   name       otype  dclass          dim</span>
<span class="co">#&gt; 0     / counts H5I_DATASET INTEGER 33694 x 4340</span></pre></body></html></div>
</div>
</div>
<div id="concluding-remarks-1" class="section level2">
<h2 class="hasAnchor">
<a href="#concluding-remarks-1" class="anchor"></a>Concluding remarks</h2>
<p>We’ve now spent nearly 2 hours learning about the DelayedArray framework. Now comes the kicker:</p>
<blockquote>
<p><strong>Don’t use a <em>DelayedArray</em> if you can avoid it!</strong></p>
</blockquote>
<p>I write that as someone who has a professional love for the DelayedArray framework. I’ve spent many, many hours developing software that supports and extends it, leveraging it to analyse large WGBS and scRNA-seq datasets, and writing and presenting talks and workshops to teach it. But the simple fact remains:</p>
<blockquote>
<p><strong>If you can load your data into memory and still compute on it then you’re always going to have a better time doing it that way.</strong></p>
</blockquote>
<p>Analyses will be faster, simpler, and you will have more options available to you. But when this isn’t an option then the DelayedArray framework is a powerful set of packages to help you get your work done. I find it pretty remarkable that a first-class single-cell analysis workflow can so seamlessly support the use of in-memory and disk-backed data.</p>
<p>A huge amount of gratitude is owed to Hervé Pagès, the developer of the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package and the broader framework, and Aaron Lun, the main developer many of the key packages for analysing scRNA-seq data and the <em><a href="https://bioconductor.org/packages/3.12/simpleSingleCell">simpleSingleCell</a></em> workflow that I cribbed from in <a href="#real-world-encounter-with-delayedarray-analysing-scrna-seq-data">Real world encounter with DelayedArray analysing scRNA-seq data</a>. Thank you!</p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<p>This document was prepared using the following software:</p>
<div class="sourceCode" id="cb67"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.0 (2020-04-24)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.4 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] scater_1.17.1               ggplot2_3.3.1              </span>
<span class="co">#&gt;  [3] scuttle_0.99.8              scran_1.17.1               </span>
<span class="co">#&gt;  [5] TENxPBMCData_1.7.0          DelayedMatrixStats_1.11.0  </span>
<span class="co">#&gt;  [7] TENxBrainData_1.9.0         HDF5Array_1.17.0           </span>
<span class="co">#&gt;  [9] rhdf5_2.33.0                SingleCellExperiment_1.11.2</span>
<span class="co">#&gt; [11] SummarizedExperiment_1.19.4 Biobase_2.49.0             </span>
<span class="co">#&gt; [13] GenomicRanges_1.41.3        GenomeInfoDb_1.25.1        </span>
<span class="co">#&gt; [15] DelayedArray_0.15.1         IRanges_2.23.7             </span>
<span class="co">#&gt; [17] S4Vectors_0.27.11           BiocGenerics_0.35.3        </span>
<span class="co">#&gt; [19] matrixStats_0.56.0         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] ggbeeswarm_0.6.0              colorspace_1.4-1             </span>
<span class="co">#&gt;   [3] ellipsis_0.3.1                rprojroot_1.3-2              </span>
<span class="co">#&gt;   [5] XVector_0.29.1                BiocNeighbors_1.7.0          </span>
<span class="co">#&gt;   [7] fs_1.4.1                      DT_0.13                      </span>
<span class="co">#&gt;   [9] gh_1.1.0                      bit64_0.9-7                  </span>
<span class="co">#&gt;  [11] interactiveDisplayBase_1.27.5 AnnotationDbi_1.51.0         </span>
<span class="co">#&gt;  [13] fansi_0.4.1                   xml2_1.3.2                   </span>
<span class="co">#&gt;  [15] knitr_1.28                    jsonlite_1.6.1               </span>
<span class="co">#&gt;  [17] dbplyr_1.4.4                  graph_1.67.1                 </span>
<span class="co">#&gt;  [19] shiny_1.4.0.2                 BiocManager_1.30.10          </span>
<span class="co">#&gt;  [21] readr_1.3.1                   compiler_4.0.0               </span>
<span class="co">#&gt;  [23] httr_1.4.1                    dqrng_0.2.1                  </span>
<span class="co">#&gt;  [25] backports_1.1.7               assertthat_0.2.1             </span>
<span class="co">#&gt;  [27] Matrix_1.2-18                 fastmap_1.0.1                </span>
<span class="co">#&gt;  [29] limma_3.45.0                  cli_2.0.2                    </span>
<span class="co">#&gt;  [31] later_1.0.0                   BiocSingular_1.5.0           </span>
<span class="co">#&gt;  [33] htmltools_0.4.0               tools_4.0.0                  </span>
<span class="co">#&gt;  [35] rsvd_1.0.3                    igraph_1.2.5                 </span>
<span class="co">#&gt;  [37] gtable_0.3.0                  glue_1.4.1                   </span>
<span class="co">#&gt;  [39] GenomeInfoDbData_1.2.3        dplyr_1.0.0                  </span>
<span class="co">#&gt;  [41] rappdirs_0.3.1                Rcpp_1.0.4.6                 </span>
<span class="co">#&gt;  [43] pkgdown_1.5.1                 vctrs_0.3.0                  </span>
<span class="co">#&gt;  [45] ExperimentHub_1.15.0          xfun_0.14                    </span>
<span class="co">#&gt;  [47] stringr_1.4.0                 beachmat_2.5.0               </span>
<span class="co">#&gt;  [49] rvest_0.3.5                   mime_0.9                     </span>
<span class="co">#&gt;  [51] lifecycle_0.2.0               irlba_2.3.3                  </span>
<span class="co">#&gt;  [53] statmod_1.4.34                XML_3.99-0.3                 </span>
<span class="co">#&gt;  [55] AnnotationHub_2.21.0          edgeR_3.31.3                 </span>
<span class="co">#&gt;  [57] scales_1.1.1                  zlibbioc_1.35.0              </span>
<span class="co">#&gt;  [59] MASS_7.3-51.5                 BiocPkgTools_1.7.0           </span>
<span class="co">#&gt;  [61] BiocStyle_2.17.0              hms_0.5.3                    </span>
<span class="co">#&gt;  [63] promises_1.1.0                rex_1.2.0                    </span>
<span class="co">#&gt;  [65] RBGL_1.65.0                   yaml_2.2.1                   </span>
<span class="co">#&gt;  [67] curl_4.3                      gridExtra_2.3                </span>
<span class="co">#&gt;  [69] memoise_1.1.0                 stringi_1.4.6                </span>
<span class="co">#&gt;  [71] RSQLite_2.2.0                 BiocVersion_3.12.0           </span>
<span class="co">#&gt;  [73] desc_1.2.0                    BiocParallel_1.23.0          </span>
<span class="co">#&gt;  [75] rlang_0.4.6                   pkgconfig_2.0.3              </span>
<span class="co">#&gt;  [77] bitops_1.0-6                  evaluate_0.14                </span>
<span class="co">#&gt;  [79] lattice_0.20-41               purrr_0.3.4                  </span>
<span class="co">#&gt;  [81] Rhdf5lib_1.11.0               htmlwidgets_1.5.1            </span>
<span class="co">#&gt;  [83] bit_1.1-15.2                  tidyselect_1.1.0             </span>
<span class="co">#&gt;  [85] magrittr_1.5                  R6_2.4.1                     </span>
<span class="co">#&gt;  [87] generics_0.0.2                RUnit_0.4.32                 </span>
<span class="co">#&gt;  [89] DBI_1.1.0                     withr_2.2.0                  </span>
<span class="co">#&gt;  [91] pillar_1.4.4                  biocViews_1.57.0             </span>
<span class="co">#&gt;  [93] RCurl_1.98-1.2                tibble_3.0.1                 </span>
<span class="co">#&gt;  [95] crayon_1.3.4                  BiocFileCache_1.13.0         </span>
<span class="co">#&gt;  [97] rmarkdown_2.2                 viridis_0.5.1                </span>
<span class="co">#&gt;  [99] locfit_1.5-9.4                grid_4.0.0                   </span>
<span class="co">#&gt; [101] blob_1.2.1                    digest_0.6.25                </span>
<span class="co">#&gt; [103] xtable_1.8-4                  tidyr_1.1.0                  </span>
<span class="co">#&gt; [105] httpuv_1.5.3.1                munsell_0.5.0                </span>
<span class="co">#&gt; [107] viridisLite_0.3.0             beeswarm_0.2.3               </span>
<span class="co">#&gt; [109] vipor_0.4.5</span></pre></body></html></div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-lun2016pooling">
<p>Lun, A. T., K. Bach, and J. C. Marioni. 2016. “Pooling across cells to normalize single-cell RNA sequencing data with many zero counts.” <em>Genome Biol.</em> 17 (April): 75.</p>
</div>
<div id="ref-rizzardi2019neuronal">
<p>Rizzardi, Lindsay F, Peter F Hickey, Varenka Rodriguez DiBlasi, Rakel Tryggvadóttir, Colin M Callahan, Adrian Idrizi, Kasper D Hansen, and Andrew P Feinberg. 2019. “Neuronal Brain-Region-Specific Dna Methylation and Chromatin Accessibility Are Associated with Neuropsychiatric Trait Heritability.” <em>Nature Neuroscience</em> 22 (2). Nature Publishing Group: 307.</p>
</div>
<div id="ref-zheng2017massively">
<p>Zheng, G. X., J. M. Terry, P. Belgrader, P. Ryvkin, Z. W. Bent, R. Wilson, S. B. Ziraldo, et al. 2017. “Massively parallel digital transcriptional profiling of single cells.” <em>Nat Commun</em> 8 (January): 14049.</p>
</div>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Including space-efficient formats like sparse arrays and run length encoded arrays<a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p>Higher-dimensional arrays may be appropriate for some types of assays.<a href="#fnref2" class="footnote-back">↩</a></p></li>
<li id="fn3"><p>In R, a <em>matrix</em> is just a 2-dimensional <em>array</em><a href="#fnref3" class="footnote-back">↩</a></p></li>
<li id="fn4"><p>You may have seen similar pretty printing with other Bioconductor objects such as <em>GRanges</em> and <em>DataFrame</em> or with the <em>data.table</em> and <em>tibble</em> extensions to the <em>data.frame</em>. I can’t say enough how much I appreciate these thoughtful touches when doing interactive data analysis.<a href="#fnref4" class="footnote-back">↩</a></p></li>
<li id="fn5"><p>As with a 2-dimensional <em>array</em> in base R being commonly known as a <em>matrix</em>, a 2-dimensional <em>DelayedArray</em> is also known as a <em>DelayedMatrix</em> and a 2-dimensional <em>HDF5Array</em> is also known as a <em>HDF5Matrix</em>.<a href="#fnref5" class="footnote-back">↩</a></p></li>
<li id="fn6"><p>We will revisit the <code>chunkdim</code> component of the <em>HDF5Matrix</em> object later in the workshop in <a href="#chunking">Chunking</a><a href="#fnref6" class="footnote-back">↩</a></p></li>
<li id="fn7"><p>The technical names of each type of delayed operation are not important.<a href="#fnref7" class="footnote-back">↩</a></p></li>
<li id="fn8"><p>The <a href="#no-op">No-op</a> example is the obvious exception<a href="#fnref8" class="footnote-back">↩</a></p></li>
<li id="fn9"><p>Here, I use the ‘namespaced function’ notation for didactic purposes to distinguish <code><a href="https://rdrr.io/r/base/colSums.html">base::colSums()</a></code> from the S4 generic <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-stats.html">colSums()</a></code> and its associated methods. When writing code, however, it is generally not necessary to use the <code>package::function()</code> notation, except for all the times that it is …<a href="#fnref9" class="footnote-back">↩</a></p></li>
<li id="fn10"><p>For example, recall that the <code>tenx_counts</code> data would occupy &gt; 146 GB in memory.<a href="#fnref10" class="footnote-back">↩</a></p></li>
<li id="fn11"><p>In the following code example we could replace <code><a href="https://rdrr.io/r/base/array.html">as.array(colSums(tenx_counts_subset[, j, drop = FALSE]))</a></code> with <code><a href="https://rdrr.io/r/base/sum.html">sum(tenx_counts_subset[, j, drop = TRUE])</a></code> or, more simply, <code><a href="https://rdrr.io/r/base/sum.html">sum(tenx_counts_subset[, j])</a></code>. This is because subsetting a <em>DelayedArray</em> returns an ordinary vector when <code>drop = TRUE</code> and the result has only one dimension; see ‘Subsetting’ in <code><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-class.html">help("DelayedArray", package = "DelayedArray")</a></code>.<a href="#fnref11" class="footnote-back">↩</a></p></li>
<li id="fn12"><p>This may seem artificial, but suppose we were dealing with a very ‘tall’ matrix.<a href="#fnref12" class="footnote-back">↩</a></p></li>
<li id="fn13"><p>In the above example it’s safe because <code>tenx_counts_subset_realized</code> only requires around 100 Mb of memory.<a href="#fnref13" class="footnote-back">↩</a></p></li>
<li id="fn14"><p>The <em><a href="https://bioconductor.org/packages/3.12/TENxBrainData">TENxBrainData</a></em> package depends upon <em><a href="https://bioconductor.org/packages/3.12/HDF5Array">HDF5Array</a></em>, so it is already installed. If you’ve been following this workflow then you’ve also already attached it to the search path by running <code><a href="https://rdrr.io/pkg/TENxBrainData/man">library(TENxBrainData)</a></code>.<a href="#fnref14" class="footnote-back">↩</a></p></li>
<li id="fn15"><p>We’ll discuss why you might want this in the second half of the workshop.<a href="#fnref15" class="footnote-back">↩</a></p></li>
<li id="fn16"><p>In fact, if you use any package that makes use of the <em>SummarizedExperiment</em> class, then you will almost certainly load the <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em> package during the course of your analysis, whether you know it or not! This is because <em><a href="https://bioconductor.org/packages/3.12/SummarizedExperiment">SummarizedExperiment</a></em> depends upon <em><a href="https://bioconductor.org/packages/3.12/DelayedArray">DelayedArray</a></em>.<a href="#fnref16" class="footnote-back">↩</a></p></li>
<li id="fn17"><p>As listed in the <code>Depends</code> field of the package <code>DESCRIPTION</code> file.<a href="#fnref17" class="footnote-back">↩</a></p></li>
<li id="fn18"><p>This is adapted from the <em><a href="https://bioconductor.org/packages/3.12/simpleSingleCell">simpleSingleCell</a></em> workflow.<a href="#fnref18" class="footnote-back">↩</a></p></li>
<li id="fn19"><p>For the sake of time, we’re going to cheat by loading the counts matrix into memory as a sparse matrix and computing the clustering using the in-memory data. But this functionality does work, albeit more slowly, for the HDF5-backed data.<a href="#fnref19" class="footnote-back">↩</a></p></li>
<li id="fn20"><p>The reality is a little bit more nuanced, but discussing the full details is not important here, and would only distract us<a href="#fnref20" class="footnote-back">↩</a></p></li>
<li id="fn21"><p>A <em>SummarizedExperiment</em> derivative<a href="#fnref21" class="footnote-back">↩</a></p></li>
<li id="fn22"><p>Chunking also applies to other storage formats, such as <em>RleArray</em> objects, but we’ll focus on the HDF5 format.<a href="#fnref22" class="footnote-back">↩</a></p></li>
<li id="fn23"><p>Unfortunately, simply typing the function name at the command prompt or using <code><a href="https://rdrr.io/r/base/args.html">args()</a></code> may not be sufficient to identify these functions. E.g., <code><a href="https://rdrr.io/r/base/args.html">args(denoisePCA)</a></code> doesn’t include <code>BPPARAM</code> but <code><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html">help("denoisePCA", "scran")</a></code> tells you this is an option (Footnote to the footnote: this is because <code>denoicePCA()</code> is an S4 generic with methods defined only for the first argument, <code>x</code>. The methods include the <code>BPPARAM</code> argument and are documented as such.)<a href="#fnref23" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Peter Hickey.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
